<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>PyProteinInference - Py Protein Inference</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "PyProteinInference";
        var mkdocs_page_input_path = "pyproteininference_reference.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Py Protein Inference
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../quickstart/">Quick Start</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../parameters/">Parameters</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../input_format/">Input Formats</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../advanced/">Advanced Usage</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../supplementary/">Supplementary Information</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Code Reference</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">PyProteinInference</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference">pyproteininference</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.datastore">datastore</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.datastore.DataStore">DataStore</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.calculate_q_values">calculate_q_values()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.check_data_consistency">check_data_consistency()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.create_scoring_input">create_scoring_input()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.exclude_non_distinguishing_peptides">exclude_non_distinguishing_peptides()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.generate_fdr_vs_target_hits">generate_fdr_vs_target_hits()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_pep_values">get_pep_values()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_data">get_protein_data()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_identifiers">get_protein_identifiers()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_identifiers_from_psm_data">get_protein_identifiers_from_psm_data()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_information">get_protein_information()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_information_dictionary">get_protein_information_dictionary()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_protein_objects">get_protein_objects()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_psm_data">get_psm_data()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_q_values">get_q_values()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.get_sorted_identifiers">get_sorted_identifiers()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.higher_or_lower">higher_or_lower()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.input_has_custom">input_has_custom()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.input_has_pep">input_has_pep()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.input_has_q">input_has_q()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.peptide_to_protein_dictionary">peptide_to_protein_dictionary()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.protein_picker">protein_picker()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.protein_to_peptide_dictionary">protein_to_peptide_dictionary()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.sort_protein_group_objects">sort_protein_group_objects()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.sort_protein_objects">sort_protein_objects()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.sort_protein_strings">sort_protein_strings()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.sort_protein_sub_groups">sort_protein_sub_groups()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.unique_to_leads_peptides">unique_to_leads_peptides()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.validate_digest">validate_digest()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.datastore.DataStore.validate_psm_data">validate_psm_data()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.export">export</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.export.Export">Export</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_all_restricted">csv_export_all_restricted()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_comma_sep_restricted">csv_export_comma_sep_restricted()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_leads_restricted">csv_export_leads_restricted()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_all">csv_export_q_value_all()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_comma_sep">csv_export_q_value_comma_sep()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_leads">csv_export_q_value_leads()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_leads_long">csv_export_q_value_leads_long()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_leads_peptides">csv_export_q_value_leads_peptides()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_leads_psm_ids">csv_export_q_value_leads_psm_ids()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.csv_export_q_value_leads_psms">csv_export_q_value_leads_psms()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.export.Export.export_to_csv">export_to_csv()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.heuristic">heuristic</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline">HeuristicPipeline</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.determine_number_stdev_from_mean">determine_number_stdev_from_mean()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.determine_optimal_inference_method">determine_optimal_inference_method()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.execute">execute()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.generate_density_plot">generate_density_plot()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.heuristic.HeuristicPipeline.generate_roc_plot">generate_roc_plot()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.in_silico_digest">in_silico_digest</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.in_silico_digest.Digest">Digest</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.in_silico_digest.PyteomicsDigest">PyteomicsDigest</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.in_silico_digest.PyteomicsDigest.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.in_silico_digest.PyteomicsDigest.digest_fasta_database">digest_fasta_database()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.inference">inference</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.Exclusion">Exclusion</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Exclusion.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Exclusion.infer_proteins">infer_proteins()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.FirstProtein">FirstProtein</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.FirstProtein.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.FirstProtein.infer_proteins">infer_proteins()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.Inclusion">Inclusion</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Inclusion.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Inclusion.infer_proteins">infer_proteins()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.Inference">Inference</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Inference.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Inference.run_inference">run_inference()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.Parsimony">Parsimony</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Parsimony.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.Parsimony.infer_proteins">infer_proteins()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.inference.PeptideCentric">PeptideCentric</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.PeptideCentric.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.inference.PeptideCentric.infer_proteins">infer_proteins()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.parameters">parameters</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.convert_to_object">convert_to_object()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.fix_parameters_from_datastore">fix_parameters_from_datastore()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.override_custom_restrict">override_custom_restrict()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.override_pep_restrict">override_pep_restrict()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.override_q_restrict">override_q_restrict()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.parameters.ProteinInferenceParameter.validate_parameters">validate_parameters()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.physical">physical</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.physical.Protein">Protein</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_num_peptides">get_num_peptides()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_num_psms">get_num_psms()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_psm_identifiers">get_psm_identifiers()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_psm_ids">get_psm_ids()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_psm_scores">get_psm_scores()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_stripped_psm_identifiers">get_stripped_psm_identifiers()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_unique_peptide_identifiers">get_unique_peptide_identifiers()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Protein.get_unique_stripped_peptide_identifiers">get_unique_stripped_peptide_identifiers()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.ProteinGroup.__init__">__init__()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.physical.Psm">Psm</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Psm.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Psm.assign_main_score">assign_main_score()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Psm.remove_peptide_mods">remove_peptide_mods()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Psm.split_peptide">split_peptide()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.physical.Psm.split_peptide_pro">split_peptide_pro()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.pipeline">pipeline</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.pipeline.ProteinInferencePipeline">ProteinInferencePipeline</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.pipeline.ProteinInferencePipeline.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.pipeline.ProteinInferencePipeline.execute">execute()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.reader">reader</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.reader.GenericReader">GenericReader</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.GenericReader.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.GenericReader.read_psms">read_psms()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.reader.PercolatorReader">PercolatorReader</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.PercolatorReader.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.PercolatorReader.read_psms">read_psms()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.reader.ProteologicPostSearchReader">ProteologicPostSearchReader</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.ProteologicPostSearchReader.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.ProteologicPostSearchReader.read_psms">read_psms()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.reader.Reader">Reader</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.Reader.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.reader.Reader.get_alternative_proteins_from_input">get_alternative_proteins_from_input()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#pyproteininference.scoring">scoring</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#pyproteininference.scoring.Score">Score</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.__init__">__init__()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.additive">additive()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.best_peptide_per_protein">best_peptide_per_protein()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.down_weighted_multiplicative_log">down_weighted_multiplicative_log()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.down_weighted_v2">down_weighted_v2()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.fishers_method">fishers_method()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.geometric_mean_log">geometric_mean_log()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.iterative_down_weighted_log">iterative_down_weighted_log()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.iterative_down_weighted_v2">iterative_down_weighted_v2()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.multiplicative_log">multiplicative_log()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.score_psms">score_psms()</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#pyproteininference.scoring.Score.top_two_combied">top_two_combied()</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Py Protein Inference</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>Code Reference &raquo;</li><li>PyProteinInference</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>

          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="py-protein-inference-module">Py Protein Inference Module</h1>


  <div class="doc doc-object doc-module">

<a id="pyproteininference"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">










  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.datastore">
        <code>datastore</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.datastore.DataStore">
        <code>
DataStore        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following Class serves as the data storage object for a protein inference analysis
The class serves as a central point that is accessed at virtually every PI processing step</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>main_data_form</code></td>
        <td><code>list</code></td>
        <td><p>List of unrestricted Psm objects.</p></td>
      </tr>
      <tr>
        <td><code>parameter_file_object</code></td>
        <td><code>ProteinInferenceParameter</code></td>
        <td><p>protein inference parameter
<a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">object</a>.</p></td>
      </tr>
      <tr>
        <td><code>restricted_peptides</code></td>
        <td><code>list</code></td>
        <td><p>List of non flaking peptide strings present in the current analysis.</p></td>
      </tr>
      <tr>
        <td><code>main_data_restricted</code></td>
        <td><code>list</code></td>
        <td><p>List of restricted <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.
Restriction is based on the parameter_file_object and the object is created by function
    <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data</a>.</p></td>
      </tr>
      <tr>
        <td><code>scored_proteins</code></td>
        <td><code>list</code></td>
        <td><p>List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.
Output from scoring methods from <a class="autorefs autorefs-internal" href="#pyproteininference.scoring">scoring</a>.</p></td>
      </tr>
      <tr>
        <td><code>grouped_scored_proteins</code></td>
        <td><code>list</code></td>
        <td><p>List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a>
objects that have been grouped and sorted. Output from
    <a class="autorefs autorefs-internal" href="#pyproteininference.inference.Inference.run_inference">run_inference</a> method.</p></td>
      </tr>
      <tr>
        <td><code>scoring_input</code></td>
        <td><code>list</code></td>
        <td><p>List of non-scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.
Output from <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.create_scoring_input">create_scoring_input</a>.</p></td>
      </tr>
      <tr>
        <td><code>picked_proteins_scored</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects that pass
the protein picker algorithm (<a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.protein_picker">protein_picker</a>).</p></td>
      </tr>
      <tr>
        <td><code>picked_proteins_removed</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects that do not
pass the protein picker algorithm (<a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.protein_picker">protein_picker</a>).</p></td>
      </tr>
      <tr>
        <td><code>protein_peptide_dictionary</code></td>
        <td><code>collections.defaultdict</code></td>
        <td><p>Dictionary of protein strings (keys) that map to sets
of peptide strings based on the peptides and proteins found in the search. Protein -&gt; set(Peptides).</p></td>
      </tr>
      <tr>
        <td><code>peptide_protein_dictionary</code></td>
        <td><code>collections.defaultdict</code></td>
        <td><p>Dictionary of peptide strings (keys) that map to sets
of protein strings based on the peptides and proteins found in the search. Peptide -&gt; set(Proteins).</p></td>
      </tr>
      <tr>
        <td><code>high_low_better</code></td>
        <td><code>str</code></td>
        <td><p>Variable that indicates whether a higher or a lower protein score is better.
This is necessary to sort Protein objects by score properly. Can either be "higher" or "lower".</p></td>
      </tr>
      <tr>
        <td><code>psm_score</code></td>
        <td><code>str</code></td>
        <td><p>Variable that indicates the <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a>
score being used in the analysis to generate <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> scores.</p></td>
      </tr>
      <tr>
        <td><code>protein_score</code></td>
        <td><code>str</code></td>
        <td><p>String to indicate the protein score method used.</p></td>
      </tr>
      <tr>
        <td><code>short_protein_score</code></td>
        <td><code>str</code></td>
        <td><p>Short String to indicate the protein score method used.</p></td>
      </tr>
      <tr>
        <td><code>protein_group_objects</code></td>
        <td><code>list</code></td>
        <td><p>List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a>
objects that have been grouped and sorted. Output from
 <a class="autorefs autorefs-internal" href="#pyproteininference.inference.Inference.run_inference">run_inference</a> method.</p></td>
      </tr>
      <tr>
        <td><code>decoy_symbol</code></td>
        <td><code>str</code></td>
        <td><p>String that is used to differentiate between decoy proteins and target proteins. Ex: "##".</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p></td>
      </tr>
      <tr>
        <td><code>SCORE_MAPPER</code></td>
        <td><code>dict</code></td>
        <td><p>Dictionary that maps potential scores in input files to internal score names.</p></td>
      </tr>
      <tr>
        <td><code>CUSTOM_SCORE_KEY</code></td>
        <td><code>str</code></td>
        <td><p>String that indicates a custom score is being used.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">DataStore</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following Class serves as the data storage object for a protein inference analysis</span>
<span class="sd">    The class serves as a central point that is accessed at virtually every PI processing step</span>


<span class="sd">    Attributes:</span>
<span class="sd">        main_data_form (list): List of unrestricted Psm objects.</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter): protein inference parameter</span>
<span class="sd">            [object][pyproteininference.parameters.ProteinInferenceParameter].</span>
<span class="sd">        restricted_peptides (list): List of non flaking peptide strings present in the current analysis.</span>
<span class="sd">        main_data_restricted (list): List of restricted [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">            Restriction is based on the parameter_file_object and the object is created by function</span>
<span class="sd">                [restrict_psm_data][pyproteininference.datastore.DataStore.restrict_psm_data].</span>
<span class="sd">        scored_proteins (list): List of scored [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            Output from scoring methods from [scoring][pyproteininference.scoring].</span>
<span class="sd">        grouped_scored_proteins (list): List of scored [Protein][pyproteininference.physical.Protein]</span>
<span class="sd">            objects that have been grouped and sorted. Output from</span>
<span class="sd">                [run_inference][pyproteininference.inference.Inference.run_inference] method.</span>
<span class="sd">        scoring_input (list): List of non-scored [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            Output from [create_scoring_input][pyproteininference.datastore.DataStore.create_scoring_input].</span>
<span class="sd">        picked_proteins_scored (list): List of [Protein][pyproteininference.physical.Protein] objects that pass</span>
<span class="sd">            the protein picker algorithm ([protein_picker][pyproteininference.datastore.DataStore.protein_picker]).</span>
<span class="sd">        picked_proteins_removed (list): List of [Protein][pyproteininference.physical.Protein] objects that do not</span>
<span class="sd">            pass the protein picker algorithm ([protein_picker][pyproteininference.datastore.DataStore.protein_picker]).</span>
<span class="sd">        protein_peptide_dictionary (collections.defaultdict): Dictionary of protein strings (keys) that map to sets</span>
<span class="sd">            of peptide strings based on the peptides and proteins found in the search. Protein -&gt; set(Peptides).</span>
<span class="sd">        peptide_protein_dictionary (collections.defaultdict): Dictionary of peptide strings (keys) that map to sets</span>
<span class="sd">            of protein strings based on the peptides and proteins found in the search. Peptide -&gt; set(Proteins).</span>
<span class="sd">        high_low_better (str): Variable that indicates whether a higher or a lower protein score is better.</span>
<span class="sd">            This is necessary to sort Protein objects by score properly. Can either be &quot;higher&quot; or &quot;lower&quot;.</span>
<span class="sd">        psm_score (str): Variable that indicates the [Psm][pyproteininference.physical.Psm]</span>
<span class="sd">            score being used in the analysis to generate [Protein][pyproteininference.physical.Protein] scores.</span>
<span class="sd">        protein_score (str): String to indicate the protein score method used.</span>
<span class="sd">        short_protein_score (str): Short String to indicate the protein score method used.</span>
<span class="sd">        protein_group_objects (list): List of scored [ProteinGroup][pyproteininference.physical.ProteinGroup]</span>
<span class="sd">            objects that have been grouped and sorted. Output from</span>
<span class="sd">             [run_inference][pyproteininference.inference.Inference.run_inference] method.</span>
<span class="sd">        decoy_symbol (str): String that is used to differentiate between decoy proteins and target proteins. Ex: &quot;##&quot;.</span>
<span class="sd">        digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        SCORE_MAPPER (dict): Dictionary that maps potential scores in input files to internal score names.</span>
<span class="sd">        CUSTOM_SCORE_KEY (str): String that indicates a custom score is being used.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">SCORE_MAPPER</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;q_value&quot;</span><span class="p">:</span> <span class="s2">&quot;qvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pep_value&quot;</span><span class="p">:</span> <span class="s2">&quot;pepvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;perc_score&quot;</span><span class="p">:</span> <span class="s2">&quot;percscore&quot;</span><span class="p">,</span>
        <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="s2">&quot;percscore&quot;</span><span class="p">,</span>
        <span class="s2">&quot;q-value&quot;</span><span class="p">:</span> <span class="s2">&quot;qvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;posterior_error_prob&quot;</span><span class="p">:</span> <span class="s2">&quot;pepvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;posterior_error_probability&quot;</span><span class="p">:</span> <span class="s2">&quot;pepvalue&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">CUSTOM_SCORE_KEY</span> <span class="o">=</span> <span class="s2">&quot;custom_score&quot;</span>

    <span class="n">HIGHER_PSM_SCORE</span> <span class="o">=</span> <span class="s2">&quot;higher&quot;</span>
    <span class="n">LOWER_PSM_SCORE</span> <span class="o">=</span> <span class="s2">&quot;lower&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            reader (Reader): Reader object [Reader][pyproteininference.reader.Reader].</span>
<span class="sd">            digest (Digest): Digest object</span>
<span class="sd">                [Digest][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">            validate (bool): True/False to indicate if the input data should be validated.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># If the reader class is from a percolator.psms then define main_data_form as reader.psms</span>
        <span class="c1"># main_data_form is the starting point for all other analyses</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_init_validate</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">parameter_file_object</span>  <span class="c1"># Parameter object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># PSM data post restriction</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of scored Protein objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of sorted scored Protein objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># List of non scored Protein objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># List of Protein objects after picker algorithm</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Protein objects removed via picker</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Variable that indicates whether a higher or lower protein score is better</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># PSM Score used</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of sorted protein group objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span>  <span class="c1"># Decoy symbol from parameter file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>  <span class="c1"># Digest object</span>

        <span class="c1"># Run Checks and Validations</span>
        <span class="k">if</span> <span class="n">validate</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">validate_psm_data</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">validate_digest</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_data_consistency</span><span class="p">()</span>

        <span class="c1"># Run method to fix our parameter object if necessary</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fix_parameters_from_datastore</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_sorted_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">scored</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves a sorted list of protein strings present in the analysis.</span>

<span class="sd">        Args:</span>
<span class="sd">            scored (bool): True/False to indicate if we should return scored or non-scored identifiers.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of sorted protein identifier strings.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; sorted_proteins = data.get_sorted_identifiers(scored=True)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">scored</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">:</span>
                <span class="n">proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scoring_input</span><span class="p">()</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">]</span>

        <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>

        <span class="n">our_target_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
        <span class="n">our_decoy_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

        <span class="n">our_target_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">our_decoy_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

        <span class="n">our_proteins_sorted</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">our_target_sp_proteins</span> <span class="o">+</span> <span class="n">our_decoy_sp_proteins</span> <span class="o">+</span> <span class="n">our_target_tr_proteins</span> <span class="o">+</span> <span class="n">our_decoy_tr_proteins</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">our_proteins_sorted</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">sort_protein_group_objects</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class Method to sort a list of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects by</span>
<span class="sd">        score and number of peptides.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_group_objects (list): list of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>
<span class="sd">            higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of sorted [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; list_of_group_objects = pyproteininference.datastore.DataStore.sort_protein_group_objects(</span>
<span class="sd">            &gt;&gt;&gt;     protein_group_objects=list_of_group_objects, higher_or_lower=&quot;higher&quot;</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>

            <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">protein_group_objects</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                    <span class="o">-</span><span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>

            <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">protein_group_objects</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                    <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">protein_group_objects</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">sort_protein_objects</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class Method to sort a list of [Protein][pyproteininference.physical.Protein] objects by score and number of</span>
<span class="sd">        peptides.</span>

<span class="sd">        Args:</span>
<span class="sd">            grouped_protein_objects (list): list of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of sorted [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; scores_grouped = pyproteininference.datastore.DataStore.sort_protein_objects(</span>
<span class="sd">            &gt;&gt;&gt;     grouped_protein_objects=scores_grouped, higher_or_lower=&quot;higher&quot;</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
            <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">grouped_protein_objects</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="o">-</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
            <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">grouped_protein_objects</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">grouped_protein_objects</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">sort_protein_sub_groups</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_list</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to sort protein sub lists.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_list (list): List of [Protein][pyproteininference.physical.Protein] objects to be sorted.</span>
<span class="sd">            higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of [Protein][pyproteininference.physical.Protein] objects to be sorted by score and number of</span>
<span class="sd">            peptides.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Sort the groups based on higher or lower indication, secondarily sort the groups based on number of unique</span>
        <span class="c1"># peptides</span>
        <span class="c1"># We use the index [1:] as we do not wish to sort the lead protein...</span>
        <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
            <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">),</span> <span class="o">-</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
            <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">protein_list</span>

    <span class="k">def</span> <span class="nf">get_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a list of [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">        Retrieves restricted data if the data has been restricted or all of the data if the data has</span>
<span class="sd">        not been restricted.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; psm_data = data.get_psm_data()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Both main_data_restricted and main_data_form variables are empty. Please re-load the DataStore &quot;</span>
                <span class="s2">&quot;object with a properly loaded Reader object.&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span><span class="p">:</span>
            <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>

        <span class="k">return</span> <span class="n">psm_data</span>

    <span class="k">def</span> <span class="nf">get_protein_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a list of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">        Retrieves picked and scored data if the data has been picked and scored or just the scored data if the data has</span>
<span class="sd">         not been picked.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; # Data must ben ran through a pyproteininference.scoring.Score method</span>
<span class="sd">            &gt;&gt;&gt; protein_data = data.get_protein_data()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">:</span>
            <span class="n">scored_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">scored_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span>

        <span class="k">return</span> <span class="n">scored_proteins</span>

    <span class="k">def</span> <span class="nf">get_protein_identifiers_from_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a list of lists of all possible protein identifiers from the psm data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of lists of protein strings.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; protein_strings = data.get_protein_identifiers_from_psm_data()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">proteins</span>

    <span class="k">def</span> <span class="nf">get_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a list of all q values for all PSMs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of floats (q values).</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; q = data.get_q_values()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">q_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">qvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">q_values</span>

    <span class="k">def</span> <span class="nf">get_pep_values</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a list of all posterior error probabilities for all PSMs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of floats (pep values).</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; pep = data.get_pep_values()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">pep_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">pep_values</span>

    <span class="k">def</span> <span class="nf">get_protein_information_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve a dictionary of scores for each peptide.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: dictionary of scores for each protein.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; protein_dict = data.get_protein_information_dictionary()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">protein_psm_score_dictionary</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="c1"># Loop through all Psms</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
            <span class="c1"># Loop through all proteins</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                <span class="n">protein_psm_score_dictionary</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;peptide&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span>
                        <span class="s2">&quot;Qvalue&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span><span class="p">,</span>
                        <span class="s2">&quot;PosteriorErrorProbability&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">,</span>
                        <span class="s2">&quot;Percscore&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">percscore</span><span class="p">,</span>
                    <span class="p">}</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">protein_psm_score_dictionary</span>

    <span class="k">def</span> <span class="nf">restrict_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">remove1pep</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to restrict the input of [Psm][pyproteininference.physical.Psm]  objects.</span>
<span class="sd">        This method is central to the pyproteininference module and is able to restrict the Psm data by:</span>
<span class="sd">        Q value, Pep Value, Percolator Score, Peptide Length, and Custom Score Input.</span>
<span class="sd">        Restriction values are pulled from</span>
<span class="sd">        the [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter]</span>
<span class="sd">        object.</span>

<span class="sd">        This method sets the `main_data_restricted` and `restricted_peptides` Attributes for the DataStore object.</span>

<span class="sd">        Args:</span>
<span class="sd">            remove1pep (bool): True/False on whether or not to remove PEP values that equal 1 even if other restrictions</span>
<span class="sd">                are set to not restrict.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; data.restrict_psm_data(remove1pep=True)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Validate that we have the main data variable</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_main_data_form</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Restricting PSM data&quot;</span><span class="p">)</span>

        <span class="n">peptide_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_peptide_length</span>
        <span class="n">posterior_error_prob_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_pep</span>
        <span class="n">q_value_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_q</span>
        <span class="n">custom_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_custom</span>

        <span class="n">main_psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of main data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)))</span>
        <span class="c1"># If restrict_main_data is called, we automatically discard everything that has a PEP of 1</span>
        <span class="k">if</span> <span class="n">remove1pep</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span><span class="p">:</span>
            <span class="n">main_psm_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">main_psm_data</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Restrict peptide length and posterior error probability</span>
        <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span>
                    <span class="n">posterior_error_prob_threshold</span>
                <span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict peptide length only</span>
        <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span><span class="p">:</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict peptide length, posterior error probability, and qvalue</span>
        <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span>
                    <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">)</span>
                    <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">)</span>
                <span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict peptide length and qvalue</span>
        <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict posterior error probability and q value</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">)</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict qvalue only</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict posterior error probability only</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">):</span>
                    <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="c1"># Restrict nothing... (only PEP gets restricted - takes everything less than 1)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
            <span class="n">restricted_data</span> <span class="o">=</span> <span class="n">main_psm_data</span>

        <span class="k">if</span> <span class="n">custom_threshold</span><span class="p">:</span>
            <span class="n">custom_restricted</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">&lt;=</span> <span class="n">custom_threshold</span><span class="p">:</span>
                        <span class="n">custom_restricted</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE_SCORE_TYPE</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">&gt;=</span> <span class="n">custom_threshold</span><span class="p">:</span>
                        <span class="n">custom_restricted</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

            <span class="n">restricted_data</span> <span class="o">=</span> <span class="n">custom_restricted</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="o">=</span> <span class="n">restricted_data</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of restricted data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">restricted_data</span><span class="p">)))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">create_scoring_input</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to create the scoring input.</span>
<span class="sd">        This method initializes a list of [Protein][pyproteininference.physical.Protein] objects to get them ready</span>
<span class="sd">        to be scored by [Score][pyproteininference.scoring.Score] methods.</span>
<span class="sd">        This method also takes into account the inference type and aggregates peptides -&gt; proteins accordingly.</span>

<span class="sd">        This method sets the `scoring_input` and `score` Attributes for the DataStore object.</span>

<span class="sd">        The score selected comes from the protein inference parameter object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; data.create_scoring_input()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating Scoring Input&quot;</span><span class="p">)</span>

        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">protein_psm_dict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">score_key</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_MAPPER</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">score_key</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">CUSTOM_SCORE_KEY</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">!=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span>
            <span class="c1"># Loop through all Psms</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
                <span class="n">psms</span><span class="o">.</span><span class="n">assign_main_score</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="n">score_key</span><span class="p">)</span>
                <span class="c1"># Loop through all proteins</span>
                <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                    <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="p">()</span>
            <span class="n">sp_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>

                <span class="c1"># Assign main score</span>
                <span class="n">psms</span><span class="o">.</span><span class="n">assign_main_score</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="n">score_key</span><span class="p">)</span>
                <span class="n">protein_set</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span><span class="p">[</span><span class="n">psms</span><span class="o">.</span><span class="n">non_flanking_peptide</span><span class="p">]</span>
                <span class="c1"># Sort protein_set by sp-alpha, decoy-sp-alpha, tr-alpha, decoy-tr-alpha</span>
                <span class="n">sorted_protein_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                    <span class="n">protein_string_list</span><span class="o">=</span><span class="n">protein_set</span><span class="p">,</span>
                    <span class="n">sp_proteins</span><span class="o">=</span><span class="n">sp_proteins</span><span class="p">,</span>
                    <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="c1"># Restrict the number of identifiers by the value in param file max_identifiers_peptide_centric</span>
                <span class="n">sorted_protein_list</span> <span class="o">=</span> <span class="n">sorted_protein_list</span><span class="p">[:</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span><span class="p">]</span>
                <span class="n">protein_name</span> <span class="o">=</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sorted_protein_list</span><span class="p">)</span>
                <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">protein_name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pkey</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">protein_psm_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">protein_object</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">pkey</span><span class="p">)</span>
            <span class="n">protein_object</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">pkey</span><span class="p">]</span>
            <span class="n">protein_object</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">pkey</span><span class="p">]])</span>
            <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_object</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="n">protein_list</span>

    <span class="k">def</span> <span class="nf">protein_to_peptide_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that returns a map of protein strings to sets of peptide strings and is essentially half</span>
<span class="sd">         of a BiPartite graph.</span>
<span class="sd">        This method sets the `protein_peptide_dictionary` Attribute for the DataStore object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            collections.defaultdict: Dictionary of protein strings (keys) that map to sets of peptide strings based</span>
<span class="sd">            on the peptides and proteins found in the search. Protein -&gt; set(Peptides).</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; protein_peptide_dict = data.protein_to_peptide_dictionary()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">res_pep_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>
        <span class="n">default_dict_proteins</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">peptide_objects</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                <span class="n">cur_peptide</span> <span class="o">=</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
                <span class="k">if</span> <span class="n">cur_peptide</span> <span class="ow">in</span> <span class="n">res_pep_set</span><span class="p">:</span>
                    <span class="n">default_dict_proteins</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_peptide</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span> <span class="o">=</span> <span class="n">default_dict_proteins</span>

        <span class="k">return</span> <span class="n">default_dict_proteins</span>

    <span class="k">def</span> <span class="nf">peptide_to_protein_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that returns a map of peptide strings to sets of protein strings and is essentially half of a</span>
<span class="sd">        BiPartite graph.</span>
<span class="sd">        This method sets the `peptide_protein_dictionary` Attribute for the DataStore object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            collections.defaultdict: Dictionary of peptide strings (keys) that map to sets of protein strings based</span>
<span class="sd">                on the peptides and proteins found in the search. Peptide -&gt; set(Proteins).</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; peptide_protein_dict = data.peptide_to_protein_dictionary()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

        <span class="n">res_pep_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>
        <span class="n">default_dict_peptides</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">peptide_objects</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                <span class="n">cur_peptide</span> <span class="o">=</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
                <span class="k">if</span> <span class="n">cur_peptide</span> <span class="ow">in</span> <span class="n">res_pep_set</span><span class="p">:</span>
                    <span class="n">default_dict_peptides</span><span class="p">[</span><span class="n">cur_peptide</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prots</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">pass</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span> <span class="o">=</span> <span class="n">default_dict_peptides</span>

        <span class="k">return</span> <span class="n">default_dict_peptides</span>

    <span class="k">def</span> <span class="nf">unique_to_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve peptides that are unique based on the data from the searches</span>
<span class="sd">        (Not based on the database digestion).</span>

<span class="sd">        Returns:</span>
<span class="sd">            set: a Set of peptide strings</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; unique_peps = data.unique_to_leads_peptides()</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">:</span>
            <span class="n">lead_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">]</span>
            <span class="n">flat_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">lead_peptides</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
            <span class="n">counted_peps</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">flat_peptides</span><span class="p">)</span>
            <span class="n">unique_to_leads_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">counted_peps</span> <span class="k">if</span> <span class="n">counted_peps</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">unique_to_leads_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">unique_to_leads_peptides</span>

    <span class="k">def</span> <span class="nf">higher_or_lower</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to determine if a higher or lower score is better for a given combination of score input and score type.</span>

<span class="sd">        This method sets the `high_low_better` Attribute for the DataStore object.</span>

<span class="sd">        This method depends on the output from the Score class to be sorted properly from best to worst score.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: String indicating &quot;higher&quot; or &quot;lower&quot; depending on if a higher or lower score is a</span>
<span class="sd">                better protein score.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; high_low = data.higher_or_lower()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Determining If a higher or lower score is better based on scored proteins&quot;</span><span class="p">)</span>
            <span class="n">worst_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>
            <span class="n">best_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>

            <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">worst_score</span><span class="p">):</span>
                <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span>

            <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">worst_score</span><span class="p">):</span>
                <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;best score = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_score</span><span class="p">))</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;worst score = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">worst_score</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">best_score</span> <span class="o">==</span> <span class="n">worst_score</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Best and Worst scores were identical, equal to </span><span class="si">{}</span><span class="s2">. Score type </span><span class="si">{}</span><span class="s2"> produced the error, &quot;</span>
                    <span class="s2">&quot;please change psm_score type.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_score</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span> <span class="o">=</span> <span class="n">higher_or_lower</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span>

        <span class="k">return</span> <span class="n">higher_or_lower</span>

    <span class="k">def</span> <span class="nf">get_protein_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_form</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve the protein string identifiers.</span>

<span class="sd">        Args:</span>
<span class="sd">            data_form (str): Can be one of the following: &quot;main&quot;, &quot;restricted&quot;, &quot;picked&quot;, &quot;picked_removed&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of protein identifier strings.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; protein_strings = data.get_protein_identifiers(data_form=&quot;main&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span>
            <span class="c1"># All the data (unrestricted)</span>
            <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>
            <span class="n">prots</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

        <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;restricted&quot;</span><span class="p">:</span>
            <span class="c1"># Proteins that pass certain restriction criteria (peptide length, pep, qvalue)</span>
            <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span>
            <span class="n">prots</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

        <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;picked&quot;</span><span class="p">:</span>
            <span class="c1"># Here we look at proteins that are &#39;picked&#39; (aka the proteins that beat out their matching target/decoy)</span>
            <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span>
            <span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

        <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;picked_removed&quot;</span><span class="p">:</span>
            <span class="c1"># Here we look at the proteins that were removed due to picking (aka the proteins that</span>
            <span class="c1"># have a worse score than their target/decoy counterpart)</span>
            <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span>
            <span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

        <span class="k">return</span> <span class="n">proteins</span>

    <span class="k">def</span> <span class="nf">get_protein_information</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to retrieve attributes for a specific scored protein.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_string (str): Protein Identifier String.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of protein attributes.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; protein_attr = data.get_protein_information(protein_string=&quot;RAF1_HUMAN|P04049&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_scored_protein_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span>
        <span class="n">identifiers</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">protein_scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">group_identification</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">reviewed</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">reviewed</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="c1"># Peptide scores currently broken...</span>
        <span class="n">peptide_scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">peptide_scores</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">picked</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">picked</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
        <span class="n">num_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">num_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>

        <span class="n">main_index</span> <span class="o">=</span> <span class="n">identifiers</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">protein_string</span><span class="p">)</span>

        <span class="n">list_structure</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;identifier&quot;</span><span class="p">,</span>
                <span class="s2">&quot;protein_score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;groups&quot;</span><span class="p">,</span>
                <span class="s2">&quot;reviewed&quot;</span><span class="p">,</span>
                <span class="s2">&quot;peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;peptide_scores&quot;</span><span class="p">,</span>
                <span class="s2">&quot;picked&quot;</span><span class="p">,</span>
                <span class="s2">&quot;num_peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="n">list_structure</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">protein_string</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_scores</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reviewed</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptides</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptide_scores</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">picked</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
        <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_peptides</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">list_structure</span>

    <span class="k">def</span> <span class="nf">exclude_non_distinguishing_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_subset_type</span><span class="o">=</span><span class="s2">&quot;hard&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to Exclude peptides that are not distinguishing on either the search or database level.</span>

<span class="sd">        The method sets the `scoring_input` and `restricted_peptides` variables for the DataStore object.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_subset_type (str): Either &quot;hard&quot; or &quot;soft&quot;. Hard will select distinguishing peptides based on</span>
<span class="sd">                the database digestion. &quot;soft&quot; will only use peptides identified in the search.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; data.exclude_non_distinguishing_peptides(protein_subset_type=&quot;hard&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Exclusion Model&quot;</span><span class="p">)</span>

        <span class="n">our_proteins_sorted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sorted_identifiers</span><span class="p">(</span><span class="n">scored</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">protein_subset_type</span> <span class="o">==</span> <span class="s2">&quot;hard&quot;</span><span class="p">:</span>
            <span class="c1"># Hard protein subsetting defines protein subsets on the digest level (Entire protein is used)</span>
            <span class="c1"># This is how Percolator PI does subsetting</span>
            <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">our_proteins_sorted</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">protein_subset_type</span> <span class="o">==</span> <span class="s2">&quot;soft&quot;</span><span class="p">:</span>
            <span class="c1"># Soft protein subsetting defines protein subsets on the Peptides identified from the search</span>
            <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If neither is dfined we do &quot;hard&quot; exclusion</span>
            <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">our_proteins_sorted</span><span class="p">]</span>

        <span class="c1"># Get frozen set of peptides....</span>
        <span class="c1"># We will also have a corresponding list of proteins...</span>
        <span class="c1"># They will have the same index...</span>
        <span class="n">peptide_sets</span> <span class="o">=</span> <span class="p">[</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">peptides</span><span class="p">]</span>
        <span class="c1"># Find a way to sort this list of sets...</span>
        <span class="c1"># We can sort the sets if we sort proteins from above...</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> number of peptide sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peptide_sets</span><span class="p">)))</span>
        <span class="n">non_subset_peptide_sets</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Get all peptide sets that are not a subset...</span>
        <span class="k">while</span> <span class="n">peptide_sets</span><span class="p">:</span>
            <span class="n">i</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">peptide_set</span> <span class="o">=</span> <span class="n">peptide_sets</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">peptide_set</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">peptide_sets</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">any</span><span class="p">(</span>
                <span class="n">peptide_set</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">non_subset_peptide_sets</span>
            <span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">non_subset_peptide_sets</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">peptide_set</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsed </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsed </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="c1"># Get their index from peptides which is the initial list of sets...</span>
        <span class="n">list_of_indeces</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pep_sets</span> <span class="ow">in</span> <span class="n">non_subset_peptide_sets</span><span class="p">:</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">peptides</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pep_sets</span><span class="p">)</span>
            <span class="n">list_of_indeces</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

        <span class="n">non_subset_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">our_proteins_sorted</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">list_of_indeces</span><span class="p">])</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing direct subset Proteins from the data&quot;</span><span class="p">)</span>
        <span class="c1"># Remove all proteins from scoring input that are a subset of another protein...</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins in scoring input after removing subset proteins&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">)))</span>

        <span class="c1"># For all the proteins that are not a complete subset of another protein...</span>
        <span class="c1"># Get the raw peptides...</span>
        <span class="n">raw_peps</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

        <span class="c1"># Make the raw peptides a flat list</span>
        <span class="n">flat_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">raw_peps</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

        <span class="c1"># Count the number of peptides in this list...</span>
        <span class="c1"># This is the number of proteins this peptide maps to....</span>
        <span class="n">counted_peptides</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">flat_peptides</span><span class="p">)</span>

        <span class="c1"># If the count is greater than 1... exclude the protein entirely from scoring input... :)</span>
        <span class="n">raw_peps_good</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">counted_peptides</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">counted_peptides</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">])</span>

        <span class="c1"># Alter self.scoring_input by removing psms and peptides that are not in raw_peps_good</span>
        <span class="n">current_score_input</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">current_score_input</span><span class="p">)):</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">psm_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">new_raw_peptides</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">current_psms</span> <span class="o">=</span> <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">psms</span>
            <span class="n">current_raw_peptides</span> <span class="o">=</span> <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">raw_peptides</span>

            <span class="k">for</span> <span class="n">psm_scores</span> <span class="ow">in</span> <span class="n">current_psms</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psm_scores</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="ow">in</span> <span class="n">raw_peps_good</span><span class="p">:</span>
                    <span class="n">psm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm_scores</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">rp</span> <span class="ow">in</span> <span class="n">current_raw_peptides</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">rp</span><span class="p">)</span> <span class="ow">in</span> <span class="n">raw_peps_good</span><span class="p">:</span>
                    <span class="n">new_raw_peptides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rp</span><span class="p">)</span>

            <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">psm_list</span>
            <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="n">new_raw_peptides</span>

            <span class="k">if</span> <span class="n">k</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Redefined </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Redefined </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>

        <span class="n">filtered_score_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">current_score_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="n">filtered_score_input</span>

        <span class="c1"># Recompute the flat peptides</span>
        <span class="n">raw_peps</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

        <span class="c1"># Make the raw peptides a flat list</span>
        <span class="n">new_flat_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">raw_peps</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_flat_peptides</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">protein_picker</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to run the protein picker algorithm.</span>

<span class="sd">        Proteins must be scored first with [score_psms][pyproteininference.scoring.Score.score_psms].</span>

<span class="sd">        The algorithm will match target and decoy proteins identified from the PSMs from the search.</span>
<span class="sd">        If a target and matching decoy is found then target/decoy competition is performed.</span>
<span class="sd">        In the Target/Decoy pair the protein with the better score is kept and the one with the worse score is</span>
<span class="sd">        discarded from the analysis.</span>

<span class="sd">        The method sets the `picked_proteins_scored` and `picked_proteins_removed` variables for</span>
<span class="sd">        the DataStore object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; data.protein_picker()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Protein Picker&quot;</span><span class="p">)</span>

        <span class="c1"># Use higher or lower class to determine if a higher protein score or lower protein score is better</span>
        <span class="c1"># based on the scoring method used</span>
        <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>
        <span class="c1"># Here we determine if a lower or higher score is better</span>
        <span class="c1"># Since all input is ordered from best to worst we can do the following</span>

        <span class="n">index_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># data.scored_proteins is simply a list of Protein objects...</span>
        <span class="c1"># Create list of all decoy proteins</span>
        <span class="n">decoy_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span><span class="p">]</span>
        <span class="c1"># Create a list of all potential matching targets (some of these may not exist in the search)</span>
        <span class="n">matching_targets</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">decoy_proteins</span><span class="p">]</span>

        <span class="c1"># Create a list of all the proteins from the scored data</span>
        <span class="n">all_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins scored&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_proteins</span><span class="p">)))</span>

        <span class="n">total_targets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">total_decoys</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">decoys_removed</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">targets_removed</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Loop over all decoys identified in the search</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Picking Proteins...&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoy_proteins</span><span class="p">)):</span>
            <span class="n">cur_decoy_index</span> <span class="o">=</span> <span class="n">all_proteins</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">decoy_proteins</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">cur_decoy_protein_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="n">cur_decoy_index</span><span class="p">]</span>
            <span class="n">total_decoys</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

            <span class="c1"># Try, Except here because the matching target to the decoy may not be a result from the search</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">cur_target_index</span> <span class="o">=</span> <span class="n">all_proteins</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">matching_targets</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">cur_target_protein_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="n">cur_target_index</span><span class="p">]</span>
                <span class="n">total_targets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">score</span> <span class="o">&gt;</span> <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>
                        <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                        <span class="n">decoys_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                        <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                        <span class="n">targets_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                        <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>

                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">score</span> <span class="o">&lt;</span> <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>
                        <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                        <span class="n">decoys_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                        <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                        <span class="n">targets_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                        <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> total decoy proteins&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_decoys</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> matching target proteins also found in search&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_targets</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> decoy proteins to be removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoys_removed</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> target proteins to be removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targets_removed</span><span class="p">)))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing Lower Scoring Proteins...&quot;</span><span class="p">)</span>
        <span class="n">picked_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">removed_proteins</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_objects</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">protein_objects</span><span class="o">.</span><span class="n">picked</span><span class="p">:</span>
                <span class="n">picked_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_objects</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">removed_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_objects</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span> <span class="o">=</span> <span class="n">picked_list</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span> <span class="o">=</span> <span class="n">removed_proteins</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished Removing Proteins&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">calculate_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">regular</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method calculates Q values FDR on the lead protein in the group on the `protein_group_objects`</span>
<span class="sd">        instance variable.</span>
<span class="sd">        FDR is calculated As (2*decoys)/total if regular is set to True and is</span>
<span class="sd">        (decoys)/total if regular is set to False.</span>

<span class="sd">        This method updates the `protein_group_objects` for the DataStore object by updating</span>
<span class="sd">        the q_value variable of the [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">            &gt;&gt;&gt; # Data must be scored first</span>
<span class="sd">            &gt;&gt;&gt; data.calculate_q_values()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_protein_group_objects</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Q values from the protein group objects&quot;</span><span class="p">)</span>

        <span class="c1"># pick out the lead scoring protein for each group... lead score is at 0 position</span>
        <span class="n">lead_score</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">]</span>
        <span class="c1"># Now pick out only the lead protein identifiers</span>
        <span class="n">lead_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lead_score</span><span class="p">]</span>

        <span class="n">lead_proteins</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating FDRs&quot;</span><span class="p">)</span>
        <span class="n">fdr_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lead_proteins</span><span class="p">)):</span>
            <span class="n">binary_decoy_target_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">elem</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">elem</span> <span class="ow">in</span> <span class="n">lead_proteins</span><span class="p">]</span>
            <span class="n">total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lead_proteins</span><span class="p">)</span>
            <span class="n">decoys</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">binary_decoy_target_list</span><span class="p">)</span>
            <span class="c1"># Calculate FDR at every step starting with the entire list...</span>
            <span class="c1"># Delete first entry (worst score) every time we go through a cycle</span>
            <span class="k">if</span> <span class="n">regular</span><span class="p">:</span>
                <span class="n">fdr</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">decoys</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">total</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">fdr</span> <span class="o">=</span> <span class="p">(</span><span class="n">decoys</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">total</span><span class="p">))</span>
            <span class="n">fdr_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fdr</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">lead_proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">qvalue_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_fdr_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Q Values&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">fdrs</span> <span class="ow">in</span> <span class="n">fdr_list</span><span class="p">:</span>
            <span class="n">new_fdr_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fdrs</span><span class="p">)</span>
            <span class="n">qvalue</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">new_fdr_list</span><span class="p">)</span>
            <span class="c1"># qvalue = fdrs</span>
            <span class="n">qvalue_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>

        <span class="n">qvalue_list</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assigning Q Values&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">)):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">q_value</span> <span class="o">=</span> <span class="n">qvalue_list</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

        <span class="n">fdr_restricted</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">q_value</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">]</span>

        <span class="n">fdr_restricted_set</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="p">))]</span>

        <span class="n">onehitwonders</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">fdr_restricted_set</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">onehitwonders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Protein Group leads that pass with more than 1 PSM with a </span><span class="si">{}</span><span class="s2"> FDR = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">,</span>
                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted_set</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">onehitwonders</span><span class="p">)),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Protein Group lead One hit Wonders that pass </span><span class="si">{}</span><span class="s2"> FDR = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">onehitwonders</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Number of Protein groups that pass a </span><span class="si">{}</span><span class="s2"> percent FDR: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted_set</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished Q value Calculation&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">validate_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that validates the PSM data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_decoys_from_data</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_isoform_from_data</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">validate_digest</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that validates the [Digest object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_reviewed_v_unreviewed</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_target_decoy_split</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">check_data_consistency</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that checks for data consistency.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_data_digest_overlap_psms</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_data_digest_overlap_proteins</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_check_data_digest_overlap_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that logs the overlap between the digested fasta file and the input files on the PSM level.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">]</span>
        <span class="n">peptides_in_digest</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">peptides_from_search_in_digest</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">peptides</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">peptides_in_digest</span><span class="p">]</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">peptides</span><span class="p">)))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">peptides_from_search_in_digest</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> PSMs identified from input files&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peptides</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> PSMs identified from input files that are also present in database digestion&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">peptides_from_search_in_digest</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">; ratio of PSMs identified from input files to those that are present in the search&quot;</span>
            <span class="s2">&quot; and in the database digestion&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">percentage</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_check_data_digest_overlap_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that logs the overlap between the digested fasta file and the input files on the Protein level.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">]</span>
        <span class="n">flat_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>
        <span class="n">proteins_in_digest</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">proteins_from_search_in_digest</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins_in_digest</span><span class="p">]</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">flat_proteins</span><span class="p">))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">proteins_from_search_in_digest</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins identified from input files&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">flat_proteins</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins identified from input files that are also present in database digestion&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">proteins_from_search_in_digest</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">; ratio of proteins identified from input files that are also present in database digestion&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">percentage</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_check_target_decoy_split</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that logs the number of target and decoy proteins from the digest.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check the number of targets vs the number of decoys from the digest</span>
        <span class="n">targets</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">x</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span>
        <span class="p">]</span>
        <span class="n">decoys</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span>
        <span class="p">]</span>
        <span class="n">ratio</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targets</span><span class="p">))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoys</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Target Proteins in Digest: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targets</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Decoy Proteins in Digest: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoys</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Ratio of Targets Proteins to Decoy Proteins: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ratio</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">_validate_decoys_from_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that checks to make sure that target and decoy proteins exist in the data files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check to see if we find decoys from our input files</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">]</span>
        <span class="n">flat_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>
        <span class="n">targets</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
        <span class="n">decoys</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Target Proteins in Data Files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targets</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Decoy Proteins in Data Files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoys</span><span class="p">)))</span>

    <span class="k">def</span> <span class="nf">_validate_isoform_from_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that validates whether or not isoforms are able to be identified in the data files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check to see if we find any proteins with isoform info in name in our input files</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">]</span>
        <span class="n">flat_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">:</span>
            <span class="n">non_iso</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">non_iso</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">:</span>
            <span class="n">iso</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">flat_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">iso</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Non Isoform Labeled Proteins in Data Files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">non_iso</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Isoform Labeled Proteins in Data Files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">iso</span><span class="p">)))</span>

    <span class="k">def</span> <span class="nf">_validate_reviewed_v_unreviewed</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that logs whether or not we can distinguish from reviewed and unreviewd protein identifiers</span>
<span class="sd">        in the digest.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check to see if we get reviewed prots in digest...</span>
        <span class="n">reviewed_proteins</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
        <span class="n">proteins_in_digest</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
        <span class="n">unreviewed_proteins</span> <span class="o">=</span> <span class="n">proteins_in_digest</span> <span class="o">-</span> <span class="n">reviewed_proteins</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Total Proteins in from Digest: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">proteins_in_digest</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Reviewed Proteins in from Digest: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">reviewed_proteins</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of Unreviewed Proteins in from Digest: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">unreviewed_proteins</span><span class="p">))</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">sort_protein_strings</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_string_list</span><span class="p">,</span> <span class="n">sp_proteins</span><span class="p">,</span> <span class="n">decoy_symbol</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that sorts protein strings in the following order: Target Reviewed, Decoy Reviewed, Target Unreviewed,</span>
<span class="sd">         Decoy Unreviewed.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_string_list (list): List of Protein Strings.</span>
<span class="sd">            sp_proteins (set): Set of Reviewed Protein Strings.</span>
<span class="sd">            decoy_symbol (str): Symbol to denote a decoy protein identifier IE &quot;##&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of sorted protein strings.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; list_of_group_objects = datastore.DataStore.sort_protein_strings(</span>
<span class="sd">            &gt;&gt;&gt;     protein_string_list=protein_string_list, sp_proteins=sp_proteins, decoy_symbol=&quot;##&quot;</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">our_target_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
        <span class="n">our_decoy_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

        <span class="n">our_target_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">our_decoy_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

        <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">our_target_sp_proteins</span> <span class="o">+</span> <span class="n">our_decoy_sp_proteins</span> <span class="o">+</span> <span class="n">our_target_tr_proteins</span> <span class="o">+</span> <span class="n">our_decoy_tr_proteins</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">identifiers_sorted</span>

    <span class="k">def</span> <span class="nf">input_has_q</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that checks to see if the input data has q values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">len_q</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">qvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">qvalue</span><span class="p">])</span>
        <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">len_q</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Q value; Can restrict by Q value&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Q value; Cannot restrict by Q value&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">status</span>

    <span class="k">def</span> <span class="nf">input_has_pep</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that checks to see if the input data has pep values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">len_pep</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">])</span>
        <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">len_pep</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Pep value; Can restrict by Pep value&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Pep value; Cannot restrict by Pep value&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">status</span>

    <span class="k">def</span> <span class="nf">input_has_custom</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that checks to see if the input data has custom score values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">len_c</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">custom_score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">custom_score</span><span class="p">])</span>
        <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">len_c</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Custom value; Can restrict by Custom value&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Custom value; Cannot restrict by Custom value&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">status</span>

    <span class="k">def</span> <span class="nf">get_protein_objects</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method retrieves protein objects. Either retrieves FDR restricted list of protien objects,</span>
<span class="sd">        or retrieves all objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            fdr_restricted (bool): True/False on whether to restrict the list of objects based on FDR.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of scored [ProteinGroup][pyproteininference.physical.ProteinGroup] objects</span>
<span class="sd">                that have been grouped and sorted.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">false_discovery_rate</span><span class="p">:</span>
            <span class="n">false_discovery_rate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span>
        <span class="k">if</span> <span class="n">fdr_restricted</span><span class="p">:</span>
            <span class="n">protein_objects</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">q_value</span> <span class="o">&lt;=</span> <span class="n">false_discovery_rate</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span>

        <span class="k">return</span> <span class="n">protein_objects</span>

    <span class="k">def</span> <span class="nf">_init_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reader</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure the reader object is properly loaded and validated.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">psms</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">psms</span>  <span class="c1"># Unrestricted PSM data</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Psms variable from Reader object is either empty or does not exist. &quot;</span>
                <span class="s2">&quot;Make sure your files contain proper data and that you run the &#39;read_psms&#39; &quot;</span>
                <span class="s2">&quot;method on your Reader object.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_main_data_form</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure the Main data has been defined to run DataStore methods.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Main Data is not defined, thus method cannot be ran. Please make sure PSM data is properly&quot;</span>
                <span class="s2">&quot; loaded from the Reader object&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_main_data_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure the Main data Restricted has been defined to run DataStore methods.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Main Data Restricted is not defined, thus method cannot be ran. Please make sure PSM data is properly&quot;</span>
                <span class="s2">&quot; loaded from the Reader object and make sure to run DataStore method &#39;restrict_psm_data&#39;.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_scored_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure that proteins have been scored to run certain subsequent methods.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Proteins have not been scored, Please initialize a Score object and run a score method with&quot;</span>
                <span class="s2">&quot; &#39;score_psms&#39; instance method.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_scoring_input</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure that Scoring Input has been created to be able to run scoring methods.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Scoring input has not been created, Please run &#39;create_scoring_input&#39; method from the DataStore &quot;</span>
                <span class="s2">&quot;object to continue.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_protein_group_objects</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal Method that checks to make sure inference has been run before proceeding.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Either &#39;protein_group_objects&#39; or &#39;grouped_scored_proteins&#39; or both DataStore variables are undefined.&quot;</span>
                <span class="s2">&quot; Please make sure you run an inference method from the Inference class before proceeding.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">generate_fdr_vs_target_hits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method for calculating FDR vs number of Target Proteins.</span>

<span class="sd">        Args:</span>
<span class="sd">            fdr_max (float): The maximum false discovery rate to calculate target hits for.</span>
<span class="sd">                Will stop once fdr_max is reached.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of lists of: (FDR, Number of Target Hits). Ordered by increasing number of Target Hits.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fdr_vs_count</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">count_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pg</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">:</span>
                <span class="n">count_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>
            <span class="n">fdr_vs_count</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">pg</span><span class="o">.</span><span class="n">q_value</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">count_list</span><span class="p">)])</span>

        <span class="n">fdr_vs_count</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_count</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">fdr_max</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">fdr_vs_count</span>

    <span class="k">def</span> <span class="nf">recover_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Recovering Proteins that exist in the input files but not in the database digest.&quot;</span><span class="p">)</span>
        <span class="n">all_psms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_psms</span><span class="p">]</span>
        <span class="n">flat_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

        <span class="n">missing_prots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">prot</span> <span class="ow">in</span> <span class="n">flat_proteins</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">[</span><span class="n">prot</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">missing_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prot</span><span class="p">)</span>

                <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>
                <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span> <span class="k">if</span> <span class="n">prot</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">pep</span> <span class="ow">in</span> <span class="n">peptides</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">pep</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prot</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pep</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">missing_prots</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins not found in mapping objects, please double check that your database&quot;</span>
                <span class="s2">&quot; provided is accurate for the given input data.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">missing_prots</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No missing proteins in the mapping objects.&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">













  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>reader</strong> (<code>Reader</code>)  Reader object <a class="autorefs autorefs-internal" href="#pyproteininference.reader.Reader">Reader</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  Digest object
<a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest</a>.</p></li>
            <li><p><strong>validate</strong> (<code>bool</code>)  True/False to indicate if the input data should be validated.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        reader (Reader): Reader object [Reader][pyproteininference.reader.Reader].</span>
<span class="sd">        digest (Digest): Digest object</span>
<span class="sd">            [Digest][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        validate (bool): True/False to indicate if the input data should be validated.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># If the reader class is from a percolator.psms then define main_data_form as reader.psms</span>
    <span class="c1"># main_data_form is the starting point for all other analyses</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_init_validate</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">parameter_file_object</span>  <span class="c1"># Parameter object</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># PSM data post restriction</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of scored Protein objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of sorted scored Protein objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># List of non scored Protein objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># List of Protein objects after picker algorithm</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Protein objects removed via picker</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Variable that indicates whether a higher or lower protein score is better</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># PSM Score used</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of sorted protein group objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span>  <span class="c1"># Decoy symbol from parameter file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>  <span class="c1"># Digest object</span>

    <span class="c1"># Run Checks and Validations</span>
    <span class="k">if</span> <span class="n">validate</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">validate_psm_data</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">validate_digest</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_data_consistency</span><span class="p">()</span>

    <span class="c1"># Run method to fix our parameter object if necessary</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fix_parameters_from_datastore</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.calculate_q_values">
<code class="highlight language-python"><span class="n">calculate_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">regular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method calculates Q values FDR on the lead protein in the group on the <code>protein_group_objects</code>
instance variable.
FDR is calculated As (2*decoys)/total if regular is set to True and is
(decoys)/total if regular is set to False.</p>
<p>This method updates the <code>protein_group_objects</code> for the DataStore object by updating
the q_value variable of the <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="c1"># Data must be scored first</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">data</span><span class="o">.</span><span class="n">calculate_q_values</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">calculate_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">regular</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method calculates Q values FDR on the lead protein in the group on the `protein_group_objects`</span>
<span class="sd">    instance variable.</span>
<span class="sd">    FDR is calculated As (2*decoys)/total if regular is set to True and is</span>
<span class="sd">    (decoys)/total if regular is set to False.</span>

<span class="sd">    This method updates the `protein_group_objects` for the DataStore object by updating</span>
<span class="sd">    the q_value variable of the [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; # Data must be scored first</span>
<span class="sd">        &gt;&gt;&gt; data.calculate_q_values()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_protein_group_objects</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Q values from the protein group objects&quot;</span><span class="p">)</span>

    <span class="c1"># pick out the lead scoring protein for each group... lead score is at 0 position</span>
    <span class="n">lead_score</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">]</span>
    <span class="c1"># Now pick out only the lead protein identifiers</span>
    <span class="n">lead_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lead_score</span><span class="p">]</span>

    <span class="n">lead_proteins</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating FDRs&quot;</span><span class="p">)</span>
    <span class="n">fdr_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lead_proteins</span><span class="p">)):</span>
        <span class="n">binary_decoy_target_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">elem</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">elem</span> <span class="ow">in</span> <span class="n">lead_proteins</span><span class="p">]</span>
        <span class="n">total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lead_proteins</span><span class="p">)</span>
        <span class="n">decoys</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">binary_decoy_target_list</span><span class="p">)</span>
        <span class="c1"># Calculate FDR at every step starting with the entire list...</span>
        <span class="c1"># Delete first entry (worst score) every time we go through a cycle</span>
        <span class="k">if</span> <span class="n">regular</span><span class="p">:</span>
            <span class="n">fdr</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">decoys</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">total</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fdr</span> <span class="o">=</span> <span class="p">(</span><span class="n">decoys</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">total</span><span class="p">))</span>
        <span class="n">fdr_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fdr</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">lead_proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">qvalue_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">new_fdr_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Q Values&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">fdrs</span> <span class="ow">in</span> <span class="n">fdr_list</span><span class="p">:</span>
        <span class="n">new_fdr_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fdrs</span><span class="p">)</span>
        <span class="n">qvalue</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">new_fdr_list</span><span class="p">)</span>
        <span class="c1"># qvalue = fdrs</span>
        <span class="n">qvalue_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">qvalue</span><span class="p">)</span>

    <span class="n">qvalue_list</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assigning Q Values&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">)):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">q_value</span> <span class="o">=</span> <span class="n">qvalue_list</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

    <span class="n">fdr_restricted</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">q_value</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">]</span>

    <span class="n">fdr_restricted_set</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="p">))]</span>

    <span class="n">onehitwonders</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">fdr_restricted_set</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">onehitwonders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Protein Group leads that pass with more than 1 PSM with a </span><span class="si">{}</span><span class="s2"> FDR = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">,</span>
            <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted_set</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">onehitwonders</span><span class="p">)),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Protein Group lead One hit Wonders that pass </span><span class="si">{}</span><span class="s2"> FDR = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">onehitwonders</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Number of Protein groups that pass a </span><span class="si">{}</span><span class="s2"> percent FDR: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">fdr_restricted_set</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished Q value Calculation&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.check_data_consistency">
<code class="highlight language-python"><span class="n">check_data_consistency</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that checks for data consistency.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">check_data_consistency</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that checks for data consistency.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_data_digest_overlap_psms</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_data_digest_overlap_proteins</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.create_scoring_input">
<code class="highlight language-python"><span class="n">create_scoring_input</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to create the scoring input.
This method initializes a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects to get them ready
to be scored by <a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score">Score</a> methods.
This method also takes into account the inference type and aggregates peptides -&gt; proteins accordingly.</p>
<p>This method sets the <code>scoring_input</code> and <code>score</code> Attributes for the DataStore object.</p>
<p>The score selected comes from the protein inference parameter object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">data</span><span class="o">.</span><span class="n">create_scoring_input</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">create_scoring_input</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to create the scoring input.</span>
<span class="sd">    This method initializes a list of [Protein][pyproteininference.physical.Protein] objects to get them ready</span>
<span class="sd">    to be scored by [Score][pyproteininference.scoring.Score] methods.</span>
<span class="sd">    This method also takes into account the inference type and aggregates peptides -&gt; proteins accordingly.</span>

<span class="sd">    This method sets the `scoring_input` and `score` Attributes for the DataStore object.</span>

<span class="sd">    The score selected comes from the protein inference parameter object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; data.create_scoring_input()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating Scoring Input&quot;</span><span class="p">)</span>

    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">protein_psm_dict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">score_key</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_MAPPER</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="n">score_key</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">CUSTOM_SCORE_KEY</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">!=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span>
        <span class="c1"># Loop through all Psms</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
            <span class="n">psms</span><span class="o">.</span><span class="n">assign_main_score</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="n">score_key</span><span class="p">)</span>
            <span class="c1"># Loop through all proteins</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="p">()</span>
        <span class="n">sp_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>

            <span class="c1"># Assign main score</span>
            <span class="n">psms</span><span class="o">.</span><span class="n">assign_main_score</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="n">score_key</span><span class="p">)</span>
            <span class="n">protein_set</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span><span class="p">[</span><span class="n">psms</span><span class="o">.</span><span class="n">non_flanking_peptide</span><span class="p">]</span>
            <span class="c1"># Sort protein_set by sp-alpha, decoy-sp-alpha, tr-alpha, decoy-tr-alpha</span>
            <span class="n">sorted_protein_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                <span class="n">protein_string_list</span><span class="o">=</span><span class="n">protein_set</span><span class="p">,</span>
                <span class="n">sp_proteins</span><span class="o">=</span><span class="n">sp_proteins</span><span class="p">,</span>
                <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># Restrict the number of identifiers by the value in param file max_identifiers_peptide_centric</span>
            <span class="n">sorted_protein_list</span> <span class="o">=</span> <span class="n">sorted_protein_list</span><span class="p">[:</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span><span class="p">]</span>
            <span class="n">protein_name</span> <span class="o">=</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sorted_protein_list</span><span class="p">)</span>
            <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">protein_name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">pkey</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">protein_psm_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="n">protein_object</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">pkey</span><span class="p">)</span>
        <span class="n">protein_object</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">pkey</span><span class="p">]</span>
        <span class="n">protein_object</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_psm_dict</span><span class="p">[</span><span class="n">pkey</span><span class="p">]])</span>
        <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_object</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="n">protein_list</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.exclude_non_distinguishing_peptides">
<code class="highlight language-python"><span class="n">exclude_non_distinguishing_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_subset_type</span><span class="o">=</span><span class="s1">&#39;hard&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to Exclude peptides that are not distinguishing on either the search or database level.</p>
<p>The method sets the <code>scoring_input</code> and <code>restricted_peptides</code> variables for the DataStore object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>protein_subset_type</strong> (<code>str</code>)  Either "hard" or "soft". Hard will select distinguishing peptides based on
the database digestion. "soft" will only use peptides identified in the search.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">data</span><span class="o">.</span><span class="n">exclude_non_distinguishing_peptides</span><span class="p">(</span><span class="n">protein_subset_type</span><span class="o">=</span><span class="s2">&quot;hard&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">exclude_non_distinguishing_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_subset_type</span><span class="o">=</span><span class="s2">&quot;hard&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to Exclude peptides that are not distinguishing on either the search or database level.</span>

<span class="sd">    The method sets the `scoring_input` and `restricted_peptides` variables for the DataStore object.</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_subset_type (str): Either &quot;hard&quot; or &quot;soft&quot;. Hard will select distinguishing peptides based on</span>
<span class="sd">            the database digestion. &quot;soft&quot; will only use peptides identified in the search.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; data.exclude_non_distinguishing_peptides(protein_subset_type=&quot;hard&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Exclusion Model&quot;</span><span class="p">)</span>

    <span class="n">our_proteins_sorted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sorted_identifiers</span><span class="p">(</span><span class="n">scored</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">protein_subset_type</span> <span class="o">==</span> <span class="s2">&quot;hard&quot;</span><span class="p">:</span>
        <span class="c1"># Hard protein subsetting defines protein subsets on the digest level (Entire protein is used)</span>
        <span class="c1"># This is how Percolator PI does subsetting</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">our_proteins_sorted</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">protein_subset_type</span> <span class="o">==</span> <span class="s2">&quot;soft&quot;</span><span class="p">:</span>
        <span class="c1"># Soft protein subsetting defines protein subsets on the Peptides identified from the search</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># If neither is dfined we do &quot;hard&quot; exclusion</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">our_proteins_sorted</span><span class="p">]</span>

    <span class="c1"># Get frozen set of peptides....</span>
    <span class="c1"># We will also have a corresponding list of proteins...</span>
    <span class="c1"># They will have the same index...</span>
    <span class="n">peptide_sets</span> <span class="o">=</span> <span class="p">[</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">e</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">peptides</span><span class="p">]</span>
    <span class="c1"># Find a way to sort this list of sets...</span>
    <span class="c1"># We can sort the sets if we sort proteins from above...</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> number of peptide sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peptide_sets</span><span class="p">)))</span>
    <span class="n">non_subset_peptide_sets</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># Get all peptide sets that are not a subset...</span>
    <span class="k">while</span> <span class="n">peptide_sets</span><span class="p">:</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">peptide_set</span> <span class="o">=</span> <span class="n">peptide_sets</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">peptide_set</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">peptide_sets</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">any</span><span class="p">(</span>
            <span class="n">peptide_set</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">non_subset_peptide_sets</span>
        <span class="p">):</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">non_subset_peptide_sets</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">peptide_set</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsed </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parsed </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="c1"># Get their index from peptides which is the initial list of sets...</span>
    <span class="n">list_of_indeces</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">pep_sets</span> <span class="ow">in</span> <span class="n">non_subset_peptide_sets</span><span class="p">:</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">peptides</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pep_sets</span><span class="p">)</span>
        <span class="n">list_of_indeces</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

    <span class="n">non_subset_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">our_proteins_sorted</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">list_of_indeces</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing direct subset Proteins from the data&quot;</span><span class="p">)</span>
    <span class="c1"># Remove all proteins from scoring input that are a subset of another protein...</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins in scoring input after removing subset proteins&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">)))</span>

    <span class="c1"># For all the proteins that are not a complete subset of another protein...</span>
    <span class="c1"># Get the raw peptides...</span>
    <span class="n">raw_peps</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

    <span class="c1"># Make the raw peptides a flat list</span>
    <span class="n">flat_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">raw_peps</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

    <span class="c1"># Count the number of peptides in this list...</span>
    <span class="c1"># This is the number of proteins this peptide maps to....</span>
    <span class="n">counted_peptides</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">flat_peptides</span><span class="p">)</span>

    <span class="c1"># If the count is greater than 1... exclude the protein entirely from scoring input... :)</span>
    <span class="n">raw_peps_good</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">counted_peptides</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">counted_peptides</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">])</span>

    <span class="c1"># Alter self.scoring_input by removing psms and peptides that are not in raw_peps_good</span>
    <span class="n">current_score_input</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">current_score_input</span><span class="p">)):</span>
        <span class="n">k</span> <span class="o">=</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">psm_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_raw_peptides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">current_psms</span> <span class="o">=</span> <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">psms</span>
        <span class="n">current_raw_peptides</span> <span class="o">=</span> <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">raw_peptides</span>

        <span class="k">for</span> <span class="n">psm_scores</span> <span class="ow">in</span> <span class="n">current_psms</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">psm_scores</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="ow">in</span> <span class="n">raw_peps_good</span><span class="p">:</span>
                <span class="n">psm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm_scores</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">rp</span> <span class="ow">in</span> <span class="n">current_raw_peptides</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">rp</span><span class="p">)</span> <span class="ow">in</span> <span class="n">raw_peps_good</span><span class="p">:</span>
                <span class="n">new_raw_peptides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rp</span><span class="p">)</span>

        <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">psm_list</span>
        <span class="n">current_score_input</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="n">new_raw_peptides</span>

        <span class="k">if</span> <span class="n">k</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Redefined </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Redefined </span><span class="si">{}</span><span class="s2"> Peptide Sets&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>

    <span class="n">filtered_score_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">current_score_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="n">filtered_score_input</span>

    <span class="c1"># Recompute the flat peptides</span>
    <span class="n">raw_peps</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">non_subset_proteins</span><span class="p">]</span>

    <span class="c1"># Make the raw peptides a flat list</span>
    <span class="n">new_flat_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">raw_peps</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_flat_peptides</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.generate_fdr_vs_target_hits">
<code class="highlight language-python"><span class="n">generate_fdr_vs_target_hits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method for calculating FDR vs number of Target Proteins.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>fdr_max</strong> (<code>float</code>)  The maximum false discovery rate to calculate target hits for.
Will stop once fdr_max is reached.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of lists of: (FDR, Number of Target Hits). Ordered by increasing number of Target Hits.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">generate_fdr_vs_target_hits</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method for calculating FDR vs number of Target Proteins.</span>

<span class="sd">    Args:</span>
<span class="sd">        fdr_max (float): The maximum false discovery rate to calculate target hits for.</span>
<span class="sd">            Will stop once fdr_max is reached.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of lists of: (FDR, Number of Target Hits). Ordered by increasing number of Target Hits.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fdr_vs_count</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">count_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">pg</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">:</span>
            <span class="n">count_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>
        <span class="n">fdr_vs_count</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">pg</span><span class="o">.</span><span class="n">q_value</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">count_list</span><span class="p">)])</span>

    <span class="n">fdr_vs_count</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_count</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">fdr_max</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">fdr_vs_count</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_pep_values">
<code class="highlight language-python"><span class="n">get_pep_values</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a list of all posterior error probabilities for all PSMs.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of floats (pep values).</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pep</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_pep_values</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_pep_values</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a list of all posterior error probabilities for all PSMs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of floats (pep values).</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; pep = data.get_pep_values()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">pep_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">pep_values</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_data">
<code class="highlight language-python"><span class="n">get_protein_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.
Retrieves picked and scored data if the data has been picked and scored or just the scored data if the data has
 not been picked.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="c1"># Data must ben ran through a pyproteininference.scoring.Score method</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a list of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">    Retrieves picked and scored data if the data has been picked and scored or just the scored data if the data has</span>
<span class="sd">     not been picked.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; # Data must ben ran through a pyproteininference.scoring.Score method</span>
<span class="sd">        &gt;&gt;&gt; protein_data = data.get_protein_data()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">:</span>
        <span class="n">scored_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">scored_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span>

    <span class="k">return</span> <span class="n">scored_proteins</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_identifiers">
<code class="highlight language-python"><span class="n">get_protein_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_form</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve the protein string identifiers.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data_form</strong> (<code>str</code>)  Can be one of the following: "main", "restricted", "picked", "picked_removed".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of protein identifier strings.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_strings</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_protein_identifiers</span><span class="p">(</span><span class="n">data_form</span><span class="o">=</span><span class="s2">&quot;main&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_form</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve the protein string identifiers.</span>

<span class="sd">    Args:</span>
<span class="sd">        data_form (str): Can be one of the following: &quot;main&quot;, &quot;restricted&quot;, &quot;picked&quot;, &quot;picked_removed&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of protein identifier strings.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; protein_strings = data.get_protein_identifiers(data_form=&quot;main&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span>
        <span class="c1"># All the data (unrestricted)</span>
        <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>
        <span class="n">prots</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

    <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;restricted&quot;</span><span class="p">:</span>
        <span class="c1"># Proteins that pass certain restriction criteria (peptide length, pep, qvalue)</span>
        <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span>
        <span class="n">prots</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

    <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;picked&quot;</span><span class="p">:</span>
        <span class="c1"># Here we look at proteins that are &#39;picked&#39; (aka the proteins that beat out their matching target/decoy)</span>
        <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span>
        <span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

    <span class="k">if</span> <span class="n">data_form</span> <span class="o">==</span> <span class="s2">&quot;picked_removed&quot;</span><span class="p">:</span>
        <span class="c1"># Here we look at the proteins that were removed due to picking (aka the proteins that</span>
        <span class="c1"># have a worse score than their target/decoy counterpart)</span>
        <span class="n">data_to_select</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span>
        <span class="n">prots</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_to_select</span><span class="p">]</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="n">prots</span>

    <span class="k">return</span> <span class="n">proteins</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_identifiers_from_psm_data">
<code class="highlight language-python"><span class="n">get_protein_identifiers_from_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a list of lists of all possible protein identifiers from the psm data.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of lists of protein strings.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_strings</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_protein_identifiers_from_psm_data</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_identifiers_from_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a list of lists of all possible protein identifiers from the psm data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of lists of protein strings.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; protein_strings = data.get_protein_identifiers_from_psm_data()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">proteins</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_information">
<code class="highlight language-python"><span class="n">get_protein_information</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_string</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve attributes for a specific scored protein.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>protein_string</strong> (<code>str</code>)  Protein Identifier String.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of protein attributes.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_attr</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_protein_information</span><span class="p">(</span><span class="n">protein_string</span><span class="o">=</span><span class="s2">&quot;RAF1_HUMAN|P04049&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_information</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_string</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve attributes for a specific scored protein.</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_string (str): Protein Identifier String.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of protein attributes.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; protein_attr = data.get_protein_information(protein_string=&quot;RAF1_HUMAN|P04049&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_scored_protein_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span>
    <span class="n">identifiers</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">protein_scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">group_identification</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">reviewed</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">reviewed</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="c1"># Peptide scores currently broken...</span>
    <span class="n">peptide_scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">peptide_scores</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">picked</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">picked</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>
    <span class="n">num_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">num_peptides</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_protein_data</span><span class="p">]</span>

    <span class="n">main_index</span> <span class="o">=</span> <span class="n">identifiers</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">protein_string</span><span class="p">)</span>

    <span class="n">list_structure</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;identifier&quot;</span><span class="p">,</span>
            <span class="s2">&quot;protein_score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;groups&quot;</span><span class="p">,</span>
            <span class="s2">&quot;reviewed&quot;</span><span class="p">,</span>
            <span class="s2">&quot;peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;peptide_scores&quot;</span><span class="p">,</span>
            <span class="s2">&quot;picked&quot;</span><span class="p">,</span>
            <span class="s2">&quot;num_peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="n">list_structure</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">protein_string</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_scores</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reviewed</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptides</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptide_scores</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">picked</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>
    <span class="n">list_structure</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_peptides</span><span class="p">[</span><span class="n">main_index</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">list_structure</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_information_dictionary">
<code class="highlight language-python"><span class="n">get_protein_information_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a dictionary of scores for each peptide.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>dict</code>  dictionary of scores for each protein.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_dict</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_protein_information_dictionary</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_information_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a dictionary of scores for each peptide.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: dictionary of scores for each protein.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; protein_dict = data.get_protein_information_dictionary()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">protein_psm_score_dictionary</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

    <span class="c1"># Loop through all Psms</span>
    <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
        <span class="c1"># Loop through all proteins</span>
        <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
            <span class="n">protein_psm_score_dictionary</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;peptide&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span>
                    <span class="s2">&quot;Qvalue&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span><span class="p">,</span>
                    <span class="s2">&quot;PosteriorErrorProbability&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">,</span>
                    <span class="s2">&quot;Percscore&quot;</span><span class="p">:</span> <span class="n">psms</span><span class="o">.</span><span class="n">percscore</span><span class="p">,</span>
                <span class="p">}</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">protein_psm_score_dictionary</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_protein_objects">
<code class="highlight language-python"><span class="n">get_protein_objects</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method retrieves protein objects. Either retrieves FDR restricted list of protien objects,
or retrieves all objects.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>fdr_restricted</strong> (<code>bool</code>)  True/False on whether to restrict the list of objects based on FDR.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects
    that have been grouped and sorted.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_protein_objects</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method retrieves protein objects. Either retrieves FDR restricted list of protien objects,</span>
<span class="sd">    or retrieves all objects.</span>

<span class="sd">    Args:</span>
<span class="sd">        fdr_restricted (bool): True/False on whether to restrict the list of objects based on FDR.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of scored [ProteinGroup][pyproteininference.physical.ProteinGroup] objects</span>
<span class="sd">            that have been grouped and sorted.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">false_discovery_rate</span><span class="p">:</span>
        <span class="n">false_discovery_rate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span>
    <span class="k">if</span> <span class="n">fdr_restricted</span><span class="p">:</span>
        <span class="n">protein_objects</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">proteins</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">q_value</span> <span class="o">&lt;=</span> <span class="n">false_discovery_rate</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span>

    <span class="k">return</span> <span class="n">protein_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_psm_data">
<code class="highlight language-python"><span class="n">get_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.
Retrieves restricted data if the data has been restricted or all of the data if the data has
not been restricted.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">psm_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a list of [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">    Retrieves restricted data if the data has been restricted or all of the data if the data has</span>
<span class="sd">    not been restricted.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; psm_data = data.get_psm_data()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Both main_data_restricted and main_data_form variables are empty. Please re-load the DataStore &quot;</span>
            <span class="s2">&quot;object with a properly loaded Reader object.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span><span class="p">:</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>

    <span class="k">return</span> <span class="n">psm_data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_q_values">
<code class="highlight language-python"><span class="n">get_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve a list of all q values for all PSMs.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of floats (q values).</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">q</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_q_values</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_q_values</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve a list of all q values for all PSMs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of floats (q values).</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; q = data.get_q_values()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">q_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">qvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">q_values</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.get_sorted_identifiers">
<code class="highlight language-python"><span class="n">get_sorted_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">scored</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves a sorted list of protein strings present in the analysis.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>scored</strong> (<code>bool</code>)  True/False to indicate if we should return scored or non-scored identifiers.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of sorted protein identifier strings.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">sorted_proteins</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get_sorted_identifiers</span><span class="p">(</span><span class="n">scored</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_sorted_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">scored</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves a sorted list of protein strings present in the analysis.</span>

<span class="sd">    Args:</span>
<span class="sd">        scored (bool): True/False to indicate if we should return scored or non-scored identifiers.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of sorted protein identifier strings.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; sorted_proteins = data.get_sorted_identifiers(scored=True)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">scored</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">:</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scoring_input</span><span class="p">()</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">]</span>

    <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>

    <span class="n">our_target_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
    <span class="n">our_decoy_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

    <span class="n">our_target_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">our_decoy_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_sp_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

    <span class="n">our_proteins_sorted</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">our_target_sp_proteins</span> <span class="o">+</span> <span class="n">our_decoy_sp_proteins</span> <span class="o">+</span> <span class="n">our_target_tr_proteins</span> <span class="o">+</span> <span class="n">our_decoy_tr_proteins</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">our_proteins_sorted</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.higher_or_lower">
<code class="highlight language-python"><span class="n">higher_or_lower</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to determine if a higher or lower score is better for a given combination of score input and score type.</p>
<p>This method sets the <code>high_low_better</code> Attribute for the DataStore object.</p>
<p>This method depends on the output from the Score class to be sorted properly from best to worst score.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>str</code>  String indicating "higher" or "lower" depending on if a higher or lower score is a
    better protein score.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">high_low</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">higher_or_lower</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to determine if a higher or lower score is better for a given combination of score input and score type.</span>

<span class="sd">    This method sets the `high_low_better` Attribute for the DataStore object.</span>

<span class="sd">    This method depends on the output from the Score class to be sorted properly from best to worst score.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: String indicating &quot;higher&quot; or &quot;lower&quot; depending on if a higher or lower score is a</span>
<span class="sd">            better protein score.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; high_low = data.higher_or_lower()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Determining If a higher or lower score is better based on scored proteins&quot;</span><span class="p">)</span>
        <span class="n">worst_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>
        <span class="n">best_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>

        <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span> <span class="o">&gt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">worst_score</span><span class="p">):</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span>

        <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">worst_score</span><span class="p">):</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;best score = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_score</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;worst score = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">worst_score</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">best_score</span> <span class="o">==</span> <span class="n">worst_score</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Best and Worst scores were identical, equal to </span><span class="si">{}</span><span class="s2">. Score type </span><span class="si">{}</span><span class="s2"> produced the error, &quot;</span>
                <span class="s2">&quot;please change psm_score type.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">best_score</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span> <span class="o">=</span> <span class="n">higher_or_lower</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">high_low_better</span>

    <span class="k">return</span> <span class="n">higher_or_lower</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.input_has_custom">
<code class="highlight language-python"><span class="n">input_has_custom</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that checks to see if the input data has custom score values.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">input_has_custom</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that checks to see if the input data has custom score values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">len_c</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">custom_score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">custom_score</span><span class="p">])</span>
    <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">len_c</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Custom value; Can restrict by Custom value&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Custom value; Cannot restrict by Custom value&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">status</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.input_has_pep">
<code class="highlight language-python"><span class="n">input_has_pep</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that checks to see if the input data has pep values.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">input_has_pep</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that checks to see if the input data has pep values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">len_pep</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">])</span>
    <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">len_pep</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Pep value; Can restrict by Pep value&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Pep value; Cannot restrict by Pep value&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">status</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.input_has_q">
<code class="highlight language-python"><span class="n">input_has_q</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that checks to see if the input data has q values.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">input_has_q</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that checks to see if the input data has q values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">len_q</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">qvalue</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">qvalue</span><span class="p">])</span>
    <span class="n">len_all</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">len_q</span> <span class="o">==</span> <span class="n">len_all</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Input has Q value; Can restrict by Q value&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">status</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Input does not have Q value; Cannot restrict by Q value&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">status</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.peptide_to_protein_dictionary">
<code class="highlight language-python"><span class="n">peptide_to_protein_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that returns a map of peptide strings to sets of protein strings and is essentially half of a
BiPartite graph.
This method sets the <code>peptide_protein_dictionary</code> Attribute for the DataStore object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>collections.defaultdict</code>  Dictionary of peptide strings (keys) that map to sets of protein strings based
    on the peptides and proteins found in the search. Peptide -&gt; set(Proteins).</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">peptide_protein_dict</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">peptide_to_protein_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that returns a map of peptide strings to sets of protein strings and is essentially half of a</span>
<span class="sd">    BiPartite graph.</span>
<span class="sd">    This method sets the `peptide_protein_dictionary` Attribute for the DataStore object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        collections.defaultdict: Dictionary of peptide strings (keys) that map to sets of protein strings based</span>
<span class="sd">            on the peptides and proteins found in the search. Peptide -&gt; set(Proteins).</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; peptide_protein_dict = data.peptide_to_protein_dictionary()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">res_pep_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>
    <span class="n">default_dict_peptides</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">peptide_objects</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
            <span class="n">cur_peptide</span> <span class="o">=</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
            <span class="k">if</span> <span class="n">cur_peptide</span> <span class="ow">in</span> <span class="n">res_pep_set</span><span class="p">:</span>
                <span class="n">default_dict_peptides</span><span class="p">[</span><span class="n">cur_peptide</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">prots</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">pass</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">peptide_protein_dictionary</span> <span class="o">=</span> <span class="n">default_dict_peptides</span>

    <span class="k">return</span> <span class="n">default_dict_peptides</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.protein_picker">
<code class="highlight language-python"><span class="n">protein_picker</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to run the protein picker algorithm.</p>
<p>Proteins must be scored first with <a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score.score_psms">score_psms</a>.</p>
<p>The algorithm will match target and decoy proteins identified from the PSMs from the search.
If a target and matching decoy is found then target/decoy competition is performed.
In the Target/Decoy pair the protein with the better score is kept and the one with the worse score is
discarded from the analysis.</p>
<p>The method sets the <code>picked_proteins_scored</code> and <code>picked_proteins_removed</code> variables for
the DataStore object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">data</span><span class="o">.</span><span class="n">protein_picker</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">protein_picker</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to run the protein picker algorithm.</span>

<span class="sd">    Proteins must be scored first with [score_psms][pyproteininference.scoring.Score.score_psms].</span>

<span class="sd">    The algorithm will match target and decoy proteins identified from the PSMs from the search.</span>
<span class="sd">    If a target and matching decoy is found then target/decoy competition is performed.</span>
<span class="sd">    In the Target/Decoy pair the protein with the better score is kept and the one with the worse score is</span>
<span class="sd">    discarded from the analysis.</span>

<span class="sd">    The method sets the `picked_proteins_scored` and `picked_proteins_removed` variables for</span>
<span class="sd">    the DataStore object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; data.protein_picker()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Protein Picker&quot;</span><span class="p">)</span>

    <span class="c1"># Use higher or lower class to determine if a higher protein score or lower protein score is better</span>
    <span class="c1"># based on the scoring method used</span>
    <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>
    <span class="c1"># Here we determine if a lower or higher score is better</span>
    <span class="c1"># Since all input is ordered from best to worst we can do the following</span>

    <span class="n">index_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># data.scored_proteins is simply a list of Protein objects...</span>
    <span class="c1"># Create list of all decoy proteins</span>
    <span class="n">decoy_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">identifier</span><span class="p">]</span>
    <span class="c1"># Create a list of all potential matching targets (some of these may not exist in the search)</span>
    <span class="n">matching_targets</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">decoy_proteins</span><span class="p">]</span>

    <span class="c1"># Create a list of all the proteins from the scored data</span>
    <span class="n">all_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> proteins scored&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_proteins</span><span class="p">)))</span>

    <span class="n">total_targets</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">total_decoys</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">decoys_removed</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">targets_removed</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Loop over all decoys identified in the search</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Picking Proteins...&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoy_proteins</span><span class="p">)):</span>
        <span class="n">cur_decoy_index</span> <span class="o">=</span> <span class="n">all_proteins</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">decoy_proteins</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">cur_decoy_protein_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="n">cur_decoy_index</span><span class="p">]</span>
        <span class="n">total_decoys</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

        <span class="c1"># Try, Except here because the matching target to the decoy may not be a result from the search</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">cur_target_index</span> <span class="o">=</span> <span class="n">all_proteins</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">matching_targets</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">cur_target_protein_object</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">[</span><span class="n">cur_target_index</span><span class="p">]</span>
            <span class="n">total_targets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">score</span> <span class="o">&gt;</span> <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>
                    <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                    <span class="n">decoys_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                    <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                    <span class="n">targets_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                    <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">score</span> <span class="o">&lt;</span> <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">score</span><span class="p">:</span>
                    <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                    <span class="n">decoys_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_decoy_index</span><span class="p">)</span>
                    <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">index_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                    <span class="n">targets_removed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_target_index</span><span class="p">)</span>
                    <span class="n">cur_decoy_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">cur_target_protein_object</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">pass</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> total decoy proteins&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_decoys</span><span class="p">)))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> matching target proteins also found in search&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_targets</span><span class="p">)))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> decoy proteins to be removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">decoys_removed</span><span class="p">)))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> target proteins to be removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">targets_removed</span><span class="p">)))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing Lower Scoring Proteins...&quot;</span><span class="p">)</span>
    <span class="n">picked_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">removed_proteins</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">protein_objects</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">scored_proteins</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">protein_objects</span><span class="o">.</span><span class="n">picked</span><span class="p">:</span>
            <span class="n">picked_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_objects</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">removed_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_objects</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_scored</span> <span class="o">=</span> <span class="n">picked_list</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picked_proteins_removed</span> <span class="o">=</span> <span class="n">removed_proteins</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished Removing Proteins&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.protein_to_peptide_dictionary">
<code class="highlight language-python"><span class="n">protein_to_peptide_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that returns a map of protein strings to sets of peptide strings and is essentially half
 of a BiPartite graph.
This method sets the <code>protein_peptide_dictionary</code> Attribute for the DataStore object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>collections.defaultdict</code>  Dictionary of protein strings (keys) that map to sets of peptide strings based
on the peptides and proteins found in the search. Protein -&gt; set(Peptides).</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">protein_peptide_dict</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">protein_to_peptide_dictionary</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that returns a map of protein strings to sets of peptide strings and is essentially half</span>
<span class="sd">     of a BiPartite graph.</span>
<span class="sd">    This method sets the `protein_peptide_dictionary` Attribute for the DataStore object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        collections.defaultdict: Dictionary of protein strings (keys) that map to sets of peptide strings based</span>
<span class="sd">        on the peptides and proteins found in the search. Protein -&gt; set(Peptides).</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; protein_peptide_dict = data.protein_to_peptide_dictionary()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_psm_data</span><span class="p">()</span>

    <span class="n">res_pep_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>
    <span class="n">default_dict_proteins</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">peptide_objects</span> <span class="ow">in</span> <span class="n">psm_data</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
            <span class="n">cur_peptide</span> <span class="o">=</span> <span class="n">peptide_objects</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
            <span class="k">if</span> <span class="n">cur_peptide</span> <span class="ow">in</span> <span class="n">res_pep_set</span><span class="p">:</span>
                <span class="n">default_dict_proteins</span><span class="p">[</span><span class="n">prots</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">cur_peptide</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span> <span class="o">=</span> <span class="n">default_dict_proteins</span>

    <span class="k">return</span> <span class="n">default_dict_proteins</span>
</code></pre></div>
        </details>
    </div>

  </div>




  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.restrict_psm_data">
<code class="highlight language-python"><span class="n">restrict_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">remove1pep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to restrict the input of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a>  objects.
This method is central to the pyproteininference module and is able to restrict the Psm data by:
Q value, Pep Value, Percolator Score, Peptide Length, and Custom Score Input.
Restriction values are pulled from
the <a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a>
object.</p>
<p>This method sets the <code>main_data_restricted</code> and <code>restricted_peptides</code> Attributes for the DataStore object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>remove1pep</strong> (<code>bool</code>)  True/False on whether or not to remove PEP values that equal 1 even if other restrictions
are set to not restrict.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">data</span><span class="o">.</span><span class="n">restrict_psm_data</span><span class="p">(</span><span class="n">remove1pep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">restrict_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">remove1pep</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to restrict the input of [Psm][pyproteininference.physical.Psm]  objects.</span>
<span class="sd">    This method is central to the pyproteininference module and is able to restrict the Psm data by:</span>
<span class="sd">    Q value, Pep Value, Percolator Score, Peptide Length, and Custom Score Input.</span>
<span class="sd">    Restriction values are pulled from</span>
<span class="sd">    the [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter]</span>
<span class="sd">    object.</span>

<span class="sd">    This method sets the `main_data_restricted` and `restricted_peptides` Attributes for the DataStore object.</span>

<span class="sd">    Args:</span>
<span class="sd">        remove1pep (bool): True/False on whether or not to remove PEP values that equal 1 even if other restrictions</span>
<span class="sd">            are set to not restrict.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; data.restrict_psm_data(remove1pep=True)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Validate that we have the main data variable</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_main_data_form</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Restricting PSM data&quot;</span><span class="p">)</span>

    <span class="n">peptide_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_peptide_length</span>
    <span class="n">posterior_error_prob_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_pep</span>
    <span class="n">q_value_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_q</span>
    <span class="n">custom_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">restrict_custom</span>

    <span class="n">main_psm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of main data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">main_data_form</span><span class="p">)))</span>
    <span class="c1"># If restrict_main_data is called, we automatically discard everything that has a PEP of 1</span>
    <span class="k">if</span> <span class="n">remove1pep</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span><span class="p">:</span>
        <span class="n">main_psm_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">main_psm_data</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Restrict peptide length and posterior error probability</span>
    <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span>
                <span class="n">posterior_error_prob_threshold</span>
            <span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict peptide length only</span>
    <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span><span class="p">:</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict peptide length, posterior error probability, and qvalue</span>
    <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span>
                <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">)</span>
                <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">)</span>
            <span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict peptide length and qvalue</span>
    <span class="k">if</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms</span><span class="o">.</span><span class="n">stripped_peptide</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict posterior error probability and q value</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">)</span> <span class="ow">and</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict qvalue only</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_value_threshold</span><span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict posterior error probability only</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">main_psm_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">&lt;</span> <span class="nb">float</span><span class="p">(</span><span class="n">posterior_error_prob_threshold</span><span class="p">):</span>
                <span class="n">restricted_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="c1"># Restrict nothing... (only PEP gets restricted - takes everything less than 1)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide_length</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">posterior_error_prob_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">q_value_threshold</span><span class="p">:</span>
        <span class="n">restricted_data</span> <span class="o">=</span> <span class="n">main_psm_data</span>

    <span class="k">if</span> <span class="n">custom_threshold</span><span class="p">:</span>
        <span class="n">custom_restricted</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">&lt;=</span> <span class="n">custom_threshold</span><span class="p">:</span>
                    <span class="n">custom_restricted</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE_SCORE_TYPE</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">psms</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">&gt;=</span> <span class="n">custom_threshold</span><span class="p">:</span>
                    <span class="n">custom_restricted</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="n">restricted_data</span> <span class="o">=</span> <span class="n">custom_restricted</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">main_data_restricted</span> <span class="o">=</span> <span class="n">restricted_data</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of restricted data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">restricted_data</span><span class="p">)))</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">restricted_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">restricted_data</span><span class="p">]</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.sort_protein_group_objects">
<code class="highlight language-python"><span class="n">sort_protein_group_objects</span><span class="p">(</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Class Method to sort a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects by
score and number of peptides.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>protein_group_objects</strong> (<code>list</code>)  list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p></li>
            <li><p><strong>higher_or_lower</strong> (<code>str</code>)  String to indicate if a "higher" or "lower" protein score is "better".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of sorted <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">list_of_group_objects</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">list_of_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="s2">&quot;higher&quot;</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">sort_protein_group_objects</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class Method to sort a list of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects by</span>
<span class="sd">    score and number of peptides.</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_group_objects (list): list of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>
<span class="sd">        higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of sorted [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; list_of_group_objects = pyproteininference.datastore.DataStore.sort_protein_group_objects(</span>
<span class="sd">        &gt;&gt;&gt;     protein_group_objects=list_of_group_objects, higher_or_lower=&quot;higher&quot;</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>

        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                <span class="o">-</span><span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>

        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                <span class="n">k</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.sort_protein_objects">
<code class="highlight language-python"><span class="n">sort_protein_objects</span><span class="p">(</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Class Method to sort a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects by score and number of
peptides.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>grouped_protein_objects</strong> (<code>list</code>)  list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></li>
            <li><p><strong>higher_or_lower</strong> (<code>str</code>)  String to indicate if a "higher" or "lower" protein score is "better".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  list of sorted <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">scores_grouped</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">scores_grouped</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="s2">&quot;higher&quot;</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">sort_protein_objects</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class Method to sort a list of [Protein][pyproteininference.physical.Protein] objects by score and number of</span>
<span class="sd">    peptides.</span>

<span class="sd">    Args:</span>
<span class="sd">        grouped_protein_objects (list): list of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">        higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: list of sorted [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; scores_grouped = pyproteininference.datastore.DataStore.sort_protein_objects(</span>
<span class="sd">        &gt;&gt;&gt;     grouped_protein_objects=scores_grouped, higher_or_lower=&quot;higher&quot;</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="o">-</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">grouped_protein_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.sort_protein_strings">
<code class="highlight language-python"><span class="n">sort_protein_strings</span><span class="p">(</span><span class="n">protein_string_list</span><span class="p">,</span> <span class="n">sp_proteins</span><span class="p">,</span> <span class="n">decoy_symbol</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Method that sorts protein strings in the following order: Target Reviewed, Decoy Reviewed, Target Unreviewed,
 Decoy Unreviewed.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>protein_string_list</strong> (<code>list</code>)  List of Protein Strings.</p></li>
            <li><p><strong>sp_proteins</strong> (<code>set</code>)  Set of Reviewed Protein Strings.</p></li>
            <li><p><strong>decoy_symbol</strong> (<code>str</code>)  Symbol to denote a decoy protein identifier IE "##".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of sorted protein strings.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">list_of_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">protein_string_list</span><span class="o">=</span><span class="n">protein_string_list</span><span class="p">,</span> <span class="n">sp_proteins</span><span class="o">=</span><span class="n">sp_proteins</span><span class="p">,</span> <span class="n">decoy_symbol</span><span class="o">=</span><span class="s2">&quot;##&quot;</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">sort_protein_strings</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_string_list</span><span class="p">,</span> <span class="n">sp_proteins</span><span class="p">,</span> <span class="n">decoy_symbol</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that sorts protein strings in the following order: Target Reviewed, Decoy Reviewed, Target Unreviewed,</span>
<span class="sd">     Decoy Unreviewed.</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_string_list (list): List of Protein Strings.</span>
<span class="sd">        sp_proteins (set): Set of Reviewed Protein Strings.</span>
<span class="sd">        decoy_symbol (str): Symbol to denote a decoy protein identifier IE &quot;##&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of sorted protein strings.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; list_of_group_objects = datastore.DataStore.sort_protein_strings(</span>
<span class="sd">        &gt;&gt;&gt;     protein_string_list=protein_string_list, sp_proteins=sp_proteins, decoy_symbol=&quot;##&quot;</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">our_target_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>
    <span class="n">our_decoy_sp_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

    <span class="n">our_target_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">our_decoy_tr_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_string_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sp_proteins</span> <span class="ow">and</span> <span class="n">decoy_symbol</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span>

    <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">our_target_sp_proteins</span> <span class="o">+</span> <span class="n">our_decoy_sp_proteins</span> <span class="o">+</span> <span class="n">our_target_tr_proteins</span> <span class="o">+</span> <span class="n">our_decoy_tr_proteins</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">identifiers_sorted</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.sort_protein_sub_groups">
<code class="highlight language-python"><span class="n">sort_protein_sub_groups</span><span class="p">(</span><span class="n">protein_list</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Method to sort protein sub lists.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>protein_list</strong> (<code>list</code>)  List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects to be sorted.</p></li>
            <li><p><strong>higher_or_lower</strong> (<code>str</code>)  String to indicate if a "higher" or "lower" protein score is "better".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects to be sorted by score and number of
peptides.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">sort_protein_sub_groups</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">protein_list</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to sort protein sub lists.</span>

<span class="sd">    Args:</span>
<span class="sd">        protein_list (list): List of [Protein][pyproteininference.physical.Protein] objects to be sorted.</span>
<span class="sd">        higher_or_lower (str): String to indicate if a &quot;higher&quot; or &quot;lower&quot; protein score is &quot;better&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of [Protein][pyproteininference.physical.Protein] objects to be sorted by score and number of</span>
<span class="sd">        peptides.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Sort the groups based on higher or lower indication, secondarily sort the groups based on number of unique</span>
    <span class="c1"># peptides</span>
    <span class="c1"># We use the index [1:] as we do not wish to sort the lead protein...</span>
    <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
        <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">),</span> <span class="o">-</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="bp">cls</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
        <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">protein_list</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.unique_to_leads_peptides">
<code class="highlight language-python"><span class="n">unique_to_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to retrieve peptides that are unique based on the data from the searches
(Not based on the database digestion).</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>set</code>  a Set of peptide strings</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span> <span class="o">=</span> <span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">unique_peps</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">unique_to_leads_peptides</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">unique_to_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to retrieve peptides that are unique based on the data from the searches</span>
<span class="sd">    (Not based on the database digestion).</span>

<span class="sd">    Returns:</span>
<span class="sd">        set: a Set of peptide strings</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; data = pyproteininference.datastore.DataStore(reader = reader, digest=digest)</span>
<span class="sd">        &gt;&gt;&gt; unique_peps = data.unique_to_leads_peptides()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">:</span>
        <span class="n">lead_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">]</span>
        <span class="n">flat_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">lead_peptides</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">counted_peps</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">flat_peptides</span><span class="p">)</span>
        <span class="n">unique_to_leads_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">counted_peps</span> <span class="k">if</span> <span class="n">counted_peps</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">unique_to_leads_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">unique_to_leads_peptides</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.validate_digest">
<code class="highlight language-python"><span class="n">validate_digest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that validates the <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">validate_digest</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that validates the [Digest object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_reviewed_v_unreviewed</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_target_decoy_split</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.datastore.DataStore.validate_psm_data">
<code class="highlight language-python"><span class="n">validate_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that validates the PSM data.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/datastore.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">validate_psm_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that validates the PSM data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_decoys_from_data</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_isoform_from_data</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.export">
        <code>export</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.export.Export">
        <code>
Export        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Class that handles exporting protein inference results to filesystem as csv files.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p></td>
      </tr>
      <tr>
        <td><code>filepath</code></td>
        <td><code>str</code></td>
        <td><p>Path to file to be written.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Export</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class that handles exporting protein inference results to filesystem as csv files.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        filepath (str): Path to file to be written.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">EXPORT_LEADS</span> <span class="o">=</span> <span class="s2">&quot;leads&quot;</span>
    <span class="n">EXPORT_ALL</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
    <span class="n">EXPORT_COMMA_SEP</span> <span class="o">=</span> <span class="s2">&quot;comma_sep&quot;</span>
    <span class="n">EXPORT_Q_VALUE_COMMA_SEP</span> <span class="o">=</span> <span class="s2">&quot;q_value_comma_sep&quot;</span>
    <span class="n">EXPORT_Q_VALUE</span> <span class="o">=</span> <span class="s2">&quot;q_value&quot;</span>
    <span class="n">EXPORT_Q_VALUE_ALL</span> <span class="o">=</span> <span class="s2">&quot;q_value_all&quot;</span>
    <span class="n">EXPORT_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;peptides&quot;</span>
    <span class="n">EXPORT_PSMS</span> <span class="o">=</span> <span class="s2">&quot;psms&quot;</span>
    <span class="n">EXPORT_PSM_IDS</span> <span class="o">=</span> <span class="s2">&quot;psm_ids&quot;</span>
    <span class="n">EXPORT_LONG</span> <span class="o">=</span> <span class="s2">&quot;long&quot;</span>

    <span class="n">EXPORT_TYPES</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">EXPORT_LEADS</span><span class="p">,</span>
        <span class="n">EXPORT_ALL</span><span class="p">,</span>
        <span class="n">EXPORT_COMMA_SEP</span><span class="p">,</span>
        <span class="n">EXPORT_Q_VALUE_COMMA_SEP</span><span class="p">,</span>
        <span class="n">EXPORT_Q_VALUE</span><span class="p">,</span>
        <span class="n">EXPORT_Q_VALUE_ALL</span><span class="p">,</span>
        <span class="n">EXPORT_PEPTIDES</span><span class="p">,</span>
        <span class="n">EXPORT_PSMS</span><span class="p">,</span>
        <span class="n">EXPORT_PSM_IDS</span><span class="p">,</span>
        <span class="n">EXPORT_LONG</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method for the Export class.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; export = pyproteininference.export.Export(data=data)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">export_to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">export_type</span><span class="o">=</span><span class="s2">&quot;q_value&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that dispatches to one of the many export methods given an export_type input.</span>

<span class="sd">        filepath is determined based on directory arg and information from</span>
<span class="sd">        [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        This method sets the `filepath` variable.</span>

<span class="sd">        Args:</span>
<span class="sd">            output_filename (str): Filepath to write to. If set as None will auto generate filename and</span>
<span class="sd">                will write to directory variable.</span>
<span class="sd">            directory (str): Directory to write the result file to. If None, will write to current working directory.</span>
<span class="sd">            export_type (str): Must be a value in `EXPORT_TYPES` and determines the output format.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; export = pyproteininference.export.Export(data=data)</span>
<span class="sd">            &gt;&gt;&gt; export.export_to_csv(output_filename=None, directory=&quot;/path/to/output/dir/&quot;, export_type=&quot;psms&quot;)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">directory</span><span class="p">:</span>
            <span class="n">directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>

        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
        <span class="n">tag</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">tag</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_LEADS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_leads_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_leads_restricted</span><span class="p">(</span><span class="n">filename_out</span><span class="o">=</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_ALL</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_all_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_all_restricted</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_COMMA_SEP</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_comma_sep_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_comma_sep_restricted</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE_COMMA_SEP</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_comma_sep_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_comma_sep</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE_ALL</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_all_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_all</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PEPTIDES</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_peptides_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_peptides</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PSMS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_psms_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_psms</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PSM_IDS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_psm_ids_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_psm_ids</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_LONG</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_long_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
                <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_long</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="s2">&quot;protein_inference_results.csv&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">complete_filepath</span>

    <span class="k">def</span> <span class="nf">csv_export_all_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs a subset of the passing proteins based on FDR.</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">prots</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">prots</span><span class="o">.</span><span class="n">peptides</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_leads_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs a subset of the passing proteins based on FDR.</span>
<span class="sd">        Only Proteins that pass FDR will be output and only Lead proteins will be output</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_comma_sep_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs a subset of the passing proteins based on FDR.</span>
<span class="sd">        Only Proteins that pass FDR will be output and only Lead proteins will be output.</span>
<span class="sd">        Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Other_Potential_Identifiers&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">prots</span> <span class="o">==</span> <span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">prots</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_leads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values.</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="n">peptides</span> <span class="o">=</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">peptides</span><span class="p">):</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_comma_sep</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values.</span>
<span class="sd">        Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Other_Potential_Identifiers&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">other_prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">other_prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_all</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all proteins with Q values.</span>
<span class="sd">        Non Lead proteins are also output - entire group gets output.</span>
<span class="sd">        Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">        This method returns a non-square CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">proteins</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">proteins</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_all_proteologic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">proteins</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">proteins</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_leads_long</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values.</span>

<span class="sd">        This method returns a long formatted result file with one peptide on each row.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">        This method outputs unique peptides per protein.</span>

<span class="sd">        This method returns a rectangular CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>
<span class="sd">            peptide_delimiter (str): String to separate peptides by in the &quot;Peptides&quot; column of the csv file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="n">peptides</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span><span class="p">)))</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptides</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_leads_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">        This method outputs all PSMs for the protein not just unique peptide identifiers.</span>

<span class="sd">        This method returns a rectangular CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>
<span class="sd">            peptide_delimiter (str): String to separate peptides by in the &quot;Peptides&quot; column of the csv file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="n">psms</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">psms</span><span class="p">]))</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">csv_export_q_value_leads_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">        Psms are output as the psm_id value. So sequence information is not output.</span>

<span class="sd">        This method returns a rectangular CSV file.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename_out (str): Filename for the data to be written to.</span>
<span class="sd">            peptide_delimiter (str): String to separate psm_ids by in the &quot;Peptides&quot; column of the csv file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="n">psms</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">get_psm_ids</span><span class="p">()))</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">




















  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method for the Export class.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method for the Export class.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; export = pyproteininference.export.Export(data=data)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_all_restricted">
<code class="highlight language-python"><span class="n">csv_export_all_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs a subset of the passing proteins based on FDR.</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_all_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs a subset of the passing proteins based on FDR.</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">prots</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">prots</span><span class="o">.</span><span class="n">peptides</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_comma_sep_restricted">
<code class="highlight language-python"><span class="n">csv_export_comma_sep_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs a subset of the passing proteins based on FDR.
Only Proteins that pass FDR will be output and only Lead proteins will be output.
Proteins in the groups of lead proteins will also be output in the same row as the lead.</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_comma_sep_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs a subset of the passing proteins based on FDR.</span>
<span class="sd">    Only Proteins that pass FDR will be output and only Lead proteins will be output.</span>
<span class="sd">    Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Other_Potential_Identifiers&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">prots</span> <span class="o">==</span> <span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">prots</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_leads_restricted">
<code class="highlight language-python"><span class="n">csv_export_leads_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs a subset of the passing proteins based on FDR.
Only Proteins that pass FDR will be output and only Lead proteins will be output</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_leads_restricted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs a subset of the passing proteins based on FDR.</span>
<span class="sd">    Only Proteins that pass FDR will be output and only Lead proteins will be output</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span><span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="n">protein_objects</span><span class="p">:</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">group_identification</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_all">
<code class="highlight language-python"><span class="n">csv_export_q_value_all</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all proteins with Q values.
Non Lead proteins are also output - entire group gets output.
Proteins in the groups of lead proteins will also be output in the same row as the lead.</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_all</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all proteins with Q values.</span>
<span class="sd">    Non Lead proteins are also output - entire group gets output.</span>
<span class="sd">    Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">proteins</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">proteins</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">proteins</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>




  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_comma_sep">
<code class="highlight language-python"><span class="n">csv_export_q_value_comma_sep</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values.
Proteins in the groups of lead proteins will also be output in the same row as the lead.</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_comma_sep</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values.</span>
<span class="sd">    Proteins in the groups of lead proteins will also be output in the same row as the lead.</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Other_Potential_Identifiers&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">other_prots</span> <span class="ow">in</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">other_prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_leads">
<code class="highlight language-python"><span class="n">csv_export_q_value_leads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values.</p>
<p>This method returns a non-square CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_leads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values.</span>

<span class="sd">    This method returns a non-square CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span>
        <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">peptides</span><span class="p">):</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_leads_long">
<code class="highlight language-python"><span class="n">csv_export_q_value_leads_long</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values.</p>
<p>This method returns a long formatted result file with one peptide on each row.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_leads_long</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values.</span>

<span class="sd">    This method returns a long formatted result file with one peptide on each row.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
            <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peps</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_leads_peptides">
<code class="highlight language-python"><span class="n">csv_export_q_value_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values in rectangular format.
This method outputs unique peptides per protein.</p>
<p>This method returns a rectangular CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
            <li><p><strong>peptide_delimiter</strong> (<code>str</code>)  String to separate peptides by in the "Peptides" column of the csv file</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_leads_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">    This method outputs unique peptides per protein.</span>

<span class="sd">    This method returns a rectangular CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>
<span class="sd">        peptide_delimiter (str): String to separate peptides by in the &quot;Peptides&quot; column of the csv file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">peptides</span><span class="p">)))</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptides</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_leads_psm_ids">
<code class="highlight language-python"><span class="n">csv_export_q_value_leads_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values in rectangular format.
Psms are output as the psm_id value. So sequence information is not output.</p>
<p>This method returns a rectangular CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
            <li><p><strong>peptide_delimiter</strong> (<code>str</code>)  String to separate psm_ids by in the "Peptides" column of the csv file.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_leads_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">    Psms are output as the psm_id value. So sequence information is not output.</span>

<span class="sd">    This method returns a rectangular CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>
<span class="sd">        peptide_delimiter (str): String to separate psm_ids by in the &quot;Peptides&quot; column of the csv file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
        <span class="n">psms</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">get_psm_ids</span><span class="p">()))</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.csv_export_q_value_leads_psms">
<code class="highlight language-python"><span class="n">csv_export_q_value_leads_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that outputs all lead proteins with Q values in rectangular format.
This method outputs all PSMs for the protein not just unique peptide identifiers.</p>
<p>This method returns a rectangular CSV file.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>filename_out</strong> (<code>str</code>)  Filename for the data to be written to.</p></li>
            <li><p><strong>peptide_delimiter</strong> (<code>str</code>)  String to separate peptides by in the "Peptides" column of the csv file.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">csv_export_q_value_leads_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_out</span><span class="p">,</span> <span class="n">peptide_delimiter</span><span class="o">=</span><span class="s2">&quot; &quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that outputs all lead proteins with Q values in rectangular format.</span>
<span class="sd">    This method outputs all PSMs for the protein not just unique peptide identifiers.</span>

<span class="sd">    This method returns a rectangular CSV file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_out (str): Filename for the data to be written to.</span>
<span class="sd">        peptide_delimiter (str): String to separate peptides by in the &quot;Peptides&quot; column of the csv file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_export_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Score&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Q_Value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number_of_Peptides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Identifier_Type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GroupID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Peptides&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">groups</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
        <span class="n">lead_protein</span> <span class="o">=</span> <span class="n">groups</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein_export_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">])</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lead_protein</span><span class="o">.</span><span class="n">num_peptides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Reviewed&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unreviewed&quot;</span><span class="p">)</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">groups</span><span class="o">.</span><span class="n">number_id</span><span class="p">)</span>
        <span class="n">psms</span> <span class="o">=</span> <span class="n">peptide_delimiter</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lead_protein</span><span class="o">.</span><span class="n">psms</span><span class="p">]))</span>
        <span class="n">protein_export_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerows</span><span class="p">(</span><span class="n">protein_export_list</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.export.Export.export_to_csv">
<code class="highlight language-python"><span class="n">export_to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">export_type</span><span class="o">=</span><span class="s1">&#39;q_value&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method that dispatches to one of the many export methods given an export_type input.</p>
<p>filepath is determined based on directory arg and information from
<a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>
<p>This method sets the <code>filepath</code> variable.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>output_filename</strong> (<code>str</code>)  Filepath to write to. If set as None will auto generate filename and
will write to directory variable.</p></li>
            <li><p><strong>directory</strong> (<code>str</code>)  Directory to write the result file to. If None, will write to current working directory.</p></li>
            <li><p><strong>export_type</strong> (<code>str</code>)  Must be a value in <code>EXPORT_TYPES</code> and determines the output format.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">export</span><span class="o">.</span><span class="n">export_to_csv</span><span class="p">(</span><span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="s2">&quot;/path/to/output/dir/&quot;</span><span class="p">,</span> <span class="n">export_type</span><span class="o">=</span><span class="s2">&quot;psms&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/export.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">export_to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">export_type</span><span class="o">=</span><span class="s2">&quot;q_value&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method that dispatches to one of the many export methods given an export_type input.</span>

<span class="sd">    filepath is determined based on directory arg and information from</span>
<span class="sd">    [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    This method sets the `filepath` variable.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_filename (str): Filepath to write to. If set as None will auto generate filename and</span>
<span class="sd">            will write to directory variable.</span>
<span class="sd">        directory (str): Directory to write the result file to. If None, will write to current working directory.</span>
<span class="sd">        export_type (str): Must be a value in `EXPORT_TYPES` and determines the output format.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; export = pyproteininference.export.Export(data=data)</span>
<span class="sd">        &gt;&gt;&gt; export.export_to_csv(output_filename=None, directory=&quot;/path/to/output/dir/&quot;, export_type=&quot;psms&quot;)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">directory</span><span class="p">:</span>
        <span class="n">directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>

    <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
    <span class="n">tag</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">tag</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_LEADS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_leads_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_leads_restricted</span><span class="p">(</span><span class="n">filename_out</span><span class="o">=</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_ALL</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_all_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_all_restricted</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_COMMA_SEP</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_comma_sep_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_comma_sep_restricted</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE_COMMA_SEP</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_comma_sep_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_comma_sep</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_Q_VALUE_ALL</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_all_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_all</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PEPTIDES</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_peptides_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_peptides</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PSMS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_psms_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_psms</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PSM_IDS</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_leads_psm_ids_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_psm_ids</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_LONG</span> <span class="o">==</span> <span class="n">export_type</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_q_value_long_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tag</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">psm_score</span><span class="p">)</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_filename</span><span class="p">:</span>
            <span class="n">complete_filepath</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting Protein Inference Data to File: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">csv_export_q_value_leads_long</span><span class="p">(</span><span class="n">complete_filepath</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">complete_filepath</span> <span class="o">=</span> <span class="s2">&quot;protein_inference_results.csv&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">complete_filepath</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.heuristic">
        <code>heuristic</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline">
        <code>
HeuristicPipeline            (<a class="autorefs autorefs-internal" title="pyproteininference.pipeline.ProteinInferencePipeline" href="#pyproteininference.pipeline.ProteinInferencePipeline">ProteinInferencePipeline</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>This is the Protein Inference Heuristic class which houses the logic to run the Protein Inference Heuristic method
 to determine the best inference method for the given data.
Logic is executed in the <a class="autorefs autorefs-internal" href="#pyproteininference.heuristic.HeuristicPipeline.execute">execute</a> method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>parameter_file</code></td>
        <td><code>str</code></td>
        <td><p>Path to Protein Inference Yaml Parameter File.</p></td>
      </tr>
      <tr>
        <td><code>database_file</code></td>
        <td><code>str</code></td>
        <td><p>Path to Fasta database used in proteomics search.</p></td>
      </tr>
      <tr>
        <td><code>target_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Target Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>decoy_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Decoy Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>combined_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Combined Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>target_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Target Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>decoy_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Decoy Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>combined_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Combined Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>output_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory where output will be written.</p></td>
      </tr>
      <tr>
        <td><code>output_filename</code></td>
        <td><code>str</code></td>
        <td><p>Path to Filename where output will be written. Will override output_directory.</p></td>
      </tr>
      <tr>
        <td><code>id_splitting</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether to split protein IDs in the digest.
Advanced usage only.</p></td>
      </tr>
      <tr>
        <td><code>append_alt_from_db</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether to append
alternative proteins from the DB digestion in Reader class.</p></td>
      </tr>
      <tr>
        <td><code>pdf_filename</code></td>
        <td><code>str</code></td>
        <td><p>Filepath to be written to by Heuristic Plotting method.
This is optional and a default filename will be created in output_directory if this is left as None.</p></td>
      </tr>
      <tr>
        <td><code>inference_method_list</code></td>
        <td><code>list</code></td>
        <td><p>List of inference methods used in heuristic determination.</p></td>
      </tr>
      <tr>
        <td><code>datastore_dict</code></td>
        <td><code>dict</code></td>
        <td><p>Dictionary of <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore</a>
objects generated in heuristic determination with the inference method as the key of each entry.</p></td>
      </tr>
      <tr>
        <td><code>selected_methods</code></td>
        <td><code>list</code></td>
        <td><p>a list of String representations of the selected inference methods based on the
heuristic.</p></td>
      </tr>
      <tr>
        <td><code>selected_datastores</code></td>
        <td><code>dict</code></td>
        <td><p>a Dictionary of <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a> objects as selected by the
heuristic.</p></td>
      </tr>
      <tr>
        <td><code>output_type</code></td>
        <td><code>str</code></td>
        <td><p>How to output results. Can either be "all" or "optimal". Will either output all results
or will only output the optimal results.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">HeuristicPipeline</span><span class="p">(</span><span class="n">ProteinInferencePipeline</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is the Protein Inference Heuristic class which houses the logic to run the Protein Inference Heuristic method</span>
<span class="sd">     to determine the best inference method for the given data.</span>
<span class="sd">    Logic is executed in the [execute][pyproteininference.heuristic.HeuristicPipeline.execute] method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">        database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">        target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">        decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">        combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">        target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">        decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">        combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">        output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">        output_filename (str): Path to Filename where output will be written. Will override output_directory.</span>
<span class="sd">        id_splitting (bool): True/False on whether to split protein IDs in the digest.</span>
<span class="sd">            Advanced usage only.</span>
<span class="sd">        append_alt_from_db (bool): True/False on whether to append</span>
<span class="sd">            alternative proteins from the DB digestion in Reader class.</span>
<span class="sd">        pdf_filename (str): Filepath to be written to by Heuristic Plotting method.</span>
<span class="sd">            This is optional and a default filename will be created in output_directory if this is left as None.</span>
<span class="sd">        inference_method_list (list): List of inference methods used in heuristic determination.</span>
<span class="sd">        datastore_dict (dict): Dictionary of [DataStore][pyproteininference.datastore.DataStore]</span>
<span class="sd">            objects generated in heuristic determination with the inference method as the key of each entry.</span>
<span class="sd">        selected_methods (list): a list of String representations of the selected inference methods based on the</span>
<span class="sd">            heuristic.</span>
<span class="sd">        selected_datastores (dict):</span>
<span class="sd">            a Dictionary of [DataStore object][pyproteininference.datastore.DataStore] objects as selected by the</span>
<span class="sd">            heuristic.</span>
<span class="sd">        output_type (str): How to output results. Can either be &quot;all&quot; or &quot;optimal&quot;. Will either output all results</span>
<span class="sd">            or will only output the optimal results.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">RATIO_CONSTANT</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">OUTPUT_TYPES</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="s2">&quot;optimal&quot;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">parameter_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">            database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">            target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">            decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">            combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">            target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">            decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">            combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">            output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">            output_filename (str): Path to Filename where output will be written.</span>
<span class="sd">                Will override output_directory.</span>
<span class="sd">            id_splitting (bool): True/False on whether to split protein IDs in the digest.</span>
<span class="sd">                Advanced usage only.</span>
<span class="sd">            append_alt_from_db (bool): True/False on whether to append alternative proteins</span>
<span class="sd">                from the DB digestion in Reader class.</span>
<span class="sd">            pdf_filename (str): Filepath to be written to by Heuristic Plotting method.</span>
<span class="sd">                This is optional and a default filename will be created in output_directory if this is left as None</span>
<span class="sd">            output_type (str): How to output results. Can either be &quot;all&quot; or &quot;optimal&quot;. Will either output all results</span>
<span class="sd">                        or will only output the optimal results.</span>

<span class="sd">        Returns:</span>
<span class="sd">            HeuristicPipeline: [HeuristicPipeline][pyproteininference.heuristic.HeuristicPipeline] object</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; heuristic = pyproteininference.heuristic.HeuristicPipeline(</span>
<span class="sd">            &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">            &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">            &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">            &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">            &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">            &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">            &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">            &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">            &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">            &gt;&gt;&gt;     pdf_filename=pdf_filename,</span>
<span class="sd">            &gt;&gt;&gt;     output_type=&quot;all&quot;</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span> <span class="o">=</span> <span class="n">parameter_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span> <span class="o">=</span> <span class="n">database_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span> <span class="o">=</span> <span class="n">target_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span> <span class="o">=</span> <span class="n">decoy_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span> <span class="o">=</span> <span class="n">target_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span> <span class="o">=</span> <span class="n">decoy_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span> <span class="o">=</span> <span class="n">combined_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">output_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">=</span> <span class="n">output_type</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">OUTPUT_TYPES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The variable output_type must be set to either &#39;all&#39; or &#39;optimal&#39;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pdf_filename</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">pdf_filename</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">inference_method_list</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">,</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">,</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">,</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">selected_datastores</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_output_directory</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_log_append_alt_from_db</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method is the main driver of the heuristic method.</span>
<span class="sd">        This method calls other classes and methods that make up the heuristic pipeline.</span>
<span class="sd">        This includes but is not limited to:</span>

<span class="sd">        1. Loops over the main inference methods: Inclusion, Exclusion, Parsimony, and Peptide Centric.</span>
<span class="sd">        2. Determines the optimal inference method based on the input data as well as the database file.</span>
<span class="sd">        3. Outputs the results and indicates the optimal results.</span>

<span class="sd">        Args:</span>
<span class="sd">            fdr_threshold (float): The Qvalue/FDR threshold the heuristic method uses to base calculations from.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; heuristic = pyproteininference.heuristic.HeuristicPipeline(</span>
<span class="sd">            &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">            &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">            &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">            &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">            &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">            &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">            &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">            &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">            &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">            &gt;&gt;&gt;     pdf_filename=pdf_filename,</span>
<span class="sd">            &gt;&gt;&gt;     output_type=&quot;all&quot;</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">            &gt;&gt;&gt; heuristic.execute(fdr_threshold=0.05)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">pyproteininference_parameters</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ProteinInferenceParameter</span><span class="p">(</span>
            <span class="n">yaml_param_filepath</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span>
        <span class="p">)</span>

        <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
            <span class="n">database_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">,</span>
            <span class="n">digest_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">digest_type</span><span class="p">,</span>
            <span class="n">missed_cleavages</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
            <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
            <span class="n">max_peptide_length</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">,</span>
            <span class="n">id_splitting</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running In Silico Database Digest on file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">))</span>
            <span class="n">digest</span><span class="o">.</span><span class="n">digest_fasta_database</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;No Database File provided, Skipping database digest and only taking protein-peptide mapping from the &quot;</span>
                <span class="s2">&quot;input files.&quot;</span>
            <span class="p">)</span>

        <span class="k">for</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">inference_method_list</span><span class="p">:</span>

            <span class="n">method_specific_parameters</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">pyproteininference_parameters</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Overriding inference type </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span><span class="p">))</span>

            <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="n">inference_method</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;New inference type </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span><span class="p">))</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;FDR Threshold Set to </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">fdr</span><span class="p">))</span>

            <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span>
                <span class="n">target_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">target_files</span><span class="p">,</span>
                <span class="n">decoy_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span><span class="p">,</span>
                <span class="n">combined_files</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span>
                <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">,</span>
                <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span>
                <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>

            <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

            <span class="n">data</span><span class="o">.</span><span class="n">restrict_psm_data</span><span class="p">()</span>

            <span class="n">data</span><span class="o">.</span><span class="n">recover_mapping</span><span class="p">()</span>

            <span class="n">data</span><span class="o">.</span><span class="n">create_scoring_input</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">exclude_non_distinguishing_peptides</span><span class="p">()</span>

            <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
            <span class="n">score</span><span class="o">.</span><span class="n">score_psms</span><span class="p">(</span><span class="n">score_method</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">protein_score</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">picker</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">protein_picker</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">pass</span>

            <span class="n">pyproteininference</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">Inference</span><span class="o">.</span><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

            <span class="n">data</span><span class="o">.</span><span class="n">calculate_q_values</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">determine_optimal_inference_method</span><span class="p">(</span>
            <span class="n">false_discovery_rate_threshold</span><span class="o">=</span><span class="n">fdr_threshold</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">selected_datastores</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">}</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_all_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">==</span> <span class="s2">&quot;optimal&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_optimal_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_optimal_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">generate_roc_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method produces a PDF ROC plot overlaying the 4 inference methods apart of the heuristic algorithm.</span>

<span class="sd">        Args:</span>
<span class="sd">            fdr_max (float): Max FDR to display on the plot.</span>
<span class="sd">            pdf_filename (str): Filename to write roc plot to.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">f</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">fdr_vs_target_hits</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span><span class="o">.</span><span class="n">generate_fdr_vs_target_hits</span><span class="p">(</span><span class="n">fdr_max</span><span class="o">=</span><span class="n">fdr_max</span><span class="p">)</span>
            <span class="n">fdrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_target_hits</span><span class="p">]</span>
            <span class="n">target_hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_target_hits</span><span class="p">]</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">fdrs</span><span class="p">,</span> <span class="n">target_hits</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">inference_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
            <span class="n">target_fdr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span>
            <span class="k">if</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">:</span>
                <span class="n">best_value</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">fdrs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">abs</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">target_fdr</span><span class="p">))</span>
                <span class="n">best_index</span> <span class="o">=</span> <span class="n">fdrs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">best_value</span><span class="p">)</span>
                <span class="n">best_target_hit_value</span> <span class="o">=</span> <span class="n">target_hits</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>  <span class="c1"># noqa F841</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">target_fdr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Target FDR&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Decoy FDR&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Target Protein Hits&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">fdr_max</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;FDR vs Target Protein Hits per Inference Method&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pdf_filename</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing ROC plot to: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">))</span>
            <span class="n">f</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_write_all_results</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameters</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method that loops over all results and writes them out.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">datastore</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">:</span>
                <span class="n">inference_method_string</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;optimal_method&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">inference_method_string</span> <span class="o">=</span> <span class="n">method</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">:</span>
                <span class="c1"># If a filename is not provided then construct one using output_directory</span>
                <span class="c1"># Note: output_directory will always get set even if its set as None - gets set to cwd</span>
                <span class="n">inference_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
                    <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">inference_method_string</span><span class="p">,</span>
                        <span class="n">parameters</span><span class="o">.</span><span class="n">tag</span><span class="p">,</span>
                        <span class="n">datastore</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span>
                        <span class="n">datastore</span><span class="o">.</span><span class="n">psm_score</span><span class="p">,</span>
                        <span class="s2">&quot;protein_inference_results.csv&quot;</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
                <span class="c1"># If the user specified an output filename then split it apart and insert the inference method</span>
                <span class="c1"># Then reconstruct the file</span>
                <span class="n">split</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">)</span>
                <span class="n">path</span> <span class="o">=</span> <span class="n">split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">inference_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">inference_method_string</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>
            <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">method</span><span class="p">])</span>
            <span class="n">export</span><span class="o">.</span><span class="n">export_to_csv</span><span class="p">(</span>
                <span class="n">output_filename</span><span class="o">=</span><span class="n">inference_filename</span><span class="p">,</span>
                <span class="n">directory</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
                <span class="n">export_type</span><span class="o">=</span><span class="n">parameters</span><span class="o">.</span><span class="n">export</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_write_optimal_results</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameters</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method that writes out the optimized results.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">:</span>
            <span class="n">datastore</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
            <span class="n">inference_method_string</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;optimal_method&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">:</span>
                <span class="c1"># If a filename is not provided then construct one using output_directory</span>
                <span class="c1"># Note: output_directory will always get set even if its set as None - gets set to cwd</span>
                <span class="n">inference_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
                    <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">inference_method_string</span><span class="p">,</span>
                        <span class="n">parameters</span><span class="o">.</span><span class="n">tag</span><span class="p">,</span>
                        <span class="n">datastore</span><span class="o">.</span><span class="n">short_protein_score</span><span class="p">,</span>
                        <span class="n">datastore</span><span class="o">.</span><span class="n">psm_score</span><span class="p">,</span>
                        <span class="s2">&quot;protein_inference_results.csv&quot;</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
                <span class="c1"># If the user specified an output filename then split it apart and insert the inference method</span>
                <span class="c1"># Then reconstruct the file</span>
                <span class="n">split</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">)</span>
                <span class="n">path</span> <span class="o">=</span> <span class="n">split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">inference_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">inference_method_string</span><span class="p">,</span> <span class="n">filename</span><span class="p">))</span>
            <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">selected_datastores</span><span class="p">[</span><span class="n">method</span><span class="p">])</span>
            <span class="n">export</span><span class="o">.</span><span class="n">export_to_csv</span><span class="p">(</span>
                <span class="n">output_filename</span><span class="o">=</span><span class="n">inference_filename</span><span class="p">,</span>
                <span class="n">directory</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
                <span class="n">export_type</span><span class="o">=</span><span class="n">parameters</span><span class="o">.</span><span class="n">export</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">determine_optimal_inference_method</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">false_discovery_rate_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
        <span class="n">upper_empirical_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">lower_empirical_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
        <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method determines the optimal inference method from Inclusion, Exclusion, Parsimony, Peptide-Centric.</span>

<span class="sd">        Args:</span>
<span class="sd">            false_discovery_rate_threshold (float): The fdr threshold to use in heuristic algorithm -</span>
<span class="sd">                This parameter determines the maximum fdr used when creating a range of finite FDR values.</span>
<span class="sd">            upper_empirical_threshold (float): Upper Threshold used for parsimony/peptide centric cutoff for</span>
<span class="sd">                the heuristic algorithm.</span>
<span class="sd">            lower_empirical_threshold (float): Lower Threshold used for inclusion/exclusion cutoff for</span>
<span class="sd">                the heuristic algorithm.</span>
<span class="sd">            pdf_filename (str): Filename to write heuristic density plot to.</span>


<span class="sd">        Returns:</span>
<span class="sd">            list: List of string representations of the recommended inference methods.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the number of passing proteins</span>
        <span class="n">number_stdev_from_mean_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">fdrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">false_discovery_rate_threshold</span> <span class="o">*</span> <span class="mf">0.01</span> <span class="o">*</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">fdr</span> <span class="ow">in</span> <span class="n">fdrs</span><span class="p">:</span>
            <span class="n">stdev_from_mean</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">determine_number_stdev_from_mean</span><span class="p">(</span><span class="n">false_discovery_rate</span><span class="o">=</span><span class="n">fdr</span><span class="p">)</span>
            <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">]</span> <span class="o">=</span> <span class="n">stdev_from_mean</span>

        <span class="n">stdev_collection</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">fdr</span> <span class="ow">in</span> <span class="n">fdrs</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">]:</span>
                <span class="n">stdev_collection</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">][</span><span class="n">key</span><span class="p">])</span>

        <span class="n">heuristic_scores</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_density_plot</span><span class="p">(</span>
            <span class="n">number_stdevs_from_mean</span><span class="o">=</span><span class="n">stdev_collection</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="n">pdf_filename</span>
        <span class="p">)</span>

        <span class="c1"># Apply conditional statement with lower and upper thresholds</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
            <span class="ow">or</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
        <span class="p">):</span>
            <span class="c1"># If parsimony or peptide centric are less than the lower empirical threshold</span>
            <span class="c1"># Then select the best method of the two</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Either parsimony </span><span class="si">{}</span><span class="s2"> or peptide centric </span><span class="si">{}</span><span class="s2"> pass empirical threshold </span><span class="si">{}</span><span class="s2">. &quot;</span>
                <span class="s2">&quot;Selecting the best method of the two.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">],</span>
                    <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">],</span>
                    <span class="n">lower_empirical_threshold</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">sub_dict</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">],</span>
                <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">],</span>
            <span class="p">}</span>

            <span class="k">if</span> <span class="p">(</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
                <span class="ow">and</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
            <span class="p">):</span>
                <span class="c1"># If both are under the threshold return both</span>
                <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">sub_dict</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sub_dict</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

        <span class="c1"># If the above condition does not apply</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
            <span class="ow">or</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
        <span class="p">):</span>
            <span class="c1"># If exclusion or inclusion are less than the upper empirical threshold</span>
            <span class="c1"># Then select the best method of the two</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Either inclusion </span><span class="si">{}</span><span class="s2"> or exclusion </span><span class="si">{}</span><span class="s2"> pass empirical threshold </span><span class="si">{}</span><span class="s2">. &quot;</span>
                <span class="s2">&quot;Selecting the best method of the two.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">],</span>
                    <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">],</span>
                    <span class="n">upper_empirical_threshold</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">sub_dict</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">],</span>
                <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">],</span>
            <span class="p">}</span>

            <span class="k">if</span> <span class="p">(</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
                <span class="ow">and</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
            <span class="p">):</span>
                <span class="c1"># If both are under the threshold return both</span>
                <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">sub_dict</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sub_dict</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If we have no conditional scenarios...</span>
            <span class="c1"># Select the best method</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No methods pass empirical thresholds, selecting the best method&quot;</span><span class="p">)</span>
            <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">heuristic_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">heuristic_scores</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Method(s) </span><span class="si">{}</span><span class="s2"> selected with the heuristic algorithm&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">selected_methods</span><span class="p">)))</span>
        <span class="k">return</span> <span class="n">selected_methods</span>

    <span class="k">def</span> <span class="nf">generate_density_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_stdevs_from_mean</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method produces a PDF Density Plot plot overlaying the 4 inference methods part of the heuristic algorithm.</span>

<span class="sd">        Args:</span>
<span class="sd">            number_stdevs_from_mean (dict): a dictionary of the number of standard deviations from the mean per</span>
<span class="sd">                inference method for a range of FDRs.</span>
<span class="sd">            pdf_filename (str): Filename to write heuristic density plot to.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a dictionary of heuristic scores per inference method which correlates to the</span>
<span class="sd">                maximum point of the density plot per inference method.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">f</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

        <span class="n">heuristic_scores</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">number_stdevs_from_mean</span><span class="p">:</span>
            <span class="n">readible_method_name</span> <span class="o">=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">INFERENCE_NAME_MAP</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
            <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">histtype</span><span class="o">=</span><span class="s1">&#39;stepfilled&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">ec</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">readible_method_name</span><span class="p">)</span>
            <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">number_stdevs_from_mean</span><span class="p">[</span><span class="n">method</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="n">center</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
            <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">center</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Density Plot of the Number of Standard Deviations from the Mean&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of Standard Deviations from the Mean&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Observations&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pdf_filename</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing Heuristic Density plot to: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">))</span>
            <span class="n">f</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Heuristic Scores&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">heuristic_scores</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">heuristic_scores</span>

    <span class="k">def</span> <span class="nf">determine_number_stdev_from_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method calculates the mean of the number of proteins identified at a specific FDR of all</span>
<span class="sd">        4 methods and then for each method calculates the number of standard deviations</span>
<span class="sd">        from the previous calculated mean.</span>

<span class="sd">        Args:</span>
<span class="sd">            false_discovery_rate (float): The false discovery rate used as a cutoff for calculations.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a dictionary of the number of standard deviations away from the mean per inference method.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">filtered_protein_objects</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span>
                <span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="o">=</span><span class="n">false_discovery_rate</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="n">number_passing_proteins</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_protein_objects</span><span class="p">[</span><span class="n">x</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">filtered_protein_objects</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

        <span class="c1"># Calculate how similar the number of passing proteins is for each method</span>
        <span class="n">all_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">number_passing_proteins</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
        <span class="n">mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">all_values</span><span class="p">)</span>
        <span class="n">standard_deviation</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">stdev</span><span class="p">(</span><span class="n">all_values</span><span class="p">)</span>
        <span class="n">number_stdev_from_mean_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">number_passing_proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">cur_value</span> <span class="o">=</span> <span class="n">number_passing_proteins</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">cur_value</span> <span class="o">-</span> <span class="n">mean</span><span class="p">)</span> <span class="o">/</span> <span class="n">standard_deviation</span>

        <span class="k">return</span> <span class="n">number_stdev_from_mean_dict</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">











  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameter_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>parameter_file</strong> (<code>str</code>)  Path to Protein Inference Yaml Parameter File.</p></li>
            <li><p><strong>database_file</strong> (<code>str</code>)  Path to Fasta database used in proteomics search.</p></li>
            <li><p><strong>target_files</strong> (<code>str/list</code>)  Path to Target Psm File (Or a list of files).</p></li>
            <li><p><strong>decoy_files</strong> (<code>str/list</code>)  Path to Decoy Psm File (Or a list of files).</p></li>
            <li><p><strong>combined_files</strong> (<code>str/list</code>)  Path to Combined Psm File (Or a list of files).</p></li>
            <li><p><strong>target_directory</strong> (<code>str</code>)  Path to Directory containing Target Psm Files.</p></li>
            <li><p><strong>decoy_directory</strong> (<code>str</code>)  Path to Directory containing Decoy Psm Files.</p></li>
            <li><p><strong>combined_directory</strong> (<code>str</code>)  Path to Directory containing Combined Psm Files.</p></li>
            <li><p><strong>output_directory</strong> (<code>str</code>)  Path to Directory where output will be written.</p></li>
            <li><p><strong>output_filename</strong> (<code>str</code>)  Path to Filename where output will be written.
Will override output_directory.</p></li>
            <li><p><strong>id_splitting</strong> (<code>bool</code>)  True/False on whether to split protein IDs in the digest.
Advanced usage only.</p></li>
            <li><p><strong>append_alt_from_db</strong> (<code>bool</code>)  True/False on whether to append alternative proteins
from the DB digestion in Reader class.</p></li>
            <li><p><strong>pdf_filename</strong> (<code>str</code>)  Filepath to be written to by Heuristic Plotting method.
This is optional and a default filename will be created in output_directory if this is left as None</p></li>
            <li><p><strong>output_type</strong> (<code>str</code>)  How to output results. Can either be "all" or "optimal". Will either output all results
        or will only output the optimal results.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>HeuristicPipeline</code>  <a class="autorefs autorefs-internal" href="#pyproteininference.heuristic.HeuristicPipeline">HeuristicPipeline</a> object</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">heuristic</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">heuristic</span><span class="o">.</span><span class="n">HeuristicPipeline</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">parameter_file</span><span class="o">=</span><span class="n">yaml_params</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_file</span><span class="o">=</span><span class="n">database</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_files</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_files</span><span class="o">=</span><span class="n">decoy</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_files</span><span class="o">=</span><span class="n">combined_files</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_directory</span><span class="o">=</span><span class="n">target_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_directory</span><span class="o">=</span><span class="n">decoy_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_directory</span><span class="o">=</span><span class="n">combined_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_directory</span><span class="o">=</span><span class="n">dir_name</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_filename</span><span class="o">=</span><span class="n">output_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">append_alt_from_db</span><span class="o">=</span><span class="n">append_alt</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">pdf_filename</span><span class="o">=</span><span class="n">pdf_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">parameter_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">output_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">        database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">        target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">        decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">        combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">        target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">        decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">        combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">        output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">        output_filename (str): Path to Filename where output will be written.</span>
<span class="sd">            Will override output_directory.</span>
<span class="sd">        id_splitting (bool): True/False on whether to split protein IDs in the digest.</span>
<span class="sd">            Advanced usage only.</span>
<span class="sd">        append_alt_from_db (bool): True/False on whether to append alternative proteins</span>
<span class="sd">            from the DB digestion in Reader class.</span>
<span class="sd">        pdf_filename (str): Filepath to be written to by Heuristic Plotting method.</span>
<span class="sd">            This is optional and a default filename will be created in output_directory if this is left as None</span>
<span class="sd">        output_type (str): How to output results. Can either be &quot;all&quot; or &quot;optimal&quot;. Will either output all results</span>
<span class="sd">                    or will only output the optimal results.</span>

<span class="sd">    Returns:</span>
<span class="sd">        HeuristicPipeline: [HeuristicPipeline][pyproteininference.heuristic.HeuristicPipeline] object</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; heuristic = pyproteininference.heuristic.HeuristicPipeline(</span>
<span class="sd">        &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">        &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">        &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">        &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">        &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">        &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">        &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">        &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">        &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">        &gt;&gt;&gt;     pdf_filename=pdf_filename,</span>
<span class="sd">        &gt;&gt;&gt;     output_type=&quot;all&quot;</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span> <span class="o">=</span> <span class="n">parameter_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span> <span class="o">=</span> <span class="n">database_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span> <span class="o">=</span> <span class="n">target_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span> <span class="o">=</span> <span class="n">decoy_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span> <span class="o">=</span> <span class="n">target_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span> <span class="o">=</span> <span class="n">decoy_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span> <span class="o">=</span> <span class="n">combined_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">output_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="o">=</span> <span class="n">output_filename</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">=</span> <span class="n">output_type</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">OUTPUT_TYPES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The variable output_type must be set to either &#39;all&#39; or &#39;optimal&#39;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">pdf_filename</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span> <span class="s2">&quot;heuristic_plot.pdf&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span> <span class="o">=</span> <span class="n">pdf_filename</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">inference_method_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">,</span>
        <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">,</span>
        <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">,</span>
        <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">selected_datastores</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_set_output_directory</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_log_append_alt_from_db</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.determine_number_stdev_from_mean">
<code class="highlight language-python"><span class="n">determine_number_stdev_from_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method calculates the mean of the number of proteins identified at a specific FDR of all
4 methods and then for each method calculates the number of standard deviations
from the previous calculated mean.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>false_discovery_rate</strong> (<code>float</code>)  The false discovery rate used as a cutoff for calculations.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>dict</code>  a dictionary of the number of standard deviations away from the mean per inference method.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">determine_number_stdev_from_mean</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method calculates the mean of the number of proteins identified at a specific FDR of all</span>
<span class="sd">    4 methods and then for each method calculates the number of standard deviations</span>
<span class="sd">    from the previous calculated mean.</span>

<span class="sd">    Args:</span>
<span class="sd">        false_discovery_rate (float): The false discovery rate used as a cutoff for calculations.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: a dictionary of the number of standard deviations away from the mean per inference method.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">filtered_protein_objects</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">get_protein_objects</span><span class="p">(</span>
            <span class="n">fdr_restricted</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">false_discovery_rate</span><span class="o">=</span><span class="n">false_discovery_rate</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="n">number_passing_proteins</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_protein_objects</span><span class="p">[</span><span class="n">x</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">filtered_protein_objects</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="c1"># Calculate how similar the number of passing proteins is for each method</span>
    <span class="n">all_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">number_passing_proteins</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">all_values</span><span class="p">)</span>
    <span class="n">standard_deviation</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">stdev</span><span class="p">(</span><span class="n">all_values</span><span class="p">)</span>
    <span class="n">number_stdev_from_mean_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">number_passing_proteins</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">cur_value</span> <span class="o">=</span> <span class="n">number_passing_proteins</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
        <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">cur_value</span> <span class="o">-</span> <span class="n">mean</span><span class="p">)</span> <span class="o">/</span> <span class="n">standard_deviation</span>

    <span class="k">return</span> <span class="n">number_stdev_from_mean_dict</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.determine_optimal_inference_method">
<code class="highlight language-python"><span class="n">determine_optimal_inference_method</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">false_discovery_rate_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">upper_empirical_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">lower_empirical_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method determines the optimal inference method from Inclusion, Exclusion, Parsimony, Peptide-Centric.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>false_discovery_rate_threshold</strong> (<code>float</code>)  The fdr threshold to use in heuristic algorithm -
This parameter determines the maximum fdr used when creating a range of finite FDR values.</p></li>
            <li><p><strong>upper_empirical_threshold</strong> (<code>float</code>)  Upper Threshold used for parsimony/peptide centric cutoff for
the heuristic algorithm.</p></li>
            <li><p><strong>lower_empirical_threshold</strong> (<code>float</code>)  Lower Threshold used for inclusion/exclusion cutoff for
the heuristic algorithm.</p></li>
            <li><p><strong>pdf_filename</strong> (<code>str</code>)  Filename to write heuristic density plot to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of string representations of the recommended inference methods.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">determine_optimal_inference_method</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">false_discovery_rate_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
    <span class="n">upper_empirical_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">lower_empirical_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method determines the optimal inference method from Inclusion, Exclusion, Parsimony, Peptide-Centric.</span>

<span class="sd">    Args:</span>
<span class="sd">        false_discovery_rate_threshold (float): The fdr threshold to use in heuristic algorithm -</span>
<span class="sd">            This parameter determines the maximum fdr used when creating a range of finite FDR values.</span>
<span class="sd">        upper_empirical_threshold (float): Upper Threshold used for parsimony/peptide centric cutoff for</span>
<span class="sd">            the heuristic algorithm.</span>
<span class="sd">        lower_empirical_threshold (float): Lower Threshold used for inclusion/exclusion cutoff for</span>
<span class="sd">            the heuristic algorithm.</span>
<span class="sd">        pdf_filename (str): Filename to write heuristic density plot to.</span>


<span class="sd">    Returns:</span>
<span class="sd">        list: List of string representations of the recommended inference methods.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Get the number of passing proteins</span>
    <span class="n">number_stdev_from_mean_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">fdrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">false_discovery_rate_threshold</span> <span class="o">*</span> <span class="mf">0.01</span> <span class="o">*</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">)]</span>
    <span class="k">for</span> <span class="n">fdr</span> <span class="ow">in</span> <span class="n">fdrs</span><span class="p">:</span>
        <span class="n">stdev_from_mean</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">determine_number_stdev_from_mean</span><span class="p">(</span><span class="n">false_discovery_rate</span><span class="o">=</span><span class="n">fdr</span><span class="p">)</span>
        <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">]</span> <span class="o">=</span> <span class="n">stdev_from_mean</span>

    <span class="n">stdev_collection</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">fdr</span> <span class="ow">in</span> <span class="n">fdrs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">]:</span>
            <span class="n">stdev_collection</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">number_stdev_from_mean_dict</span><span class="p">[</span><span class="n">fdr</span><span class="p">][</span><span class="n">key</span><span class="p">])</span>

    <span class="n">heuristic_scores</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_density_plot</span><span class="p">(</span>
        <span class="n">number_stdevs_from_mean</span><span class="o">=</span><span class="n">stdev_collection</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="n">pdf_filename</span>
    <span class="p">)</span>

    <span class="c1"># Apply conditional statement with lower and upper thresholds</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
        <span class="ow">or</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
    <span class="p">):</span>
        <span class="c1"># If parsimony or peptide centric are less than the lower empirical threshold</span>
        <span class="c1"># Then select the best method of the two</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Either parsimony </span><span class="si">{}</span><span class="s2"> or peptide centric </span><span class="si">{}</span><span class="s2"> pass empirical threshold </span><span class="si">{}</span><span class="s2">. &quot;</span>
            <span class="s2">&quot;Selecting the best method of the two.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">],</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">],</span>
                <span class="n">lower_empirical_threshold</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">sub_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">],</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">],</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
            <span class="ow">and</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">lower_empirical_threshold</span>
        <span class="p">):</span>
            <span class="c1"># If both are under the threshold return both</span>
            <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">sub_dict</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sub_dict</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

    <span class="c1"># If the above condition does not apply</span>
    <span class="k">elif</span> <span class="p">(</span>
        <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
        <span class="ow">or</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
    <span class="p">):</span>
        <span class="c1"># If exclusion or inclusion are less than the upper empirical threshold</span>
        <span class="c1"># Then select the best method of the two</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Either inclusion </span><span class="si">{}</span><span class="s2"> or exclusion </span><span class="si">{}</span><span class="s2"> pass empirical threshold </span><span class="si">{}</span><span class="s2">. &quot;</span>
            <span class="s2">&quot;Selecting the best method of the two.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">],</span>
                <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">],</span>
                <span class="n">upper_empirical_threshold</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">sub_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">],</span>
            <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">:</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">],</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
            <span class="ow">and</span> <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_empirical_threshold</span>
        <span class="p">):</span>
            <span class="c1"># If both are under the threshold return both</span>
            <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">sub_dict</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">sub_dict</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># If we have no conditional scenarios...</span>
        <span class="c1"># Select the best method</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No methods pass empirical thresholds, selecting the best method&quot;</span><span class="p">)</span>
        <span class="n">selected_methods</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">heuristic_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">heuristic_scores</span><span class="o">.</span><span class="n">get</span><span class="p">)]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Method(s) </span><span class="si">{}</span><span class="s2"> selected with the heuristic algorithm&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">selected_methods</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">selected_methods</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.execute">
<code class="highlight language-python"><span class="n">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method is the main driver of the heuristic method.
This method calls other classes and methods that make up the heuristic pipeline.
This includes but is not limited to:</p>
<ol>
<li>Loops over the main inference methods: Inclusion, Exclusion, Parsimony, and Peptide Centric.</li>
<li>Determines the optimal inference method based on the input data as well as the database file.</li>
<li>Outputs the results and indicates the optimal results.</li>
</ol>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>fdr_threshold</strong> (<code>float</code>)  The Qvalue/FDR threshold the heuristic method uses to base calculations from.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">heuristic</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">heuristic</span><span class="o">.</span><span class="n">HeuristicPipeline</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">parameter_file</span><span class="o">=</span><span class="n">yaml_params</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_file</span><span class="o">=</span><span class="n">database</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_files</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_files</span><span class="o">=</span><span class="n">decoy</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_files</span><span class="o">=</span><span class="n">combined_files</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_directory</span><span class="o">=</span><span class="n">target_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_directory</span><span class="o">=</span><span class="n">decoy_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_directory</span><span class="o">=</span><span class="n">combined_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_directory</span><span class="o">=</span><span class="n">dir_name</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_filename</span><span class="o">=</span><span class="n">output_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">append_alt_from_db</span><span class="o">=</span><span class="n">append_alt</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">pdf_filename</span><span class="o">=</span><span class="n">pdf_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">heuristic</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method is the main driver of the heuristic method.</span>
<span class="sd">    This method calls other classes and methods that make up the heuristic pipeline.</span>
<span class="sd">    This includes but is not limited to:</span>

<span class="sd">    1. Loops over the main inference methods: Inclusion, Exclusion, Parsimony, and Peptide Centric.</span>
<span class="sd">    2. Determines the optimal inference method based on the input data as well as the database file.</span>
<span class="sd">    3. Outputs the results and indicates the optimal results.</span>

<span class="sd">    Args:</span>
<span class="sd">        fdr_threshold (float): The Qvalue/FDR threshold the heuristic method uses to base calculations from.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; heuristic = pyproteininference.heuristic.HeuristicPipeline(</span>
<span class="sd">        &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">        &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">        &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">        &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">        &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">        &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">        &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">        &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">        &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">        &gt;&gt;&gt;     pdf_filename=pdf_filename,</span>
<span class="sd">        &gt;&gt;&gt;     output_type=&quot;all&quot;</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">        &gt;&gt;&gt; heuristic.execute(fdr_threshold=0.05)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">pyproteininference_parameters</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ProteinInferenceParameter</span><span class="p">(</span>
        <span class="n">yaml_param_filepath</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span>
    <span class="p">)</span>

    <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
        <span class="n">database_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">,</span>
        <span class="n">digest_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">digest_type</span><span class="p">,</span>
        <span class="n">missed_cleavages</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
        <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
        <span class="n">max_peptide_length</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">,</span>
        <span class="n">id_splitting</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running In Silico Database Digest on file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">))</span>
        <span class="n">digest</span><span class="o">.</span><span class="n">digest_fasta_database</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;No Database File provided, Skipping database digest and only taking protein-peptide mapping from the &quot;</span>
            <span class="s2">&quot;input files.&quot;</span>
        <span class="p">)</span>

    <span class="k">for</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">inference_method_list</span><span class="p">:</span>

        <span class="n">method_specific_parameters</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">pyproteininference_parameters</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Overriding inference type </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span><span class="p">))</span>

        <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="n">inference_method</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;New inference type </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;FDR Threshold Set to </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">fdr</span><span class="p">))</span>

        <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span>
            <span class="n">target_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">target_files</span><span class="p">,</span>
            <span class="n">decoy_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span><span class="p">,</span>
            <span class="n">combined_files</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span>
            <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">,</span>
            <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span>
            <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

        <span class="n">data</span><span class="o">.</span><span class="n">restrict_psm_data</span><span class="p">()</span>

        <span class="n">data</span><span class="o">.</span><span class="n">recover_mapping</span><span class="p">()</span>

        <span class="n">data</span><span class="o">.</span><span class="n">create_scoring_input</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
            <span class="n">data</span><span class="o">.</span><span class="n">exclude_non_distinguishing_peptides</span><span class="p">()</span>

        <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        <span class="n">score</span><span class="o">.</span><span class="n">score_psms</span><span class="p">(</span><span class="n">score_method</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">protein_score</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">method_specific_parameters</span><span class="o">.</span><span class="n">picker</span><span class="p">:</span>
            <span class="n">data</span><span class="o">.</span><span class="n">protein_picker</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="n">pyproteininference</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">Inference</span><span class="o">.</span><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

        <span class="n">data</span><span class="o">.</span><span class="n">calculate_q_values</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">determine_optimal_inference_method</span><span class="p">(</span>
        <span class="n">false_discovery_rate_threshold</span><span class="o">=</span><span class="n">fdr_threshold</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pdf_filename</span>
    <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">selected_datastores</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">}</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_all_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>
    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_type</span> <span class="o">==</span> <span class="s2">&quot;optimal&quot;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_optimal_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_optimal_results</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="n">method_specific_parameters</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.generate_density_plot">
<code class="highlight language-python"><span class="n">generate_density_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_stdevs_from_mean</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method produces a PDF Density Plot plot overlaying the 4 inference methods part of the heuristic algorithm.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>number_stdevs_from_mean</strong> (<code>dict</code>)  a dictionary of the number of standard deviations from the mean per
inference method for a range of FDRs.</p></li>
            <li><p><strong>pdf_filename</strong> (<code>str</code>)  Filename to write heuristic density plot to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>dict</code>  a dictionary of heuristic scores per inference method which correlates to the
    maximum point of the density plot per inference method.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">generate_density_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_stdevs_from_mean</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method produces a PDF Density Plot plot overlaying the 4 inference methods part of the heuristic algorithm.</span>

<span class="sd">    Args:</span>
<span class="sd">        number_stdevs_from_mean (dict): a dictionary of the number of standard deviations from the mean per</span>
<span class="sd">            inference method for a range of FDRs.</span>
<span class="sd">        pdf_filename (str): Filename to write heuristic density plot to.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: a dictionary of heuristic scores per inference method which correlates to the</span>
<span class="sd">            maximum point of the density plot per inference method.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">f</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

    <span class="n">heuristic_scores</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">number_stdevs_from_mean</span><span class="p">:</span>
        <span class="n">readible_method_name</span> <span class="o">=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">INFERENCE_NAME_MAP</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">histtype</span><span class="o">=</span><span class="s1">&#39;stepfilled&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">ec</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">readible_method_name</span><span class="p">)</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">number_stdevs_from_mean</span><span class="p">[</span><span class="n">method</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">center</span> <span class="o">=</span> <span class="n">y</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
        <span class="n">heuristic_scores</span><span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">center</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Density Plot of the Number of Standard Deviations from the Mean&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of Standard Deviations from the Mean&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Observations&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">pdf_filename</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing Heuristic Density plot to: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">))</span>
        <span class="n">f</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Heuristic Scores&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">heuristic_scores</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">heuristic_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.heuristic.HeuristicPipeline.generate_roc_plot">
<code class="highlight language-python"><span class="n">generate_roc_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method produces a PDF ROC plot overlaying the 4 inference methods apart of the heuristic algorithm.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>fdr_max</strong> (<code>float</code>)  Max FDR to display on the plot.</p></li>
            <li><p><strong>pdf_filename</strong> (<code>str</code>)  Filename to write roc plot to.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/heuristic.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">generate_roc_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fdr_max</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">pdf_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method produces a PDF ROC plot overlaying the 4 inference methods apart of the heuristic algorithm.</span>

<span class="sd">    Args:</span>
<span class="sd">        fdr_max (float): Max FDR to display on the plot.</span>
<span class="sd">        pdf_filename (str): Filename to write roc plot to.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">f</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">fdr_vs_target_hits</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span><span class="o">.</span><span class="n">generate_fdr_vs_target_hits</span><span class="p">(</span><span class="n">fdr_max</span><span class="o">=</span><span class="n">fdr_max</span><span class="p">)</span>
        <span class="n">fdrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_target_hits</span><span class="p">]</span>
        <span class="n">target_hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fdr_vs_target_hits</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">fdrs</span><span class="p">,</span> <span class="n">target_hits</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">inference_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">))</span>
        <span class="n">target_fdr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datastore_dict</span><span class="p">[</span><span class="n">inference_method</span><span class="p">]</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">fdr</span>
        <span class="k">if</span> <span class="n">inference_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">selected_methods</span><span class="p">:</span>
            <span class="n">best_value</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">fdrs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">abs</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">target_fdr</span><span class="p">))</span>
            <span class="n">best_index</span> <span class="o">=</span> <span class="n">fdrs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">best_value</span><span class="p">)</span>
            <span class="n">best_target_hit_value</span> <span class="o">=</span> <span class="n">target_hits</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>  <span class="c1"># noqa F841</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">target_fdr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Target FDR&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Decoy FDR&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Target Protein Hits&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">fdr_max</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;FDR vs Target Protein Hits per Inference Method&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">pdf_filename</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing ROC plot to: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">))</span>
        <span class="n">f</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">pdf_filename</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.in_silico_digest">
        <code>in_silico_digest</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.in_silico_digest.Digest">
        <code>
Digest        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class handles data storage of in silico digest data from a fasta formatted sequence database.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>peptide_to_protein_dictionary</code></td>
        <td><code>dict</code></td>
        <td><p>Dictionary of peptides (keys) to protein sets (values).</p></td>
      </tr>
      <tr>
        <td><code>protein_to_peptide_dictionary</code></td>
        <td><code>dict</code></td>
        <td><p>Dictionary of proteins (keys) to peptide sets (values).</p></td>
      </tr>
      <tr>
        <td><code>swiss_prot_protein_set</code></td>
        <td><code>set</code></td>
        <td><p>Set of reviewed proteins if they are able to be distinguished from unreviewed
proteins.</p></td>
      </tr>
      <tr>
        <td><code>database_path</code></td>
        <td><code>str</code></td>
        <td><p>Path to fasta database file to digest.</p></td>
      </tr>
      <tr>
        <td><code>missed_cleavages</code></td>
        <td><code>int</code></td>
        <td><p>The number of missed cleavages to allow.</p></td>
      </tr>
      <tr>
        <td><code>id_splitting</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether or not to split a given regex off identifiers.
This is used to split of "sp|" and "tr|"
from the database protein strings as sometimes the database will contain those strings while
the input data will have the strings split already.
Advanced usage only.</p></td>
      </tr>
      <tr>
        <td><code>reviewed_identifier_symbol</code></td>
        <td><code>str/None</code></td>
        <td><p>Identifier that distinguishes reviewed from unreviewed proteins.
Typically this is "sp|". Can also be None type.</p></td>
      </tr>
      <tr>
        <td><code>digest_type</code></td>
        <td><code>str</code></td>
        <td><p>can be any value in <code>LIST_OF_DIGEST_TYPES</code>.</p></td>
      </tr>
      <tr>
        <td><code>max_peptide_length</code></td>
        <td><code>int</code></td>
        <td><p>Max peptide length to keep for analysis.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/in_silico_digest.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Digest</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class handles data storage of in silico digest data from a fasta formatted sequence database.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        peptide_to_protein_dictionary (dict): Dictionary of peptides (keys) to protein sets (values).</span>
<span class="sd">        protein_to_peptide_dictionary (dict): Dictionary of proteins (keys) to peptide sets (values).</span>
<span class="sd">        swiss_prot_protein_set (set): Set of reviewed proteins if they are able to be distinguished from unreviewed</span>
<span class="sd">            proteins.</span>
<span class="sd">        database_path (str): Path to fasta database file to digest.</span>
<span class="sd">        missed_cleavages (int): The number of missed cleavages to allow.</span>
<span class="sd">        id_splitting (bool): True/False on whether or not to split a given regex off identifiers.</span>
<span class="sd">            This is used to split of &quot;sp|&quot; and &quot;tr|&quot;</span>
<span class="sd">            from the database protein strings as sometimes the database will contain those strings while</span>
<span class="sd">            the input data will have the strings split already.</span>
<span class="sd">            Advanced usage only.</span>
<span class="sd">        reviewed_identifier_symbol (str/None): Identifier that distinguishes reviewed from unreviewed proteins.</span>
<span class="sd">            Typically this is &quot;sp|&quot;. Can also be None type.</span>
<span class="sd">        digest_type (str): can be any value in `LIST_OF_DIGEST_TYPES`.</span>
<span class="sd">        max_peptide_length (int): Max peptide length to keep for analysis.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">TRYPSIN</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
    <span class="n">LYSC</span> <span class="o">=</span> <span class="s2">&quot;lysc&quot;</span>
    <span class="n">LIST_OF_DIGEST_TYPES</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">parser</span><span class="o">.</span><span class="n">expasy_rules</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="n">AA_LIST</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;N&quot;</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">,</span>
        <span class="s2">&quot;E&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
        <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;I&quot;</span><span class="p">,</span>
        <span class="s2">&quot;L&quot;</span><span class="p">,</span>
        <span class="s2">&quot;K&quot;</span><span class="p">,</span>
        <span class="s2">&quot;M&quot;</span><span class="p">,</span>
        <span class="s2">&quot;F&quot;</span><span class="p">,</span>
        <span class="s2">&quot;P&quot;</span><span class="p">,</span>
        <span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="s2">&quot;T&quot;</span><span class="p">,</span>
        <span class="s2">&quot;W&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
        <span class="s2">&quot;V&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">UNIPROT_STRS</span> <span class="o">=</span> <span class="s2">&quot;sp\||tr\|&quot;</span>  <span class="c1"># noqa W605</span>
    <span class="n">UNIPROT_STR_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">UNIPROT_STRS</span><span class="p">)</span>
    <span class="n">SP_STRING</span> <span class="o">=</span> <span class="s2">&quot;sp|&quot;</span>
    <span class="n">METHIONINE</span> <span class="o">=</span> <span class="s2">&quot;M&quot;</span>
    <span class="n">ANY_AMINO_ACID</span> <span class="o">=</span> <span class="s2">&quot;X&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">





















  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.in_silico_digest.PyteomicsDigest">
        <code>
PyteomicsDigest            (<a class="autorefs autorefs-internal" title="pyproteininference.in_silico_digest.Digest" href="#pyproteininference.in_silico_digest.Digest">Digest</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>This class represents a pyteomics implementation of an in silico digest.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/in_silico_digest.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">PyteomicsDigest</span><span class="p">(</span><span class="n">Digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class represents a pyteomics implementation of an in silico digest.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">database_path</span><span class="p">,</span>
        <span class="n">digest_type</span><span class="p">,</span>
        <span class="n">missed_cleavages</span><span class="p">,</span>
        <span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
        <span class="n">max_peptide_length</span><span class="p">,</span>
        <span class="n">id_splitting</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The following class creates protein to peptide, peptide to protein, and reviewed protein mappings.</span>

<span class="sd">        The input is a fasta database, a protein inference parameter object, and whether or not to split IDs.</span>

<span class="sd">        This class sets important attributes for the Digest object such as: `peptide_to_protein_dictionary`,</span>
<span class="sd">        `protein_to_peptide_dictionary`, and `swiss_prot_protein_set`.</span>

<span class="sd">        Args:</span>
<span class="sd">            database_path (str): Path to fasta database file to digest.</span>
<span class="sd">            digest_type (str): Must be a value in `LIST_OF_DIGEST_TYPES`.</span>
<span class="sd">            missed_cleavages (int): Integer that indicates the maximum number of allowable missed cleavages from</span>
<span class="sd">                the ms search.</span>
<span class="sd">            reviewed_identifier_symbol (str/None): Symbol that indicates a reviewed identifier.</span>
<span class="sd">                If using Uniprot this is typically &#39;sp|&#39;.</span>
<span class="sd">            max_peptide_length (int): The maximum length of peptides to keep for the analysis.</span>
<span class="sd">            id_splitting (bool): True/False on whether or not to split a given regex off identifiers.</span>
<span class="sd">                This is used to split of &quot;sp|&quot; and &quot;tr|&quot;</span>
<span class="sd">                from the database protein strings as sometimes the database will contain those</span>
<span class="sd">                strings while the input data will have the strings split already.</span>
<span class="sd">                Advanced usage only.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; digest = pyproteininference.in_silico_digest.PyteomicsDigest(</span>
<span class="sd">            &gt;&gt;&gt;     database_path=database_file,</span>
<span class="sd">            &gt;&gt;&gt;     digest_type=&#39;trypsin&#39;,</span>
<span class="sd">            &gt;&gt;&gt;     missed_cleavages=2,</span>
<span class="sd">            &gt;&gt;&gt;     reviewed_identifier_symbol=&#39;sp|&#39;,</span>
<span class="sd">            &gt;&gt;&gt;     max_peptide_length=7,</span>
<span class="sd">            &gt;&gt;&gt;     id_splitting=False,</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">database_path</span> <span class="o">=</span> <span class="n">database_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="n">missed_cleavages</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="n">reviewed_identifier_symbol</span>
        <span class="k">if</span> <span class="n">digest_type</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="n">digest_type</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;digest_type must be equal to one of the following </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_peptide_length</span> <span class="o">=</span> <span class="n">max_peptide_length</span>

    <span class="k">def</span> <span class="nf">digest_fasta_database</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method reads in and prepares the fasta database for database digestion and assigns</span>
<span class="sd">        the several attributes for the Digest object: `peptide_to_protein_dictionary`,</span>
<span class="sd">        `protein_to_peptide_dictionary`, and `swiss_prot_protein_set`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; digest = pyproteininference.in_silico_digest.PyteomicsDigest(</span>
<span class="sd">            &gt;&gt;&gt;     database_path=database_file,</span>
<span class="sd">            &gt;&gt;&gt;     digest_type=&#39;trypsin&#39;,</span>
<span class="sd">            &gt;&gt;&gt;     missed_cleavages=2,</span>
<span class="sd">            &gt;&gt;&gt;     reviewed_identifier_symbol=&#39;sp|&#39;,</span>
<span class="sd">            &gt;&gt;&gt;     max_peptide_length=7,</span>
<span class="sd">            &gt;&gt;&gt;     id_splitting=False,</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">            &gt;&gt;&gt; digest.digest_fasta_database()</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting Pyteomics Digest...&quot;</span><span class="p">)</span>
        <span class="n">pep_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">prot_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">sp_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">description</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_path</span><span class="p">):</span>
            <span class="n">new_peptides</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">cleave</span><span class="p">(</span>
                <span class="n">sequence</span><span class="p">,</span>
                <span class="n">parser</span><span class="o">.</span><span class="n">expasy_rules</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
                <span class="n">min_length</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">max_peptide_length</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Hopefully this splitting works...</span>
            <span class="c1"># IDK how robust this is...</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

            <span class="c1"># Handle ID Splitting...</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">:</span>
                <span class="n">identifier_stripped</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">UNIPROT_STR_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">identifier</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">identifier_stripped</span> <span class="o">=</span> <span class="n">identifier</span>

            <span class="c1"># If reviewed add to sp_set</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">):</span>
                    <span class="n">sp_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>

            <span class="n">prot_dict</span><span class="p">[</span><span class="n">identifier_stripped</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_peptides</span>
            <span class="n">met_cleaved_peps</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">peptide</span> <span class="ow">in</span> <span class="n">new_peptides</span><span class="p">:</span>
                <span class="n">pep_dict</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>
                <span class="c1"># Need to account for potential N-term Methionine Cleavage</span>
                <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">peptide</span><span class="p">)</span> <span class="ow">and</span> <span class="n">peptide</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">METHIONINE</span><span class="p">):</span>
                    <span class="c1"># If our sequence starts with the current peptide... and our current peptide starts with methionine</span>
                    <span class="c1"># Then we remove the methionine from the peptide and add it to our dicts...</span>
                    <span class="n">methionine_cleaved_peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="n">met_cleaved_peps</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">methionine_cleaved_peptide</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">met_peps</span> <span class="ow">in</span> <span class="n">met_cleaved_peps</span><span class="p">:</span>
                <span class="n">pep_dict</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">met_peps</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>
                <span class="n">prot_dict</span><span class="p">[</span><span class="n">identifier_stripped</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">met_peps</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span> <span class="o">=</span> <span class="n">sp_set</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="n">pep_dict</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span> <span class="o">=</span> <span class="n">prot_dict</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Pyteomics Digest Finished...&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.in_silico_digest.PyteomicsDigest.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">database_path</span><span class="p">,</span> <span class="n">digest_type</span><span class="p">,</span> <span class="n">missed_cleavages</span><span class="p">,</span> <span class="n">reviewed_identifier_symbol</span><span class="p">,</span> <span class="n">max_peptide_length</span><span class="p">,</span> <span class="n">id_splitting</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>The following class creates protein to peptide, peptide to protein, and reviewed protein mappings.</p>
<p>The input is a fasta database, a protein inference parameter object, and whether or not to split IDs.</p>
<p>This class sets important attributes for the Digest object such as: <code>peptide_to_protein_dictionary</code>,
<code>protein_to_peptide_dictionary</code>, and <code>swiss_prot_protein_set</code>.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>database_path</strong> (<code>str</code>)  Path to fasta database file to digest.</p></li>
            <li><p><strong>digest_type</strong> (<code>str</code>)  Must be a value in <code>LIST_OF_DIGEST_TYPES</code>.</p></li>
            <li><p><strong>missed_cleavages</strong> (<code>int</code>)  Integer that indicates the maximum number of allowable missed cleavages from
the ms search.</p></li>
            <li><p><strong>reviewed_identifier_symbol</strong> (<code>str/None</code>)  Symbol that indicates a reviewed identifier.
If using Uniprot this is typically 'sp|'.</p></li>
            <li><p><strong>max_peptide_length</strong> (<code>int</code>)  The maximum length of peptides to keep for the analysis.</p></li>
            <li><p><strong>id_splitting</strong> (<code>bool</code>)  True/False on whether or not to split a given regex off identifiers.
This is used to split of "sp|" and "tr|"
from the database protein strings as sometimes the database will contain those
strings while the input data will have the strings split already.
Advanced usage only.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_path</span><span class="o">=</span><span class="n">database_file</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">digest_type</span><span class="o">=</span><span class="s1">&#39;trypsin&#39;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">missed_cleavages</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="s1">&#39;sp|&#39;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">max_peptide_length</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/in_silico_digest.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">database_path</span><span class="p">,</span>
    <span class="n">digest_type</span><span class="p">,</span>
    <span class="n">missed_cleavages</span><span class="p">,</span>
    <span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
    <span class="n">max_peptide_length</span><span class="p">,</span>
    <span class="n">id_splitting</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class creates protein to peptide, peptide to protein, and reviewed protein mappings.</span>

<span class="sd">    The input is a fasta database, a protein inference parameter object, and whether or not to split IDs.</span>

<span class="sd">    This class sets important attributes for the Digest object such as: `peptide_to_protein_dictionary`,</span>
<span class="sd">    `protein_to_peptide_dictionary`, and `swiss_prot_protein_set`.</span>

<span class="sd">    Args:</span>
<span class="sd">        database_path (str): Path to fasta database file to digest.</span>
<span class="sd">        digest_type (str): Must be a value in `LIST_OF_DIGEST_TYPES`.</span>
<span class="sd">        missed_cleavages (int): Integer that indicates the maximum number of allowable missed cleavages from</span>
<span class="sd">            the ms search.</span>
<span class="sd">        reviewed_identifier_symbol (str/None): Symbol that indicates a reviewed identifier.</span>
<span class="sd">            If using Uniprot this is typically &#39;sp|&#39;.</span>
<span class="sd">        max_peptide_length (int): The maximum length of peptides to keep for the analysis.</span>
<span class="sd">        id_splitting (bool): True/False on whether or not to split a given regex off identifiers.</span>
<span class="sd">            This is used to split of &quot;sp|&quot; and &quot;tr|&quot;</span>
<span class="sd">            from the database protein strings as sometimes the database will contain those</span>
<span class="sd">            strings while the input data will have the strings split already.</span>
<span class="sd">            Advanced usage only.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; digest = pyproteininference.in_silico_digest.PyteomicsDigest(</span>
<span class="sd">        &gt;&gt;&gt;     database_path=database_file,</span>
<span class="sd">        &gt;&gt;&gt;     digest_type=&#39;trypsin&#39;,</span>
<span class="sd">        &gt;&gt;&gt;     missed_cleavages=2,</span>
<span class="sd">        &gt;&gt;&gt;     reviewed_identifier_symbol=&#39;sp|&#39;,</span>
<span class="sd">        &gt;&gt;&gt;     max_peptide_length=7,</span>
<span class="sd">        &gt;&gt;&gt;     id_splitting=False,</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">database_path</span> <span class="o">=</span> <span class="n">database_path</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="n">missed_cleavages</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="n">reviewed_identifier_symbol</span>
    <span class="k">if</span> <span class="n">digest_type</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="n">digest_type</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;digest_type must be equal to one of the following </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_peptide_length</span> <span class="o">=</span> <span class="n">max_peptide_length</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.in_silico_digest.PyteomicsDigest.digest_fasta_database">
<code class="highlight language-python"><span class="n">digest_fasta_database</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method reads in and prepares the fasta database for database digestion and assigns
the several attributes for the Digest object: <code>peptide_to_protein_dictionary</code>,
<code>protein_to_peptide_dictionary</code>, and <code>swiss_prot_protein_set</code>.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_path</span><span class="o">=</span><span class="n">database_file</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">digest_type</span><span class="o">=</span><span class="s1">&#39;trypsin&#39;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">missed_cleavages</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="s1">&#39;sp|&#39;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">max_peptide_length</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">digest</span><span class="o">.</span><span class="n">digest_fasta_database</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/in_silico_digest.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">digest_fasta_database</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method reads in and prepares the fasta database for database digestion and assigns</span>
<span class="sd">    the several attributes for the Digest object: `peptide_to_protein_dictionary`,</span>
<span class="sd">    `protein_to_peptide_dictionary`, and `swiss_prot_protein_set`.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; digest = pyproteininference.in_silico_digest.PyteomicsDigest(</span>
<span class="sd">        &gt;&gt;&gt;     database_path=database_file,</span>
<span class="sd">        &gt;&gt;&gt;     digest_type=&#39;trypsin&#39;,</span>
<span class="sd">        &gt;&gt;&gt;     missed_cleavages=2,</span>
<span class="sd">        &gt;&gt;&gt;     reviewed_identifier_symbol=&#39;sp|&#39;,</span>
<span class="sd">        &gt;&gt;&gt;     max_peptide_length=7,</span>
<span class="sd">        &gt;&gt;&gt;     id_splitting=False,</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">        &gt;&gt;&gt; digest.digest_fasta_database()</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting Pyteomics Digest...&quot;</span><span class="p">)</span>
    <span class="n">pep_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">prot_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">sp_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">description</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">fasta</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_path</span><span class="p">):</span>
        <span class="n">new_peptides</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">cleave</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">,</span>
            <span class="n">parser</span><span class="o">.</span><span class="n">expasy_rules</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
            <span class="n">min_length</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">max_peptide_length</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Hopefully this splitting works...</span>
        <span class="c1"># IDK how robust this is...</span>
        <span class="n">identifier</span> <span class="o">=</span> <span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Handle ID Splitting...</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">:</span>
            <span class="n">identifier_stripped</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">UNIPROT_STR_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">identifier</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">identifier_stripped</span> <span class="o">=</span> <span class="n">identifier</span>

        <span class="c1"># If reviewed add to sp_set</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">):</span>
                <span class="n">sp_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>

        <span class="n">prot_dict</span><span class="p">[</span><span class="n">identifier_stripped</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_peptides</span>
        <span class="n">met_cleaved_peps</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">peptide</span> <span class="ow">in</span> <span class="n">new_peptides</span><span class="p">:</span>
            <span class="n">pep_dict</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>
            <span class="c1"># Need to account for potential N-term Methionine Cleavage</span>
            <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">peptide</span><span class="p">)</span> <span class="ow">and</span> <span class="n">peptide</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">METHIONINE</span><span class="p">):</span>
                <span class="c1"># If our sequence starts with the current peptide... and our current peptide starts with methionine</span>
                <span class="c1"># Then we remove the methionine from the peptide and add it to our dicts...</span>
                <span class="n">methionine_cleaved_peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
                <span class="n">met_cleaved_peps</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">methionine_cleaved_peptide</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">met_peps</span> <span class="ow">in</span> <span class="n">met_cleaved_peps</span><span class="p">:</span>
            <span class="n">pep_dict</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">met_peps</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">identifier_stripped</span><span class="p">)</span>
            <span class="n">prot_dict</span><span class="p">[</span><span class="n">identifier_stripped</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">met_peps</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span> <span class="o">=</span> <span class="n">sp_set</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="n">pep_dict</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span> <span class="o">=</span> <span class="n">prot_dict</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Pyteomics Digest Finished...&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.inference">
        <code>inference</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.Exclusion">
        <code>
Exclusion            (<a class="autorefs autorefs-internal" title="pyproteininference.inference.Inference" href="#pyproteininference.inference.Inference">Inference</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Exclusion Inference class. This class contains methods that support the initialization of an
Exclusion inference method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>scored_data</code></td>
        <td><code>list</code></td>
        <td><p>a List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Exclusion</span><span class="p">(</span><span class="n">Inference</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Exclusion Inference class. This class contains methods that support the initialization of an</span>
<span class="sd">    Exclusion inference method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        scored_data (list): a List of scored [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method of the Exclusion Class.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">list_of_prots_not_in_db</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">list_of_peps_not_in_db</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method performs the Exclusion inference/grouping method.</span>

<span class="sd">        For the exclusion inference method groups cannot be created because all shared peptides are removed.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore Object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

        <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Exclusion Method&quot;</span><span class="p">)</span>
        <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Exclusion.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method of the Exclusion Class.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method of the Exclusion Class.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">list_of_prots_not_in_db</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">list_of_peps_not_in_db</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Exclusion.infer_proteins">
<code class="highlight language-python"><span class="n">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method performs the Exclusion inference/grouping method.</p>
<p>For the exclusion inference method groups cannot be created because all shared peptides are removed.</p>
<p>This method assigns the variables: <code>grouped_scored_proteins</code> and <code>protein_group_objects</code>.
These are both variables of the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a> and are
lists of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
and <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method performs the Exclusion inference/grouping method.</span>

<span class="sd">    For the exclusion inference method groups cannot be created because all shared peptides are removed.</span>

<span class="sd">    This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">    These are both variables of the [DataStore Object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">    lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

    <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Exclusion Method&quot;</span><span class="p">)</span>
    <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
        <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.FirstProtein">
        <code>
FirstProtein            (<a class="autorefs autorefs-internal" title="pyproteininference.inference.Inference" href="#pyproteininference.inference.Inference">Inference</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>FirstProtein Inference class. This class contains methods that support the initialization of a
FirstProtein inference method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">FirstProtein</span><span class="p">(</span><span class="n">Inference</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    FirstProtein Inference class. This class contains methods that support the initialization of a</span>
<span class="sd">    FirstProtein inference method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        FirstProtein Inference initialization method.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        Returns:</span>
<span class="sd">            object:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method performs the First Protein inference method.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the First Protein Method&quot;</span><span class="p">)</span>
        <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.FirstProtein.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>FirstProtein Inference initialization method.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>object</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    FirstProtein Inference initialization method.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    Returns:</span>
<span class="sd">        object:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.FirstProtein.infer_proteins">
<code class="highlight language-python"><span class="n">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method performs the First Protein inference method.</p>
<p>This method assigns the variables: <code>grouped_scored_proteins</code> and <code>protein_group_objects</code>.
These are both variables of the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a> and are
lists of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
and <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method performs the First Protein inference method.</span>

<span class="sd">    This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">    These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">    lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

    <span class="c1"># Get the higher or lower variable</span>
    <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the First Protein Method&quot;</span><span class="p">)</span>
    <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
        <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.Inclusion">
        <code>
Inclusion            (<a class="autorefs autorefs-internal" title="pyproteininference.inference.Inference" href="#pyproteininference.inference.Inference">Inference</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Inclusion Inference class. This class contains methods that support the initialization of an
Inclusion inference method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>scored_data</code></td>
        <td><code>list</code></td>
        <td><p>a List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Inclusion</span><span class="p">(</span><span class="n">Inference</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Inclusion Inference class. This class contains methods that support the initialization of an</span>
<span class="sd">    Inclusion inference method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        scored_data (list): a List of scored [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method of the Inclusion Inference method.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method performs the grouping for Inclusion.</span>

<span class="sd">        Inclusion actually does not do grouping as all peptides get assigned to all possible proteins</span>
<span class="sd">        and groups are not created.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore Object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

        <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Inclusion Method&quot;</span><span class="p">)</span>

        <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>

    <span class="k">def</span> <span class="nf">_apply_protein_group_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates the ProteinGroup objects for the inclusion inference type using protein groups from</span>
<span class="sd">         [_create_protein_groups][`pyproteininference.inference.Inference._create_protein_groups].</span>

<span class="sd">        Args:</span>
<span class="sd">            grouped_protein_objects (list): list of grouped [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a Dictionary that contains a list of [ProteinGroup][pyproteininference.physical.ProteinGroup]</span>
<span class="sd">                objects (key:&quot;group_objects&quot;) and a list of</span>
<span class="sd">                grouped [Protein][pyproteininference.physical.Protein] objects (key:&quot;grouped_protein_objects&quot;).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sp_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>

        <span class="n">prot_pep_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="c1"># Here we create group ID&#39;s</span>
        <span class="n">group_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">grouped_protein_objects</span><span class="p">:</span>
            <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">group_id</span> <span class="o">=</span> <span class="n">group_id</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">pg</span> <span class="o">=</span> <span class="n">ProteinGroup</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Created Protein Group with ID: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">group_id</span><span class="p">)))</span>
            <span class="k">for</span> <span class="n">prot</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="p">:</span>
                <span class="n">cur_protein</span> <span class="o">=</span> <span class="n">prot</span>
                <span class="c1"># The following loop assigns group_id&#39;s, reviewed/unreviewed status, and number of unique peptides...</span>
                <span class="k">if</span> <span class="n">group_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">cur_protein</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">sp_protein_set</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">reviewed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">unreviewed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">cur_identifier</span> <span class="o">=</span> <span class="n">cur_protein</span><span class="o">.</span><span class="n">identifier</span>
                <span class="n">cur_protein</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_identifier</span><span class="p">])</span>
                <span class="c1"># Here append the number of unique peptides... so we can use this as secondary sorting...</span>
                <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_protein</span><span class="p">)</span>
                <span class="c1"># Sorted protein_groups then becomes a list of lists... of protein objects</span>

            <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="n">protein_list</span>
            <span class="n">protein_group_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="s2">&quot;group_objects&quot;</span><span class="p">:</span> <span class="n">protein_group_objects</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Inclusion.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method of the Inclusion Inference method.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method of the Inclusion Inference method.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Inclusion.infer_proteins">
<code class="highlight language-python"><span class="n">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method performs the grouping for Inclusion.</p>
<p>Inclusion actually does not do grouping as all peptides get assigned to all possible proteins
and groups are not created.</p>
<p>This method assigns the variables: <code>grouped_scored_proteins</code> and <code>protein_group_objects</code>.
These are both variables of the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a> and are
lists of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
and <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method performs the grouping for Inclusion.</span>

<span class="sd">    Inclusion actually does not do grouping as all peptides get assigned to all possible proteins</span>
<span class="sd">    and groups are not created.</span>

<span class="sd">    This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">    These are both variables of the [DataStore Object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">    lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span><span class="n">scored_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span><span class="p">)</span>

    <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Inclusion Method&quot;</span><span class="p">)</span>

    <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
        <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.Inference">
        <code>
Inference        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Parent Inference class for all inference/grouper subset classes.
The base Inference class contains several methods that are shared across the Inference sub-classes.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Inference</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parent Inference class for all inference/grouper subset classes.</span>
<span class="sd">    The base Inference class contains several methods that are shared across the Inference sub-classes.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">PARSIMONY</span> <span class="o">=</span> <span class="s2">&quot;parsimony&quot;</span>
    <span class="n">INCLUSION</span> <span class="o">=</span> <span class="s2">&quot;inclusion&quot;</span>
    <span class="n">EXCLUSION</span> <span class="o">=</span> <span class="s2">&quot;exclusion&quot;</span>
    <span class="n">FIRST_PROTEIN</span> <span class="o">=</span> <span class="s2">&quot;first_protein&quot;</span>
    <span class="n">PEPTIDE_CENTRIC</span> <span class="o">=</span> <span class="s2">&quot;peptide_centric&quot;</span>

    <span class="n">INFERENCE_TYPES</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">PARSIMONY</span><span class="p">,</span>
        <span class="n">INCLUSION</span><span class="p">,</span>
        <span class="n">EXCLUSION</span><span class="p">,</span>
        <span class="n">FIRST_PROTEIN</span><span class="p">,</span>
        <span class="n">PEPTIDE_CENTRIC</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">INFERENCE_NAME_MAP</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">PARSIMONY</span><span class="p">:</span> <span class="s2">&quot;Parsimony&quot;</span><span class="p">,</span>
        <span class="n">INCLUSION</span><span class="p">:</span> <span class="s2">&quot;Inclusion&quot;</span><span class="p">,</span>
        <span class="n">EXCLUSION</span><span class="p">:</span> <span class="s2">&quot;Exclusion&quot;</span><span class="p">,</span>
        <span class="n">FIRST_PROTEIN</span><span class="p">:</span> <span class="s2">&quot;First Protein&quot;</span><span class="p">,</span>
        <span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span> <span class="s2">&quot;Peptide Centric&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">SUBSET_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;subset_peptides&quot;</span>
    <span class="n">SHARED_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;shared_peptides&quot;</span>
    <span class="n">NONE_GROUPING</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">GROUPING_TYPES</span> <span class="o">=</span> <span class="p">[</span><span class="n">SUBSET_PEPTIDES</span><span class="p">,</span> <span class="n">SHARED_PEPTIDES</span><span class="p">,</span> <span class="n">NONE_GROUPING</span><span class="p">]</span>

    <span class="n">PULP</span> <span class="o">=</span> <span class="s2">&quot;pulp&quot;</span>
    <span class="n">LP_SOLVERS</span> <span class="o">=</span> <span class="p">[</span><span class="n">PULP</span><span class="p">]</span>

    <span class="n">ALL_SHARED_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
    <span class="n">BEST_SHARED_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;best&quot;</span>
    <span class="n">NONE_SHARED_PEPTIDES</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">SHARED_PEPTIDE_TYPES</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">ALL_SHARED_PEPTIDES</span><span class="p">,</span>
        <span class="n">BEST_SHARED_PEPTIDES</span><span class="p">,</span>
        <span class="n">NONE_SHARED_PEPTIDES</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method of Inference object.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">run_inference</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class method dispatches to one of the five different inference classes/models</span>
<span class="sd">        based on input from the [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter]</span>
<span class="sd">        object.</span>
<span class="sd">        The methods are &quot;parsimony&quot;, &quot;inclusion&quot;, &quot;exclusion&quot;, &quot;peptide_centric&quot;, and &quot;first_protein&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pyproteininference.inference.Inference.run_inference(data=data,digest=digest)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">inference_type</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Inference with Inference Type: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">inference_type</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">:</span>
            <span class="n">group</span> <span class="o">=</span> <span class="n">Parsimony</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">:</span>
            <span class="n">group</span> <span class="o">=</span> <span class="n">Inclusion</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
            <span class="n">group</span> <span class="o">=</span> <span class="n">Exclusion</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
            <span class="n">group</span> <span class="o">=</span> <span class="n">FirstProtein</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span>
            <span class="n">group</span> <span class="o">=</span> <span class="n">PeptideCentric</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
            <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_create_protein_groups</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">scored_proteins</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method sets up protein groups for inference methods that do not need grouping.</span>

<span class="sd">        Args:</span>
<span class="sd">            scored_proteins (list): List of scored [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of lists of scored [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">scored_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">scored_proteins</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">),</span> <span class="n">k</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">prot_pep_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="n">restricted_peptides_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>

        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_objects</span> <span class="ow">in</span> <span class="n">scored_proteins</span><span class="p">:</span>
            <span class="n">cur_protein_identifier</span> <span class="o">=</span> <span class="n">protein_objects</span><span class="o">.</span><span class="n">identifier</span>

            <span class="c1"># Set peptide variable if the peptide is in the restricted peptide set</span>
            <span class="c1"># Sort the peptides alphabetically</span>
            <span class="n">protein_objects</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_protein_identifier</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">restricted_peptides_set</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">protein_list_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_objects</span><span class="p">]</span>
            <span class="n">grouped_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_list_group</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">grouped_proteins</span>

    <span class="k">def</span> <span class="nf">_apply_protein_group_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates the ProteinGroup objects from the output of</span>
<span class="sd">            [_create_protein_groups][`pyproteininference.inference.Inference._create_protein_groups].</span>

<span class="sd">        Args:</span>
<span class="sd">            grouped_protein_objects (list): list of grouped [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a Dictionary that contains a list of [ProteinGroup][pyproteininference.physical.ProteinGroup]</span>
<span class="sd">                objects (key:&quot;group_objects&quot;) and a list of grouped [Protein][pyproteininference.physical.Protein]</span>
<span class="sd">                objects (key:&quot;grouped_protein_objects&quot;).</span>


<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sp_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>

        <span class="n">prot_pep_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="c1"># Here we create group ID&#39;s</span>
        <span class="n">group_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">grouped_protein_objects</span><span class="p">:</span>
            <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">group_id</span> <span class="o">=</span> <span class="n">group_id</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">pg</span> <span class="o">=</span> <span class="n">ProteinGroup</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Created Protein Group with ID: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">group_id</span><span class="p">)))</span>
            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="p">:</span>
                <span class="n">cur_protein</span> <span class="o">=</span> <span class="n">protein</span>
                <span class="c1"># The following loop assigns group_id&#39;s, reviewed/unreviewed status, and number of unique peptides...</span>
                <span class="k">if</span> <span class="n">group_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">protein</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">sp_protein_set</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">reviewed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">unreviewed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">cur_identifier</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">identifier</span>
                <span class="n">cur_protein</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_identifier</span><span class="p">])</span>
                <span class="c1"># Here append the number of unique peptides... so we can use this as secondary sorting...</span>
                <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_protein</span><span class="p">)</span>
                <span class="c1"># Sorted protein_groups then becomes a list of lists... of protein objects</span>

            <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="n">protein_list</span>
            <span class="n">protein_group_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="s2">&quot;group_objects&quot;</span><span class="p">:</span> <span class="n">protein_group_objects</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">


























  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Inference.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method of Inference object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method of Inference object.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Inference.run_inference">
<code class="highlight language-python"><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>This class method dispatches to one of the five different inference classes/models
based on input from the <a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a>
object.
The methods are "parsimony", "inclusion", "exclusion", "peptide_centric", and "first_protein".</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">Inference</span><span class="o">.</span><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span><span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">run_inference</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class method dispatches to one of the five different inference classes/models</span>
<span class="sd">    based on input from the [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter]</span>
<span class="sd">    object.</span>
<span class="sd">    The methods are &quot;parsimony&quot;, &quot;inclusion&quot;, &quot;exclusion&quot;, &quot;peptide_centric&quot;, and &quot;first_protein&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pyproteininference.inference.Inference.run_inference(data=data,digest=digest)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">inference_type</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running Inference with Inference Type: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">inference_type</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">:</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">Parsimony</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">INCLUSION</span><span class="p">:</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">Inclusion</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">Exclusion</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">FirstProtein</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC</span><span class="p">:</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">PeptideCentric</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">infer_proteins</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.Parsimony">
        <code>
Parsimony            (<a class="autorefs autorefs-internal" title="pyproteininference.inference.Inference" href="#pyproteininference.inference.Inference">Inference</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Parsimony Inference class. This class contains methods that support the initialization of a
Parsimony inference method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>scored_data</code></td>
        <td><code>list</code></td>
        <td><p>a List of scored <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></td>
      </tr>
      <tr>
        <td><code>lead_protein_set</code></td>
        <td><code>set</code></td>
        <td><p>Set of protein strings that are classified as leads from the LP solver.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Parsimony</span><span class="p">(</span><span class="n">Inference</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parsimony Inference class. This class contains methods that support the initialization of a</span>
<span class="sd">    Parsimony inference method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        scored_data (list): a List of scored [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">        lead_protein_set (set): Set of protein strings that are classified as leads from the LP solver.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method of the Parsimony object.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lead_protein_set</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span>

    <span class="k">def</span> <span class="nf">_create_protein_groups</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">all_scored_proteins</span><span class="p">,</span>
        <span class="n">lead_protein_objects</span><span class="p">,</span>
        <span class="n">grouping_type</span><span class="o">=</span><span class="s2">&quot;shared_peptides&quot;</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method that creates a list of lists of [Protein][pyproteininference.physical.Protein]</span>
<span class="sd">        objects for the Parsimony inference object.</span>
<span class="sd">        These list of lists are &quot;groups&quot; and the proteins get grouped them according to grouping_type variable.</span>

<span class="sd">        Args:</span>
<span class="sd">            all_scored_proteins (list): list of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            lead_protein_objects (list): list of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">                Only needed if inference_type=parsimony.</span>
<span class="sd">            grouping_type: (str): One of `GROUPING_TYPES`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: list of lists of [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Grouping Peptides with Grouping Type: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">grouping_type</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Grouping Peptides with Inference Type: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY</span><span class="p">))</span>

        <span class="n">all_scored_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">all_scored_proteins</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">),</span> <span class="n">k</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">lead_scored_proteins</span> <span class="o">=</span> <span class="n">lead_protein_objects</span>
        <span class="n">lead_scored_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">lead_scored_proteins</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">),</span> <span class="n">k</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">protein_finder</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_scored_proteins</span><span class="p">]</span>

        <span class="n">prot_pep_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="n">protein_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">restricted_peptides_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">restricted_peptides</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">picked_removed</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">picked_proteins_removed</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
            <span class="n">picked_removed</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="n">missing_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">in_silico_peptides_to_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>
        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_objects</span> <span class="ow">in</span> <span class="n">lead_scored_proteins</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">protein_objects</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">protein_tracker</span><span class="p">:</span>
                <span class="n">protein_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">protein_objects</span><span class="p">)</span>
                <span class="n">cur_protein_identifier</span> <span class="o">=</span> <span class="n">protein_objects</span><span class="o">.</span><span class="n">identifier</span>

                <span class="c1"># Set peptide variable if the peptide is in the restricted peptide set</span>
                <span class="c1"># Sort the peptides alphabetically</span>
                <span class="n">protein_objects</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                    <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_protein_identifier</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">restricted_peptides_set</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="n">protein_list_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">protein_objects</span><span class="p">]</span>
                <span class="n">current_peptides</span> <span class="o">=</span> <span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_protein_identifier</span><span class="p">]</span>

                <span class="n">current_grouped_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="k">for</span> <span class="p">(</span>
                    <span class="n">peptide</span>
                <span class="p">)</span> <span class="ow">in</span> <span class="n">current_peptides</span><span class="p">:</span>  <span class="c1"># Probably put an if here... if peptide is in the list of peptide after being</span>
                    <span class="c1"># restricted by datastore.RestrictMainData</span>
                    <span class="k">if</span> <span class="n">peptide</span> <span class="ow">in</span> <span class="n">restricted_peptides_set</span><span class="p">:</span>
                        <span class="c1"># Get the proteins that map to the current peptide using in_silico_peptides_to_proteins</span>
                        <span class="c1"># First make sure our peptide is formatted properly...</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                            <span class="c1"># If the peptide is not all upper case or if its not all alphabetical...</span>
                            <span class="n">peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide</span><span class="p">)</span>
                        <span class="n">potential_protein_list</span> <span class="o">=</span> <span class="n">in_silico_peptides_to_proteins</span><span class="p">[</span><span class="n">peptide</span><span class="p">]</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">potential_protein_list</span><span class="p">:</span>
                            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                                <span class="s2">&quot;Protein </span><span class="si">{}</span><span class="s2"> and Peptide </span><span class="si">{}</span><span class="s2"> is not in database...&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                    <span class="n">protein_objects</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span> <span class="n">peptide</span>
                                <span class="p">)</span>
                            <span class="p">)</span>

                        <span class="c1"># Assign proteins to groups based on shared peptide... unless the protein is equivalent</span>
                        <span class="c1"># to the current identifier</span>
                        <span class="k">if</span> <span class="n">grouping_type</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">NONE_GROUPING</span><span class="p">:</span>
                            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">potential_protein_list</span><span class="p">:</span>
                                <span class="c1"># If statement below to avoid grouping the same protein twice and to not group the lead</span>
                                <span class="k">if</span> <span class="p">(</span>
                                    <span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">current_grouped_proteins</span>
                                    <span class="ow">and</span> <span class="n">protein</span> <span class="o">!=</span> <span class="n">cur_protein_identifier</span>
                                    <span class="ow">and</span> <span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">picked_removed</span>
                                    <span class="ow">and</span> <span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">missing_proteins</span>
                                <span class="p">):</span>
                                    <span class="k">try</span><span class="p">:</span>
                                        <span class="c1"># Try to find its object using protein_finder (list of identifiers) and</span>
                                        <span class="c1"># lead_scored_proteins (list of Protein Objects)</span>
                                        <span class="n">cur_index</span> <span class="o">=</span> <span class="n">protein_finder</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
                                        <span class="n">current_protein_object</span> <span class="o">=</span> <span class="n">all_scored_proteins</span><span class="p">[</span><span class="n">cur_index</span><span class="p">]</span>
                                        <span class="k">if</span> <span class="ow">not</span> <span class="n">current_protein_object</span><span class="o">.</span><span class="n">peptides</span><span class="p">:</span>
                                            <span class="n">current_protein_object</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                                                <span class="nb">sorted</span><span class="p">(</span>
                                                    <span class="p">[</span>
                                                        <span class="n">x</span>
                                                        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">current_protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">]</span>
                                                        <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">restricted_peptides_set</span>
                                                    <span class="p">]</span>
                                                <span class="p">)</span>
                                            <span class="p">)</span>
                                        <span class="k">if</span> <span class="n">grouping_type</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHARED_PEPTIDES</span><span class="p">:</span>
                                            <span class="n">current_grouped_proteins</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_protein_object</span><span class="p">)</span>
                                        <span class="k">elif</span> <span class="n">grouping_type</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">SUBSET_PEPTIDES</span><span class="p">:</span>
                                            <span class="k">if</span> <span class="n">current_protein_object</span><span class="o">.</span><span class="n">peptides</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">protein_objects</span><span class="o">.</span><span class="n">peptides</span><span class="p">):</span>
                                                <span class="n">current_grouped_proteins</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_protein_object</span><span class="p">)</span>
                                                <span class="n">protein_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_protein_object</span><span class="p">)</span>
                                            <span class="k">else</span><span class="p">:</span>
                                                <span class="k">pass</span>
                                        <span class="k">else</span><span class="p">:</span>
                                            <span class="k">pass</span>
                                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                                        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                                            <span class="s2">&quot;Protein from DB </span><span class="si">{}</span><span class="s2"> not found with protein finder for peptide </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                                <span class="n">protein</span><span class="p">,</span> <span class="n">peptide</span>
                                            <span class="p">)</span>
                                        <span class="p">)</span>
                                        <span class="n">missing_proteins</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

                                <span class="k">else</span><span class="p">:</span>
                                    <span class="k">pass</span>
                <span class="c1"># Add the proteins to the lead if they share peptides...</span>
                <span class="n">protein_list_group</span> <span class="o">=</span> <span class="n">protein_list_group</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">current_grouped_proteins</span><span class="p">)</span>
                <span class="c1"># protein_list_group at first is just the lead protein object...</span>
                <span class="c1"># We then try apply grouping by looking at all peptide from the lead...</span>
                <span class="c1"># For all of these peptide look at all other non lead proteins and try to assign them to the group...</span>
                <span class="c1"># We assign the entire protein object as well... in the above try/except</span>
                <span class="c1"># Then append this sub group to the main list</span>
                <span class="c1"># The variable grouped_proteins is now a list of lists which each element being a Protein object and</span>
                <span class="c1"># each list of protein objects corresponding to a group</span>
                <span class="n">grouped_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_list_group</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">grouped_proteins</span>

    <span class="k">def</span> <span class="nf">_swissprot_and_isoform_override</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">scored_data</span><span class="p">,</span>
        <span class="n">grouped_proteins</span><span class="p">,</span>
        <span class="n">override_type</span><span class="o">=</span><span class="s2">&quot;soft&quot;</span><span class="p">,</span>
        <span class="n">isoform_override</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This internal method creates and reorders protein groups based on criteria such as Reviewed/Unreviewed</span>
<span class="sd">        Identifiers as well as Canonincal/Isoform Identifiers.</span>
<span class="sd">        This method is only used with parsimony inference type.</span>

<span class="sd">        Args:</span>
<span class="sd">            scored_data (list): list of scored [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            grouped_proteins:  list of grouped [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            override_type (str): &quot;soft&quot; or &quot;hard&quot; to indicate Reviewed/Unreviewed override. &quot;soft&quot; is preferred and</span>
<span class="sd">                default.</span>
<span class="sd">            isoform_override (bool): True/False on whether to favor canonical forms vs isoforms as group leads.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a Dictionary that contains a list of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects</span>
<span class="sd">            (key:&quot;group_objects&quot;) and a list of grouped [Protein][pyproteininference.physical.Protein]</span>
<span class="sd">            objects (key:&quot;grouped_protein_objects&quot;).</span>


<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sp_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
        <span class="n">scored_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">scored_data</span><span class="p">)</span>
        <span class="n">protein_finder</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">scored_proteins</span><span class="p">]</span>

        <span class="n">prot_pep_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group IDs... and Executing </span><span class="si">{}</span><span class="s2"> Swissprot Override...&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">override_type</span><span class="p">))</span>
        <span class="c1"># Here we create group ID&#39;s for all groups and do some sorting</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">group_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">leads</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">grouped_proteins</span><span class="p">:</span>
            <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">group_id</span> <span class="o">=</span> <span class="n">group_id</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="c1"># Make a protein group</span>
            <span class="n">pg</span> <span class="o">=</span> <span class="n">ProteinGroup</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Created Protein Group with ID: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">group_id</span><span class="p">)))</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="p">:</span>
                <span class="c1"># Loop over all proteins in the original group</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="c1"># The following loop assigns group_id&#39;s, reviewed/unreviewed status, and number of unique peptides</span>
                    <span class="n">pindex</span> <span class="o">=</span> <span class="n">protein_finder</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">prots</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                    <span class="c1"># Attempt to find the protein object by identifier</span>
                    <span class="n">cur_protein</span> <span class="o">=</span> <span class="n">scored_proteins</span><span class="p">[</span><span class="n">pindex</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">group_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="p">:</span>
                        <span class="n">cur_protein</span><span class="o">.</span><span class="n">group_identification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">prots</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">in</span> <span class="n">sp_protein_set</span><span class="p">:</span>
                        <span class="n">cur_protein</span><span class="o">.</span><span class="n">reviewed</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">cur_protein</span><span class="o">.</span><span class="n">unreviewed</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">cur_identifier</span> <span class="o">=</span> <span class="n">prots</span><span class="o">.</span><span class="n">identifier</span>
                    <span class="n">cur_protein</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">prot_pep_dict</span><span class="p">[</span><span class="n">cur_identifier</span><span class="p">])</span>
                    <span class="c1"># Here append the number of unique peptides... so we can use this as secondary sorting...</span>
                    <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_protein</span><span class="p">)</span>
                    <span class="c1"># Sorted groups then becomes a list of lists... of protein objects</span>

                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="c1"># Here we pass if the protein does not have a score...</span>
                    <span class="c1"># Potentially it got &#39;picked&#39; (removed) by protein picker...</span>
                    <span class="k">pass</span>

            <span class="c1"># Sort protein sub group</span>
            <span class="n">protein_list</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_sub_groups</span><span class="p">(</span>
                <span class="n">protein_list</span><span class="o">=</span><span class="n">protein_list</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">higher_or_lower</span>
            <span class="p">)</span>

            <span class="c1"># grouped_protein_objects is the MAIN list of lists with grouped protein objects</span>
            <span class="n">grouped_protein_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_list</span><span class="p">)</span>
            <span class="c1"># If the lead is reviewed append it to leads and do nothing else...</span>
            <span class="c1"># If the lead is unreviewed then try to replace it with the best reviewed hit</span>
            <span class="c1"># Run swissprot override</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">:</span>
                <span class="n">sp_override</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_swissprot_override</span><span class="p">(</span>
                    <span class="n">protein_list</span><span class="o">=</span><span class="n">protein_list</span><span class="p">,</span>
                    <span class="n">leads</span><span class="o">=</span><span class="n">leads</span><span class="p">,</span>
                    <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span>
                    <span class="n">override_type</span><span class="o">=</span><span class="n">override_type</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">sp_override</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
                <span class="n">leads</span> <span class="o">=</span> <span class="n">sp_override</span><span class="p">[</span><span class="s2">&quot;leads&quot;</span><span class="p">]</span>
                <span class="n">protein_list</span> <span class="o">=</span> <span class="n">sp_override</span><span class="p">[</span><span class="s2">&quot;protein_list&quot;</span><span class="p">]</span>

            <span class="c1"># Run isoform override If we want to run isoform_override and if the isoform symbol exists...</span>
            <span class="k">if</span> <span class="n">isoform_override</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">:</span>
                <span class="n">iso_override</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_isoform_override</span><span class="p">(</span>
                    <span class="n">protein_list</span><span class="o">=</span><span class="n">protein_list</span><span class="p">,</span>
                    <span class="n">leads</span><span class="o">=</span><span class="n">leads</span><span class="p">,</span>
                    <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">iso_override</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
                <span class="n">leads</span> <span class="o">=</span> <span class="n">iso_override</span><span class="p">[</span><span class="s2">&quot;leads&quot;</span><span class="p">]</span>
                <span class="n">protein_list</span> <span class="o">=</span> <span class="n">iso_override</span><span class="p">[</span><span class="s2">&quot;protein_list&quot;</span><span class="p">]</span>

            <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="n">protein_list</span>
            <span class="n">protein_group_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="s2">&quot;group_objects&quot;</span><span class="p">:</span> <span class="n">protein_group_objects</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>

    <span class="k">def</span> <span class="nf">_swissprot_override</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_list</span><span class="p">,</span> <span class="n">leads</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">override_type</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method re-assigns protein group leads if the lead is an unreviewed protein and if the protein group</span>
<span class="sd">         contains a reviewed protein that contains the exact same set of peptides as the unreviewed lead.</span>
<span class="sd">        This method is here to provide consistency to the output.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_list (list): List of grouped [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            leads (set): Set of string protein identifiers that have been identified as a lead.</span>
<span class="sd">            grouped_protein_objects (list): List of protein_list lists.</span>
<span class="sd">            override_type (str): &quot;soft&quot; or &quot;hard&quot; on how to override non reviewed identifiers. &quot;soft&quot; is preferred.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: leads (set): Set of string protein identifiers that have been identified as a lead.</span>
<span class="sd">             Updated to reflect lead changes.</span>
<span class="sd">            grouped_protein_objects (list): List of protein_list lists. Updated to reflect lead changes.</span>
<span class="sd">            protein_list (list): List of grouped [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">                Updated to reflect lead changes.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
            <span class="c1"># If the lead is unreviewed attempt to replace it...</span>
            <span class="c1"># Start to loop through protein_list which is the current group...</span>
            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="c1"># Find the first reviewed it... if its not a lead protein already then do score swap and break...</span>
                <span class="k">if</span> <span class="n">protein</span><span class="o">.</span><span class="n">reviewed</span><span class="p">:</span>
                    <span class="n">best_swiss_prot_prot</span> <span class="o">=</span> <span class="n">protein</span>

                    <span class="k">if</span> <span class="n">override_type</span> <span class="o">==</span> <span class="s2">&quot;soft&quot;</span><span class="p">:</span>
                        <span class="c1"># If the lead proteins peptides are a subset of the best swissprot.... then swap the proteins.</span>
                        <span class="c1"># (meaning equal peptides or the swissprot completely covers the tremble reference)</span>
                        <span class="k">if</span> <span class="n">best_swiss_prot_prot</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leads</span> <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span><span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span>
                            <span class="nb">set</span><span class="p">(</span><span class="n">best_swiss_prot_prot</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
                        <span class="p">):</span>
                            <span class="c1"># We use -1 as the idex of grouped_protein_objects because the current &#39;protein_list&#39; is</span>
                            <span class="c1"># the last entry appended to scores grouped</span>
                            <span class="c1"># Essentially grouped_protein_objects[-1]==protein_list</span>
                            <span class="c1"># We need this syntax so we can switch the location of the unreviewed lead identifier with</span>
                            <span class="c1"># the best reviewed identifier in grouped_protein_objects</span>
                            <span class="n">swiss_prot_override_index</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">best_swiss_prot_prot</span><span class="p">)</span>
                            <span class="n">cur_tr_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="p">(</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">swiss_prot_override_index</span><span class="p">],</span>
                            <span class="p">)</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">swiss_prot_override_index</span><span class="p">],</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                            <span class="p">)</span>
                            <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">swiss_prot_override_index</span><span class="p">],</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="n">new_sp_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                                <span class="s2">&quot;Overriding Unreviewed </span><span class="si">{}</span><span class="s2"> with Reviewed </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                    <span class="n">cur_tr_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span> <span class="n">new_sp_lead</span><span class="o">.</span><span class="n">identifier</span>
                                <span class="p">)</span>
                            <span class="p">)</span>

                            <span class="c1"># Append new_sp_lead protein to leads, to make sure we dont repeat leads</span>
                            <span class="n">leads</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_sp_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                            <span class="k">break</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># If no reviewed and none not in leads then pass...</span>
                            <span class="k">pass</span>

                    <span class="k">if</span> <span class="n">override_type</span> <span class="o">==</span> <span class="s2">&quot;hard&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">best_swiss_prot_prot</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leads</span><span class="p">:</span>
                            <span class="c1"># We use -1 as the index of grouped_protein_objects because the current &#39;protein_list&#39;</span>
                            <span class="c1"># is the last entry appended to grouped_protein_objects</span>
                            <span class="c1"># Essentially grouped_protein_objects[-1]==protein_list</span>
                            <span class="c1"># We need this syntax so we can switch the location of the unreviewed lead identifier</span>
                            <span class="c1"># with the best reviewed identifier in grouped_protein_objects</span>
                            <span class="n">swiss_prot_override_index</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">best_swiss_prot_prot</span><span class="p">)</span>
                            <span class="n">cur_tr_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="c1"># Re-assigning the value within the index will also reassign the value in protein_list...</span>
                            <span class="c1"># This is because grouped_protein_objects[-1] equals protein_list</span>
                            <span class="c1"># So we do not have to reassign values in protein_list</span>
                            <span class="p">(</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">swiss_prot_override_index</span><span class="p">],</span>
                            <span class="p">)</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">swiss_prot_override_index</span><span class="p">],</span>
                                <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                            <span class="p">)</span>
                            <span class="n">new_sp_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                                <span class="s2">&quot;Overriding Unreviewed </span><span class="si">{}</span><span class="s2"> with Reviewed </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                    <span class="n">cur_tr_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span> <span class="n">new_sp_lead</span><span class="o">.</span><span class="n">identifier</span>
                                <span class="p">)</span>
                            <span class="p">)</span>

                            <span class="c1"># Append new_sp_lead protein to leads, to make sure we dont repeat leads</span>
                            <span class="n">leads</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_sp_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                            <span class="k">break</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># If no reviewed and none not in leads then pass...</span>
                            <span class="k">pass</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="k">pass</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">leads</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;leads&quot;</span><span class="p">:</span> <span class="n">leads</span><span class="p">,</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="s2">&quot;protein_list&quot;</span><span class="p">:</span> <span class="n">protein_list</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>

    <span class="k">def</span> <span class="nf">_isoform_override</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_list</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">leads</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method re-assigns protein group leads if the lead is an isoform protein and if the protein group contains</span>
<span class="sd">        a canonical protein that contains the exact same set of peptides as the isoform lead.</span>
<span class="sd">        This method is here to provide consistency to the output.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_list (list): List of grouped [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">            leads (set): Set of string protein identifiers that have been identified as a lead.</span>
<span class="sd">            grouped_protein_objects (list): List of protein_list lists.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: leads (set): Set of string protein identifiers that have been identified as a lead. Updated to</span>
<span class="sd">                reflect lead changes.</span>
<span class="sd">            grouped_protein_objects (list): List of protein_list lists. Updated to reflect lead changes.</span>
<span class="sd">            protein_list (list): List of grouped [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">                Updated to reflect lead changes.</span>


<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">:</span>
            <span class="n">pure_id</span> <span class="o">=</span> <span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="c1"># Start to loop through protein_list which is the current group...</span>
            <span class="k">for</span> <span class="n">potential_replacement</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">isoform_override</span> <span class="o">=</span> <span class="n">potential_replacement</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">isoform_override</span><span class="o">.</span><span class="n">identifier</span> <span class="o">==</span> <span class="n">pure_id</span>
                    <span class="ow">and</span> <span class="n">isoform_override</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leads</span>
                    <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span><span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">isoform_override</span><span class="o">.</span><span class="n">peptides</span><span class="p">))</span>
                <span class="p">):</span>
                    <span class="n">isoform_override_index</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">isoform_override</span><span class="p">)</span>
                    <span class="n">cur_iso_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="c1"># Re-assigning the value within the index will also reassign the value in protein_list...</span>
                    <span class="c1"># This is because grouped_protein_objects[-1] equals protein_list</span>
                    <span class="c1"># So we do not have to reassign values in protein_list</span>
                    <span class="p">(</span><span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">isoform_override_index</span><span class="p">],)</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">isoform_override_index</span><span class="p">],</span>
                        <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                    <span class="p">)</span>
                    <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">isoform_override_index</span><span class="p">],</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

                    <span class="n">new_iso_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                        <span class="s2">&quot;Overriding Isoform </span><span class="si">{}</span><span class="s2"> with </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cur_iso_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span> <span class="n">new_iso_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="n">leads</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">protein_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;leads&quot;</span><span class="p">:</span> <span class="n">leads</span><span class="p">,</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">grouped_protein_objects</span><span class="p">,</span>
            <span class="s2">&quot;protein_list&quot;</span><span class="p">:</span> <span class="n">protein_list</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>

    <span class="k">def</span> <span class="nf">_reassign_protein_group_leads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_group_objects</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This internal method corrects leads that are improperly assigned in the parsimony inference method.</span>
<span class="sd">        This method acts on the protein group objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_group_objects (list): List of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            protein_group_objects (list): List of [ProteinGroup][pyproteininference.physical.ProteinGroup] objects</span>
<span class="sd">            where leads have been reassigned properly.</span>


<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">high_low_better</span><span class="p">:</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">high_low_better</span>

        <span class="c1"># Sometimes we have cases where:</span>
        <span class="c1"># protein a maps to peptides 1,2,3</span>
        <span class="c1"># protein b maps to peptides 1,2</span>
        <span class="c1"># protein c maps to a bunch of peptides and peptide 3</span>
        <span class="c1"># Therefore, in the model proteins a and b are equivalent in that they map to 2 peptides together - 1 and 2.</span>
        <span class="c1"># peptide 3 maps to a but also to c...</span>
        <span class="c1"># Sometimes the model (pulp) will spit out protein b as the lead... we wish to swap protein b as the lead with</span>
        <span class="c1"># protein a because it will likely have a better score...</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Potentially Reassigning Protein Group leads...&quot;</span><span class="p">)</span>
        <span class="n">lead_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_group_objects</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">protein_group_objects</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">)):</span>  <span class="c1"># Loop over all sub proteins in the group...</span>
                <span class="c1"># if the lead proteins peptides are a subset of one of its proteins in the group, and the secondary</span>
                <span class="c1"># protein is not a lead protein and its score is better than the leads... and it has more peptides...</span>
                <span class="n">new_lead</span> <span class="o">=</span> <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                <span class="n">old_lead</span> <span class="o">=</span> <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">))</span>
                        <span class="ow">and</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">lead_protein_set</span>
                        <span class="ow">and</span> <span class="n">old_lead</span><span class="o">.</span><span class="n">score</span> <span class="o">&lt;=</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">score</span>
                        <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
                    <span class="p">):</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="s2">&quot;protein </span><span class="si">{}</span><span class="s2"> will replace protein </span><span class="si">{}</span><span class="s2"> as lead, with index </span><span class="si">{}</span><span class="s2">, New Num Peptides: </span><span class="si">{}</span><span class="s2">, &quot;</span>
                            <span class="s2">&quot;Old Num Peptides: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="c1"># Swap their positions in the list</span>
                        <span class="p">(</span>
                            <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="p">(</span><span class="n">new_lead</span><span class="p">,</span> <span class="n">old_lead</span><span class="p">)</span>
                        <span class="k">break</span>

                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">))</span>
                        <span class="ow">and</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">lead_protein_set</span>
                        <span class="ow">and</span> <span class="n">old_lead</span><span class="o">.</span><span class="n">score</span> <span class="o">&gt;=</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">score</span>
                        <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
                    <span class="p">):</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="s2">&quot;protein </span><span class="si">{}</span><span class="s2"> will replace protein </span><span class="si">{}</span><span class="s2"> as lead, with index </span><span class="si">{}</span><span class="s2">, New Num Peptides: </span><span class="si">{}</span><span class="s2">, &quot;</span>
                            <span class="s2">&quot;Old Num Peptides: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="c1"># Swap their positions in the list</span>
                        <span class="p">(</span>
                            <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">protein_group_objects</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="p">(</span><span class="n">new_lead</span><span class="p">,</span> <span class="n">old_lead</span><span class="p">)</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">protein_group_objects</span>

    <span class="k">def</span> <span class="nf">_reassign_protein_list_leads</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">grouped_protein_objects</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This internal method corrects leads that are improperly assigned in the parsimony inference method.</span>
<span class="sd">        This method acts on the grouped protein objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            grouped_protein_objects (list): List of [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of [Protein][pyproteininference.physical.Protein] objects where leads have been</span>
<span class="sd">                reassigned properly.</span>


<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">high_low_better</span><span class="p">:</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">higher_or_lower</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">high_low_better</span>

        <span class="c1"># Sometimes we have cases where:</span>
        <span class="c1"># protein a maps to peptides 1,2,3</span>
        <span class="c1"># protein b maps to peptides 1,2</span>
        <span class="c1"># protein c maps to a bunch of peptides and peptide 3</span>
        <span class="c1"># Therefore, in the model proteins a and b are equivalent in that they map to 2 peptides together - 1 and 2.</span>
        <span class="c1"># peptide 3 maps to a but also to c...</span>
        <span class="c1"># Sometimes the model (pulp) will spit out protein b as the lead... we wish to swap protein b as the lead with</span>
        <span class="c1"># protein a because it will likely have a better score...</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Potentially Reassigning Proteoin List leads...&quot;</span><span class="p">)</span>
        <span class="n">lead_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">grouped_protein_objects</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">grouped_protein_objects</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>  <span class="c1"># Loop over all sub proteins in the group...</span>
                <span class="c1"># if the lead proteins peptides are a subset of one of its proteins in the group, and the secondary</span>
                <span class="c1"># protein is not a lead protein and its score is better than the leads... and it has more peptides...</span>
                <span class="n">new_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span>
                <span class="n">old_lead</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">HIGHER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">))</span>
                        <span class="ow">and</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">lead_protein_set</span>
                        <span class="ow">and</span> <span class="n">old_lead</span><span class="o">.</span><span class="n">score</span> <span class="o">&lt;=</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">score</span>
                        <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
                    <span class="p">):</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="s2">&quot;protein </span><span class="si">{}</span><span class="s2"> will replace protein </span><span class="si">{}</span><span class="s2"> as lead, with index </span><span class="si">{}</span><span class="s2">, New Num Peptides: </span><span class="si">{}</span><span class="s2">, &quot;</span>
                            <span class="s2">&quot;Old Num Peptides: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="c1"># Swap their positions in the list</span>
                        <span class="p">(</span>
                            <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">],</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="p">(</span><span class="n">new_lead</span><span class="p">,</span> <span class="n">old_lead</span><span class="p">)</span>
                        <span class="k">break</span>

                <span class="k">if</span> <span class="n">higher_or_lower</span> <span class="o">==</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">LOWER_PSM_SCORE</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">))</span>
                        <span class="ow">and</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">lead_protein_set</span>
                        <span class="ow">and</span> <span class="n">old_lead</span><span class="o">.</span><span class="n">score</span> <span class="o">&gt;=</span> <span class="n">new_lead</span><span class="o">.</span><span class="n">score</span>
                        <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
                    <span class="p">):</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="s2">&quot;protein </span><span class="si">{}</span><span class="s2"> will replace protein </span><span class="si">{}</span><span class="s2"> as lead, with index </span><span class="si">{}</span><span class="s2">, New Num Peptides: </span><span class="si">{}</span><span class="s2">, &quot;</span>
                            <span class="s2">&quot;Old Num Peptides: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">peptides</span><span class="p">)),</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="n">lead_protein_set</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">old_lead</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
                        <span class="c1"># Swap their positions in the list</span>
                        <span class="p">(</span>
                            <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">grouped_protein_objects</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">],</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="p">(</span><span class="n">new_lead</span><span class="p">,</span> <span class="n">old_lead</span><span class="p">)</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">grouped_protein_objects</span>

    <span class="k">def</span> <span class="nf">_pulp_grouper</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This internal function uses pulp to solve the lp problem for parsimony then performs protein grouping with the</span>
<span class="sd">         various internal grouping functions.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Here we get the peptide to protein dictionary</span>
        <span class="n">pep_prot_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">()</span>

        <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_sorted_identifiers</span><span class="p">(</span><span class="n">scored</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Get all the proteins that we scored and the ones picked if picker was ran...</span>
        <span class="n">data_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">identifiers_sorted</span><span class="p">])</span>
        <span class="c1"># Get the set of peptides for each protein...</span>
        <span class="n">data_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_peptide_dictionary</span><span class="p">[</span><span class="n">x</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_proteins</span><span class="p">]</span>
        <span class="n">flat_peptides_in_data</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">data_peptides</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>

        <span class="n">peptide_sets</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Loop over the list of peptides...</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data_peptides</span><span class="p">)):</span>
            <span class="n">raw_peptides</span> <span class="o">=</span> <span class="n">data_peptides</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="n">peptide_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="c1"># Loop over each individual peptide per protein...</span>
            <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">raw_peptides</span><span class="p">:</span>
                <span class="n">peptide</span> <span class="o">=</span> <span class="n">peps</span>

                <span class="c1"># Remove mods...</span>
                <span class="n">new_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide</span><span class="p">)</span>
                <span class="c1"># Add it to a temporary set...</span>
                <span class="n">peptide_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">new_peptide</span><span class="p">)</span>
            <span class="c1"># Append this set to a new list...</span>
            <span class="n">peptide_sets</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">peptide_set</span><span class="p">)</span>
            <span class="c1"># Set that proteins peptides to be the unmodified ones...</span>
            <span class="n">data_peptides</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">peptide_set</span>

        <span class="c1"># Get them all...</span>
        <span class="n">all_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_peptides</span><span class="p">]</span>
        <span class="c1"># Remove redundant sets...</span>
        <span class="n">non_redundant_peptide_sets</span> <span class="o">=</span> <span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">OrderedDict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">peptide_sets</span><span class="p">)]</span>

        <span class="c1"># Loop over  the restricted list of peptides...</span>
        <span class="n">ind_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pep_sets</span> <span class="ow">in</span> <span class="n">non_redundant_peptide_sets</span><span class="p">:</span>
            <span class="c1"># Get its index in terms of the overall list...</span>
            <span class="n">ind_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">all_peptides</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pep_sets</span><span class="p">))</span>

        <span class="c1"># Get the protein based on the index</span>
        <span class="n">restricted_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">data_proteins</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data_peptides</span><span class="p">))</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ind_list</span><span class="p">]</span>

        <span class="c1"># Here we get the list of all proteins</span>
        <span class="n">plist</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">pep_prot_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">pep_prot_dict</span><span class="p">[</span><span class="n">peps</span><span class="p">]):</span>
                <span class="k">if</span> <span class="n">prots</span> <span class="ow">in</span> <span class="n">restricted_proteins</span> <span class="ow">and</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">flat_peptides_in_data</span><span class="p">:</span>
                    <span class="n">plist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prots</span><span class="p">)</span>

        <span class="c1"># Here we get the unique proteins</span>
        <span class="n">unique_prots</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">plist</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">())</span>
        <span class="n">unique_protein_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">unique_prots</span><span class="p">)</span>

        <span class="n">unique_prots_sorted</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">identifiers_sorted</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">unique_prots</span><span class="p">]</span>

        <span class="c1"># Define the protein variables with a lower bound of 0 and catgeory Integer</span>
        <span class="n">prots</span> <span class="o">=</span> <span class="n">pulp</span><span class="o">.</span><span class="n">LpVariable</span><span class="o">.</span><span class="n">dicts</span><span class="p">(</span><span class="s2">&quot;prot&quot;</span><span class="p">,</span> <span class="n">indices</span><span class="o">=</span><span class="n">unique_prots_sorted</span><span class="p">,</span> <span class="n">lowBound</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">cat</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">)</span>

        <span class="c1"># Define our Lp Problem which is to Minimize our objective function</span>
        <span class="n">prob</span> <span class="o">=</span> <span class="n">pulp</span><span class="o">.</span><span class="n">LpProblem</span><span class="p">(</span><span class="s2">&quot;Parsimony_Problem&quot;</span><span class="p">,</span> <span class="n">pulp</span><span class="o">.</span><span class="n">LpMinimize</span><span class="p">)</span>

        <span class="c1"># Define our objective function, which is to take the sum of all of our proteins and find the minimum set.</span>
        <span class="n">prob</span> <span class="o">+=</span> <span class="n">pulp</span><span class="o">.</span><span class="n">lpSum</span><span class="p">([</span><span class="n">prots</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">prots</span><span class="p">])</span>

        <span class="c1"># Set up our constraints. The constrains are as follows:</span>

        <span class="c1"># Loop over each peptide and determine the proteins it maps to...</span>
        <span class="c1"># Each peptide is a constraint with the proteins it maps to having to be greater than or equal to 1</span>
        <span class="c1"># In the case below we see that protein 3 has a unique peptide, protein 2 is redundant</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting peptides before looping&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">peptides</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pep_prot_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">prob</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="n">pulp</span><span class="o">.</span><span class="n">lpSum</span><span class="p">([</span><span class="n">prots</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pep_prot_dict</span><span class="p">[</span><span class="n">peptides</span><span class="p">]))</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">unique_protein_set</span><span class="p">])</span>
                    <span class="o">&gt;=</span> <span class="mi">1</span>
                <span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Not including protein </span><span class="si">{}</span><span class="s2"> in pulp model&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pep_prot_dict</span><span class="p">[</span><span class="n">peptides</span><span class="p">]))</span>

        <span class="n">prob</span><span class="o">.</span><span class="n">solve</span><span class="p">()</span>

        <span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
        <span class="n">scored_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">scored_data</span><span class="p">)</span>
        <span class="n">protein_finder</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">scored_proteins</span><span class="p">]</span>

        <span class="n">lead_protein_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">lead_protein_identifiers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="n">unique_prots_sorted</span><span class="p">:</span>
            <span class="n">parsimony_value</span> <span class="o">=</span> <span class="n">pulp</span><span class="o">.</span><span class="n">value</span><span class="p">(</span><span class="n">prots</span><span class="p">[</span><span class="n">proteins</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="n">protein_finder</span> <span class="ow">and</span> <span class="n">parsimony_value</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">p_ind</span> <span class="o">=</span> <span class="n">protein_finder</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">proteins</span><span class="p">)</span>
                <span class="n">protein_object</span> <span class="o">=</span> <span class="n">scored_proteins</span><span class="p">[</span><span class="n">p_ind</span><span class="p">]</span>
                <span class="n">lead_protein_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_object</span><span class="p">)</span>
                <span class="n">lead_protein_identifiers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_object</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">parsimony_value</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># Why are some proteins not being found when we run exclusion???</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Protein </span><span class="si">{}</span><span class="s2"> not found with protein finder...&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">proteins</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">pass</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">lead_protein_objects</span> <span class="o">=</span> <span class="n">lead_protein_objects</span>

        <span class="n">grouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_protein_groups</span><span class="p">(</span>
            <span class="n">all_scored_proteins</span><span class="o">=</span><span class="n">scored_data</span><span class="p">,</span>
            <span class="n">lead_protein_objects</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">lead_protein_objects</span><span class="p">,</span>
            <span class="n">grouping_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">grouping_type</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_swissprot_and_isoform_override</span><span class="p">(</span>
            <span class="n">scored_data</span><span class="o">=</span><span class="n">scored_data</span><span class="p">,</span>
            <span class="n">grouped_proteins</span><span class="o">=</span><span class="n">grouped_proteins</span><span class="p">,</span>
            <span class="n">override_type</span><span class="o">=</span><span class="s2">&quot;soft&quot;</span><span class="p">,</span>
            <span class="n">isoform_override</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="c1"># Run lead reassignment for the group objets and protein objects</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reassign_protein_group_leads</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reassign_protein_list_leads</span><span class="p">(</span><span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Re Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>

    <span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method performs the Parsimony inference method and uses pulp for the LP solver.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">PULP</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_pulp_grouper</span><span class="p">()</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Parsimony cannot run if lp_solver parameter value is not one of the following: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Inference</span><span class="o">.</span><span class="n">LP_SOLVERS</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Call assign shared peptides</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_assign_shared_peptides</span><span class="p">(</span><span class="n">shared_pep_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">shared_peptides</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_assign_shared_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">shared_pep_type</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Grouped Protein objects could not be found. Please run &#39;infer_proteins&#39; method of the Parsimony class&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">shared_pep_type</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ALL_SHARED_PEPTIDES</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="k">elif</span> <span class="n">shared_pep_type</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">BEST_SHARED_PEPTIDES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assigning Shared Peptides from Parsimony to the Best Scoring Protein&quot;</span><span class="p">)</span>
            <span class="n">raw_peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">prots</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span><span class="p">:</span>
                <span class="n">new_psms</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">new_raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="n">new_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="n">lead_prot</span> <span class="o">=</span> <span class="n">prots</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">psm</span> <span class="ow">in</span> <span class="n">lead_prot</span><span class="o">.</span><span class="n">psms</span><span class="p">:</span>
                    <span class="n">raw_pep</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">identifier</span>
                    <span class="n">pep</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
                    <span class="k">if</span> <span class="n">raw_pep</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">raw_peptide_tracker</span><span class="p">:</span>
                        <span class="n">new_raw_peptides</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">raw_pep</span><span class="p">)</span>
                        <span class="n">raw_peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">raw_pep</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">pep</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
                        <span class="n">new_peptides</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pep</span><span class="p">)</span>
                        <span class="n">new_psms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm</span><span class="p">)</span>
                        <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pep</span><span class="p">)</span>
                <span class="n">lead_prot</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">new_psms</span>
                <span class="n">lead_prot</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="n">new_raw_peptides</span>
                <span class="n">lead_prot</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="n">new_peptides</span>

            <span class="n">raw_peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span><span class="p">:</span>
                <span class="n">lead_prot</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">new_psms</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">new_raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="n">new_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">psm</span> <span class="ow">in</span> <span class="n">lead_prot</span><span class="o">.</span><span class="n">psms</span><span class="p">:</span>
                    <span class="n">raw_pep</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">identifier</span>
                    <span class="n">pep</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">non_flanking_peptide</span>
                    <span class="k">if</span> <span class="n">raw_pep</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">raw_peptide_tracker</span><span class="p">:</span>
                        <span class="n">new_raw_peptides</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">raw_pep</span><span class="p">)</span>
                        <span class="n">raw_peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">raw_pep</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">pep</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
                        <span class="n">new_peptides</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pep</span><span class="p">)</span>
                        <span class="n">new_psms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm</span><span class="p">)</span>
                        <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pep</span><span class="p">)</span>

                <span class="n">lead_prot</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">new_psms</span>
                <span class="n">lead_prot</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="n">new_raw_peptides</span>
                <span class="n">lead_prot</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="n">new_peptides</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Parsimony.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method of the Parsimony object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method of the Parsimony object.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lead_protein_set</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parameter_file_object</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.Parsimony.infer_proteins">
<code class="highlight language-python"><span class="n">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method performs the Parsimony inference method and uses pulp for the LP solver.</p>
<p>This method assigns the variables: <code>grouped_scored_proteins</code> and <code>protein_group_objects</code>.
These are both variables of the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a> and are
lists of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
and <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method performs the Parsimony inference method and uses pulp for the LP solver.</span>

<span class="sd">    This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">    These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">    lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">PULP</span><span class="p">:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_pulp_grouper</span><span class="p">()</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Parsimony cannot run if lp_solver parameter value is not one of the following: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Inference</span><span class="o">.</span><span class="n">LP_SOLVERS</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># Call assign shared peptides</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_assign_shared_peptides</span><span class="p">(</span><span class="n">shared_pep_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">shared_peptides</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.inference.PeptideCentric">
        <code>
PeptideCentric            (<a class="autorefs autorefs-internal" title="pyproteininference.inference.Inference" href="#pyproteininference.inference.Inference">Inference</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>PeptideCentric Inference class. This class contains methods that support the initialization of a
PeptideCentric inference method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">PeptideCentric</span><span class="p">(</span><span class="n">Inference</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    PeptideCentric Inference class. This class contains methods that support the initialization of a</span>
<span class="sd">    PeptideCentric inference method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        PeptideCentric Inference initialization method.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        Returns:</span>
<span class="sd">            object:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method performs the Peptide Centric inference method.</span>

<span class="sd">        This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">        These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">        lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">        and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get the higher or lower variable</span>
        <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Peptide Centric Method&quot;</span><span class="p">)</span>
        <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">()</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
            <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
            <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>

    <span class="k">def</span> <span class="nf">_apply_protein_group_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates the ProteinGroup objects for the peptide_centric inference based on protein groups</span>
<span class="sd">        from [._create_protein_groups][pyproteininference.inference.Inference._create_protein_groups].</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: a Dictionary that contains a list of [ProteinGroup]]pyproteininference.physical.ProteinGroup]</span>
<span class="sd">            objects (key:&quot;group_objects&quot;) and a list of grouped [Protein]]pyproteininference.physical.Protein]</span>
<span class="sd">            objects (key:&quot;grouped_protein_objects&quot;).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>

        <span class="c1"># Here we create group ID&#39;s</span>
        <span class="n">group_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">list_of_proteins_grouped</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">grouped_protein_objects</span><span class="p">:</span>
            <span class="n">protein_group</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
                <span class="p">[</span><span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">protein_group</span><span class="o">.</span><span class="n">raw_peptides</span><span class="p">)]</span>
            <span class="p">)</span>
            <span class="n">protein_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">group_id</span> <span class="o">=</span> <span class="n">group_id</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">pg</span> <span class="o">=</span> <span class="n">ProteinGroup</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Created Protein Group with ID: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">group_id</span><span class="p">)))</span>
            <span class="c1"># The following loop assigns group_id&#39;s, reviewed/unreviewed status, and number of unique peptides...</span>
            <span class="k">if</span> <span class="n">group_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">protein_group</span><span class="o">.</span><span class="n">group_identification</span><span class="p">:</span>
                <span class="n">protein_group</span><span class="o">.</span><span class="n">group_identification</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">group_id</span><span class="p">)</span>
            <span class="n">protein_group</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">protein_group</span><span class="o">.</span><span class="n">peptides</span><span class="p">)</span>
            <span class="c1"># Here append the number of unique peptides... so we can use this as secondary sorting...</span>
            <span class="n">protein_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_group</span><span class="p">)</span>
            <span class="c1"># Sorted protein_groups then becomes a list of lists... of protein objects</span>

            <span class="n">pg</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="n">protein_list</span>
            <span class="n">protein_group_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pg</span><span class="p">)</span>
            <span class="n">list_of_proteins_grouped</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">protein_group</span><span class="p">])</span>

        <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">:</span> <span class="n">list_of_proteins_grouped</span><span class="p">,</span>
            <span class="s2">&quot;group_objects&quot;</span><span class="p">:</span> <span class="n">protein_group_objects</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.PeptideCentric.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>PeptideCentric Inference initialization method.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>object</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    PeptideCentric Inference initialization method.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    Returns:</span>
<span class="sd">        object:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">_validate_scored_proteins</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scored_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">get_protein_data</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.inference.PeptideCentric.infer_proteins">
<code class="highlight language-python"><span class="n">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method performs the Peptide Centric inference method.</p>
<p>This method assigns the variables: <code>grouped_scored_proteins</code> and <code>protein_group_objects</code>.
These are both variables of the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a> and are
lists of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
and <a class="autorefs autorefs-internal" href="#pyproteininference.physical.ProteinGroup">ProteinGroup</a> objects.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/inference.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">infer_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method performs the Peptide Centric inference method.</span>

<span class="sd">    This method assigns the variables: `grouped_scored_proteins` and `protein_group_objects`.</span>
<span class="sd">    These are both variables of the [DataStore object][pyproteininference.datastore.DataStore] and are</span>
<span class="sd">    lists of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    and [ProteinGroup][pyproteininference.physical.ProteinGroup] objects.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Get the higher or lower variable</span>
    <span class="n">hl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">higher_or_lower</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Applying Group ID&#39;s for the Peptide Centric Method&quot;</span><span class="p">)</span>
    <span class="n">regrouped_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_apply_protein_group_ids</span><span class="p">()</span>

    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;grouped_protein_objects&quot;</span><span class="p">]</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">regrouped_proteins</span><span class="p">[</span><span class="s2">&quot;group_objects&quot;</span><span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting Results based on lead Protein Score&quot;</span><span class="p">)</span>
    <span class="n">grouped_protein_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_objects</span><span class="p">(</span>
        <span class="n">grouped_protein_objects</span><span class="o">=</span><span class="n">grouped_protein_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>
    <span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_group_objects</span><span class="p">(</span>
        <span class="n">protein_group_objects</span><span class="o">=</span><span class="n">protein_group_objects</span><span class="p">,</span> <span class="n">higher_or_lower</span><span class="o">=</span><span class="n">hl</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">grouped_scored_proteins</span> <span class="o">=</span> <span class="n">grouped_protein_objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_group_objects</span> <span class="o">=</span> <span class="n">protein_group_objects</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.parameters">
        <code>parameters</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter">
        <code>
ProteinInferenceParameter        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Class that handles data retrieval, storage, and validation of Protein Inference Parameters.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>yaml_param_filepath</code></td>
        <td><code>str</code></td>
        <td><p>path to properly formatted parameter file specific to Protein Inference.</p></td>
      </tr>
      <tr>
        <td><code>digest_type</code></td>
        <td><code>str</code></td>
        <td><p>String that determines that type of digestion in silico digestion for
<a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>. Typically "trypsin".</p></td>
      </tr>
      <tr>
        <td><code>export</code></td>
        <td><code>str</code></td>
        <td><p>String to indicate the export type for <a class="autorefs autorefs-internal" href="#pyproteininference.export.Export">Export object</a>.
Typically this is "psms", "peptides", or "psm_ids".</p></td>
      </tr>
      <tr>
        <td><code>fdr</code></td>
        <td><code>float</code></td>
        <td><p>Float to indicate FDR filtering.</p></td>
      </tr>
      <tr>
        <td><code>missed_cleavages</code></td>
        <td><code>int</code></td>
        <td><p>Integer to determine the number of missed cleavages in the database digestion
<a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest object</a>.</p></td>
      </tr>
      <tr>
        <td><code>picker</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether or not to run
the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.protein_picker">protein picker</a> algorithm.</p></td>
      </tr>
      <tr>
        <td><code>restrict_pep</code></td>
        <td><code>float/None</code></td>
        <td><p>Float to restrict the posterior error probability values by in the PSM input.
Used in <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data</a>.</p></td>
      </tr>
      <tr>
        <td><code>restrict_peptide_length</code></td>
        <td><code>int/None</code></td>
        <td><p>Float to restrict the peptide length values by in the PSM input.
Used in <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data</a>.</p></td>
      </tr>
      <tr>
        <td><code>restrict_q</code></td>
        <td><code>float/None</code></td>
        <td><p>Float to restrict the q values by in the PSM input.
Used in <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data</a>.</p></td>
      </tr>
      <tr>
        <td><code>restrict_custom</code></td>
        <td><code>float/None</code></td>
        <td><p>Float to restrict the custom values by in the PSM input.
Used in <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.restrict_psm_data">restrict_psm_data</a>.
Filtering depends on score_type variable. If score_type is multiplicative then values that are less than
restrict_custom are kept. If score_type is additive then values that are more than restrict_custom are kept.</p></td>
      </tr>
      <tr>
        <td><code>protein_score</code></td>
        <td><code>str</code></td>
        <td><p>String to determine the way in which Proteins are scored can be any of the SCORE_METHODS
in <a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score">Score object</a>.</p></td>
      </tr>
      <tr>
        <td><code>psm_score_type</code></td>
        <td><code>str</code></td>
        <td><p>String to determine the type of score that the PSM scores are
(Additive or Multiplicative) can be any of the SCORE_TYPES
in <a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score">Score object</a>.</p></td>
      </tr>
      <tr>
        <td><code>decoy_symbol</code></td>
        <td><code>str</code></td>
        <td><p>String to denote decoy proteins from target proteins. IE "##".</p></td>
      </tr>
      <tr>
        <td><code>isoform_symbol</code></td>
        <td><code>str</code></td>
        <td><p>String to denote isoforms from regular proteins. IE "-". Can also be None.</p></td>
      </tr>
      <tr>
        <td><code>reviewed_identifier_symbol</code></td>
        <td><code>str</code></td>
        <td><p>String to denote a "Reviewed" Protein. Typically this is: "sp|"
if using Uniprot Fasta database.</p></td>
      </tr>
      <tr>
        <td><code>inference_type</code></td>
        <td><code>str</code></td>
        <td><p>String to determine the inference procedure. Can be any value of INFERENCE_TYPES
of <a class="autorefs autorefs-internal" href="#pyproteininference.inference.Inference">Inference object</a>.</p></td>
      </tr>
      <tr>
        <td><code>tag</code></td>
        <td><code>str</code></td>
        <td><p>String to be added to output files.</p></td>
      </tr>
      <tr>
        <td><code>psm_score</code></td>
        <td><code>str</code></td>
        <td><p>String that indicates the PSM input score. The value should match the string in the
input data of the score you want to use for PSM score. This score will be used in scoring methods
    here: <a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score">Score object</a>.</p></td>
      </tr>
      <tr>
        <td><code>grouping_type</code></td>
        <td><code>str/None</code></td>
        <td><p>String to determine the grouping procedure. Can be any value of
GROUPING_TYPES of <a class="autorefs autorefs-internal" href="#pyproteininference.inference.Inference">Inference object</a>.</p></td>
      </tr>
      <tr>
        <td><code>max_identifiers_peptide_centric</code></td>
        <td><code>int</code></td>
        <td><p>Maximum number of identifiers to assign to a group when
running peptide_centric inference. Typically this is 10 or 5.</p></td>
      </tr>
      <tr>
        <td><code>lp_solver</code></td>
        <td><code>str/None</code></td>
        <td><p>The LP solver to use if inference_type="Parsimony".
Can be any value in LP_SOLVERS in the <a class="autorefs autorefs-internal" href="#pyproteininference.inference.Inference">Inference object</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ProteinInferenceParameter</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class that handles data retrieval, storage, and validation of Protein Inference Parameters.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        yaml_param_filepath (str): path to properly formatted parameter file specific to Protein Inference.</span>
<span class="sd">        digest_type (str): String that determines that type of digestion in silico digestion for</span>
<span class="sd">            [Digest object][pyproteininference.in_silico_digest.Digest]. Typically &quot;trypsin&quot;.</span>
<span class="sd">        export (str): String to indicate the export type for [Export object][pyproteininference.export.Export].</span>
<span class="sd">            Typically this is &quot;psms&quot;, &quot;peptides&quot;, or &quot;psm_ids&quot;.</span>
<span class="sd">        fdr (float): Float to indicate FDR filtering.</span>
<span class="sd">        missed_cleavages (int): Integer to determine the number of missed cleavages in the database digestion</span>
<span class="sd">            [Digest object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        picker (bool): True/False on whether or not to run</span>
<span class="sd">            the [protein picker][pyproteininference.datastore.DataStore.protein_picker] algorithm.</span>
<span class="sd">        restrict_pep (float/None): Float to restrict the posterior error probability values by in the PSM input.</span>
<span class="sd">            Used in [restrict_psm_data][pyproteininference.datastore.DataStore.restrict_psm_data].</span>
<span class="sd">        restrict_peptide_length (int/None): Float to restrict the peptide length values by in the PSM input.</span>
<span class="sd">            Used in [restrict_psm_data][pyproteininference.datastore.DataStore.restrict_psm_data].</span>
<span class="sd">        restrict_q (float/None): Float to restrict the q values by in the PSM input.</span>
<span class="sd">            Used in [restrict_psm_data][pyproteininference.datastore.DataStore.restrict_psm_data].</span>
<span class="sd">        restrict_custom (float/None): Float to restrict the custom values by in the PSM input.</span>
<span class="sd">            Used in [restrict_psm_data][pyproteininference.datastore.DataStore.restrict_psm_data].</span>
<span class="sd">            Filtering depends on score_type variable. If score_type is multiplicative then values that are less than</span>
<span class="sd">            restrict_custom are kept. If score_type is additive then values that are more than restrict_custom are kept.</span>
<span class="sd">        protein_score (str): String to determine the way in which Proteins are scored can be any of the SCORE_METHODS</span>
<span class="sd">            in [Score object][pyproteininference.scoring.Score].</span>
<span class="sd">        psm_score_type (str): String to determine the type of score that the PSM scores are</span>
<span class="sd">            (Additive or Multiplicative) can be any of the SCORE_TYPES</span>
<span class="sd">            in [Score object][pyproteininference.scoring.Score].</span>
<span class="sd">        decoy_symbol (str): String to denote decoy proteins from target proteins. IE &quot;##&quot;.</span>
<span class="sd">        isoform_symbol (str): String to denote isoforms from regular proteins. IE &quot;-&quot;. Can also be None.</span>
<span class="sd">        reviewed_identifier_symbol (str): String to denote a &quot;Reviewed&quot; Protein. Typically this is: &quot;sp|&quot;</span>
<span class="sd">            if using Uniprot Fasta database.</span>
<span class="sd">        inference_type (str): String to determine the inference procedure. Can be any value of INFERENCE_TYPES</span>
<span class="sd">            of [Inference object][pyproteininference.inference.Inference].</span>
<span class="sd">        tag (str): String to be added to output files.</span>
<span class="sd">        psm_score (str): String that indicates the PSM input score. The value should match the string in the</span>
<span class="sd">            input data of the score you want to use for PSM score. This score will be used in scoring methods</span>
<span class="sd">                here: [Score object][pyproteininference.scoring.Score].</span>
<span class="sd">        grouping_type (str/None): String to determine the grouping procedure. Can be any value of</span>
<span class="sd">            GROUPING_TYPES of [Inference object][pyproteininference.inference.Inference].</span>
<span class="sd">        max_identifiers_peptide_centric (int): Maximum number of identifiers to assign to a group when</span>
<span class="sd">            running peptide_centric inference. Typically this is 10 or 5.</span>
<span class="sd">        lp_solver (str/None): The LP solver to use if inference_type=&quot;Parsimony&quot;.</span>
<span class="sd">            Can be any value in LP_SOLVERS in the [Inference object][pyproteininference.inference.Inference].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">PARENT_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;parameters&quot;</span>

    <span class="n">GENERAL_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;general&quot;</span>
    <span class="n">DATA_RESTRICTION_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;data_restriction&quot;</span>
    <span class="n">SCORE_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;score&quot;</span>
    <span class="n">IDENTIFIERS_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;identifiers&quot;</span>
    <span class="n">INFERENCE_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;inference&quot;</span>
    <span class="n">DIGEST_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;digest&quot;</span>
    <span class="n">PARSIMONY_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;parsimony&quot;</span>
    <span class="n">PEPTIDE_CENTRIC_PARAMETER_KEY</span> <span class="o">=</span> <span class="s2">&quot;peptide_centric&quot;</span>

    <span class="n">PARAMETER_MAIN_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">GENERAL_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">SCORE_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">DIGEST_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">,</span>
        <span class="n">PEPTIDE_CENTRIC_PARAMETER_KEY</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">EXPORT_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;export&quot;</span>
    <span class="n">FDR_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;fdr&quot;</span>
    <span class="n">PICKER_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;picker&quot;</span>
    <span class="n">TAG_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;tag&quot;</span>

    <span class="n">GENERAL_PARAMETER_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">EXPORT_PARAMETER</span><span class="p">,</span>
        <span class="n">FDR_PARAMETER</span><span class="p">,</span>
        <span class="n">PICKER_PARAMETER</span><span class="p">,</span>
        <span class="n">TAG_PARAMETER</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">PEP_RESTRICT_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;pep_restriction&quot;</span>
    <span class="n">PEPTIDE_LENGTH_RESTRICT_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;peptide_length_restriction&quot;</span>
    <span class="n">Q_VALUE_RESTRICT_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;q_value_restriction&quot;</span>
    <span class="n">CUSTOM_RESTRICT_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;custom_restriction&quot;</span>

    <span class="n">DATA_RESTRICTION_PARAMETER_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">PEP_RESTRICT_PARAMETER</span><span class="p">,</span>
        <span class="n">PEPTIDE_LENGTH_RESTRICT_PARAMETER</span><span class="p">,</span>
        <span class="n">Q_VALUE_RESTRICT_PARAMETER</span><span class="p">,</span>
        <span class="n">CUSTOM_RESTRICT_PARAMETER</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">PROTEIN_SCORE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;protein_score&quot;</span>
    <span class="n">PSM_SCORE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;psm_score&quot;</span>
    <span class="n">PSM_SCORE_TYPE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;psm_score_type&quot;</span>

    <span class="n">SCORE_PARAMETER_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">PROTEIN_SCORE_PARAMETER</span><span class="p">,</span>
        <span class="n">PSM_SCORE_PARAMETER</span><span class="p">,</span>
        <span class="n">PSM_SCORE_TYPE_PARAMETER</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">DECOY_SYMBOL_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;decoy_symbol&quot;</span>
    <span class="n">ISOFORM_SYMBOL_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;isoform_symbol&quot;</span>
    <span class="n">REVIEWED_IDENTIFIER_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;reviewed_identifier_symbol&quot;</span>

    <span class="n">IDENTIFIER_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">DECOY_SYMBOL_PARAMETER</span><span class="p">,</span>
        <span class="n">ISOFORM_SYMBOL_PARAMETER</span><span class="p">,</span>
        <span class="n">REVIEWED_IDENTIFIER_PARAMETER</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">INFERENCE_TYPE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;inference_type&quot;</span>
    <span class="n">GROUPING_TYPE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;grouping_type&quot;</span>

    <span class="n">INFERENCE_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span><span class="n">INFERENCE_TYPE_PARAMETER</span><span class="p">,</span> <span class="n">GROUPING_TYPE_PARAMETER</span><span class="p">}</span>

    <span class="n">DIGEST_TYPE_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;digest_type&quot;</span>
    <span class="n">MISSED_CLEAV_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;missed_cleavages&quot;</span>

    <span class="n">DIGEST_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span><span class="n">DIGEST_TYPE_PARAMETER</span><span class="p">,</span> <span class="n">MISSED_CLEAV_PARAMETER</span><span class="p">}</span>

    <span class="n">LP_SOLVER_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;lp_solver&quot;</span>
    <span class="n">SHARED_PEPTIDES_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;shared_peptides&quot;</span>

    <span class="n">PARSIMONY_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">LP_SOLVER_PARAMETER</span><span class="p">,</span>
        <span class="n">SHARED_PEPTIDES_PARAMETER</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">MAX_IDENTIFIERS_PARAMETER</span> <span class="o">=</span> <span class="s2">&quot;max_identifiers&quot;</span>

    <span class="n">PEPTIDE_CENTRIC_SUB_KEYS</span> <span class="o">=</span> <span class="p">{</span><span class="n">MAX_IDENTIFIERS_PARAMETER</span><span class="p">}</span>

    <span class="n">DEFAULT_DIGEST_TYPE</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
    <span class="n">DEFAULT_EXPORT</span> <span class="o">=</span> <span class="s2">&quot;peptides&quot;</span>
    <span class="n">DEFAULT_FDR</span> <span class="o">=</span> <span class="mf">0.01</span>
    <span class="n">DEFAULT_MISSED_CLEAVAGES</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="n">DEFAULT_PICKER</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">DEFAULT_RESTRICT_PEP</span> <span class="o">=</span> <span class="mf">0.9</span>
    <span class="n">DEFAULT_RESTRICT_PEPTIDE_LENGTH</span> <span class="o">=</span> <span class="mi">7</span>
    <span class="n">DEFAULT_RESTRICT_Q</span> <span class="o">=</span> <span class="mf">0.005</span>
    <span class="n">DEFAULT_RESTRICT_CUSTOM</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span>
    <span class="n">DEFAULT_PROTEIN_SCORE</span> <span class="o">=</span> <span class="s2">&quot;multiplicative_log&quot;</span>
    <span class="n">DEFAULT_PSM_SCORE</span> <span class="o">=</span> <span class="s2">&quot;posterior_error_prob&quot;</span>
    <span class="n">DEFAULT_DECOY_SYMBOL</span> <span class="o">=</span> <span class="s2">&quot;##&quot;</span>
    <span class="n">DEFAULT_ISOFORM_SYMBOL</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
    <span class="n">DEFAULT_REVIEWED_IDENTIFIER_SYMBOL</span> <span class="o">=</span> <span class="s2">&quot;sp|&quot;</span>
    <span class="n">DEFAULT_INFERENCE_TYPE</span> <span class="o">=</span> <span class="s2">&quot;peptide_centric&quot;</span>
    <span class="n">DEFAULT_TAG</span> <span class="o">=</span> <span class="s2">&quot;py_protein_inference&quot;</span>
    <span class="n">DEFAULT_PSM_SCORE_TYPE</span> <span class="o">=</span> <span class="s2">&quot;multiplicative&quot;</span>
    <span class="n">DEFAULT_GROUPING_TYPE</span> <span class="o">=</span> <span class="s2">&quot;shared_peptides&quot;</span>
    <span class="n">DEFAULT_MAX_IDENTIFIERS_PEPTIDE_CENTRIC</span> <span class="o">=</span> <span class="mi">5</span>
    <span class="n">DEFAULT_LP_SOLVER</span> <span class="o">=</span> <span class="s2">&quot;pulp&quot;</span>
    <span class="n">DEFAULT_SHARED_PEPTIDES</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">yaml_param_filepath</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Class to store Protein Inference parameter information as an object.</span>

<span class="sd">        Args:</span>
<span class="sd">            yaml_param_filepath (str): path to properly formatted parameter file specific to Protein Inference.</span>
<span class="sd">            validate (bool): True/False on whether to validate the parameter file of interest.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pyproteininference.parameters.ProteinInferenceParameter(</span>
<span class="sd">            &gt;&gt;&gt;     yaml_param_filepath = &quot;/path/to/pyproteininference_params.yaml&quot;, validate=True</span>
<span class="sd">            &gt;&gt;&gt; )</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span> <span class="o">=</span> <span class="n">yaml_param_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DIGEST_TYPE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">export</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_EXPORT</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fdr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_FDR</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MISSED_CLEAVAGES</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PICKER</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEP</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEPTIDE_LENGTH</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_Q</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_CUSTOM</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PROTEIN_SCORE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE_TYPE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DECOY_SYMBOL</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_ISOFORM_SYMBOL</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_REVIEWED_IDENTIFIER_SYMBOL</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_INFERENCE_TYPE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_TAG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_GROUPING_TYPE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MAX_IDENTIFIERS_PEPTIDE_CENTRIC</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_LP_SOLVER</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_SHARED_PEPTIDES</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">validate</span> <span class="o">=</span> <span class="n">validate</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">convert_to_object</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">validate</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">validate_parameters</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_none_parameters</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">convert_to_object</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Function that takes a Protein Inference parameter file and converts it into a ProteinInferenceParameter object</span>
<span class="sd">        by assigning all Attributes of the ProteinInferenceParameter object.</span>

<span class="sd">        If no parameter filepath is supplied the parameter object will be loaded with default params.</span>

<span class="sd">        This function gets ran in the initialization of the ProteinInferenceParameter object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
                <span class="n">yaml_params</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="n">Loader</span><span class="o">=</span><span class="n">yaml</span><span class="o">.</span><span class="n">Loader</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_TYPE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;digest_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DIGEST_TYPE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">export</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;export set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_EXPORT</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">fdr</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">FDR_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;fdr set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_FDR</span><span class="p">))</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">MISSED_CLEAV_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;missed_cleavages set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MISSED_CLEAVAGES</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">picker</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PICKER_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;picker set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PICKER</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">PEP_RESTRICT_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_pep set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEP</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span>
                <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_LENGTH_RESTRICT_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;restrict_peptide_length set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEPTIDE_LENGTH</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE_RESTRICT_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_q set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_Q</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">CUSTOM_RESTRICT_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_custom set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_CUSTOM</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">PROTEIN_SCORE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;protein_score set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PROTEIN_SCORE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">PSM_SCORE_TYPE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;psm_score_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE_TYPE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">DECOY_SYMBOL_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;decoy_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DECOY_SYMBOL</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">ISOFORM_SYMBOL_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;isoform_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_ISOFORM_SYMBOL</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span>
                <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">REVIEWED_IDENTIFIER_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;reviewed_identifier_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_REVIEWED_IDENTIFIER_SYMBOL</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_TYPE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;inference_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_INFERENCE_TYPE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">tag</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">TAG_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;tag set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_TAG</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">PSM_SCORE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;psm_score set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">GROUPING_TYPE_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;grouping_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_GROUPING_TYPE</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC_PARAMETER_KEY</span>
                <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_IDENTIFIERS_PARAMETER</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;max_identifiers_peptide_centric set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MAX_IDENTIFIERS_PEPTIDE_CENTRIC</span>
                    <span class="p">)</span>
                <span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">LP_SOLVER_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;lp_solver set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_LP_SOLVER</span><span class="p">))</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Do try except here to make old param files backwards compatible</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">][</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">SHARED_PEPTIDES_PARAMETER</span>
                <span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;shared_peptides set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_SHARED_PEPTIDES</span><span class="p">))</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Yaml parameter file not found, all parameters set to default&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">validate_parameters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class method to validate all parameters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None:</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Run all of the parameter validations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_digest_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_export_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_floats</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_bools</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_method</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_combination</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_inference_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_grouping_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_max_id</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_lp_solver</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_identifiers</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_parsimony_shared_peptides</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_validate_digest_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the digest type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Make sure we have a valid digest type</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="ow">in</span> <span class="n">PyteomicsDigest</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using digest type &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Digest Type &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following enyzme digestions: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">PyteomicsDigest</span><span class="o">.</span><span class="n">LIST_OF_DIGEST_TYPES</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_export_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the export type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Make sure we have a valid export type</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">export</span> <span class="ow">in</span> <span class="n">Export</span><span class="o">.</span><span class="n">EXPORT_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Export type &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">export</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Export Type &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following export types: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">export</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Export</span><span class="o">.</span><span class="n">EXPORT_TYPES</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">_validate_floats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate floats.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Validate that FDR, cleavages, and restrict values are all floats and or ints if they need to be</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fdr</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;FDR Input </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fdr</span><span class="p">))</span>

        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;FDR must be a decimal between 0 and 1, FDR provided: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fdr</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;PEP restriction </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span><span class="p">))</span>

        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Not restrict by PEP Value&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;PEP restriction must be a decimal between 0 and 1, PEP restriction provided: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span>
                    <span class="p">)</span>
                <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Q Value restriction </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span><span class="p">))</span>

        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Not restrict by Q Value&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Q Value restriction must be a decimal between 0 and 1, Q Value restriction provided: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span>
                    <span class="p">)</span>
                <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Missed Cleavages selected: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Missed Cleavages must be an integer, Provided Missed Cleavages value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Peptide Length Restriction: Len </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Not Restricting by Peptide Length&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Peptide Length Restriction must be an integer, &quot;</span>
                    <span class="s2">&quot;Provided Peptide Length Restriction value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">)</span>
                <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Custom restriction </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="ow">or</span> <span class="ne">TypeError</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Not Restricting by Custom Value&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Custom restriction must be a number, Custom restriction provided: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_bools</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the bools.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Make sure picker is a bool</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">picker</span><span class="p">)</span> <span class="o">==</span> <span class="nb">bool</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">picker</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parameters loaded to run Picker&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Parameters loaded to NOT run Picker&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Picker Variable must be set to True or False, Picker Variable provided: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">picker</span><span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_score_method</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the score method.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Make sure we have the score method defined in code to use...</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="ow">in</span> <span class="n">Score</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Score Method &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Score Method &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, &quot;</span>
                <span class="s2">&quot;please use one of the following Score Methods: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Score</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_score_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the score type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Make sure score type is multiplicative or additive</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="ow">in</span> <span class="n">Score</span><span class="o">.</span><span class="n">SCORE_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Score Type &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Score Type &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, &quot;</span>
                <span class="s2">&quot;please use one of the following Score Types: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Score</span><span class="o">.</span><span class="n">SCORE_TYPES</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_score_combination</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate combination of score method and score type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check to see if combination of score (column), method(multiplicative log, additive),</span>
        <span class="c1"># and score type (multiplicative/additive) is possible...</span>
        <span class="c1"># This will be super custom</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE_SCORE_TYPE</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">!=</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;If Score type is &#39;additive&#39; (Higher PSM score is better) then you must use the &#39;additive&#39; score method&quot;</span>
            <span class="p">)</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;If Score type is &#39;multiplicative&#39; (Lower PSM score is better) &quot;</span>
                <span class="s2">&quot;then you must NOT use the &#39;additive&#39; score method please &quot;</span>
                <span class="s2">&quot;select one of the following score methods: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">Score</span><span class="o">.</span><span class="n">SCORE_METHODS</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;additive&quot;</span><span class="p">])</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Combination of Score Type: &#39;</span><span class="si">{}</span><span class="s2">&#39; and Score Method: &#39;</span><span class="si">{}</span><span class="s2">&#39; is Ok&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_inference_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the inference type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if its parsimony, exclusion, inclusion, none</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span> <span class="ow">in</span> <span class="n">Inference</span><span class="o">.</span><span class="n">INFERENCE_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using inference type &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Inferece Type &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following Inferece Types: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Inference</span><span class="o">.</span><span class="n">INFERENCE_TYPES</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_grouping_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the grouping type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if its parsimony, exclusion, inclusion, none</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="ow">in</span> <span class="n">Inference</span><span class="o">.</span><span class="n">GROUPING_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Grouping type &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Grouping type: None&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>

                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Grouping Type &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following Grouping Types: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">GROUPING_TYPES</span>
                    <span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_max_id</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the max peptide centric id.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if max_identifiers_peptide_centric param is an INT</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Max Number of Indentifiers for Peptide Centric Inference: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Max Number of Indentifiers for Peptide Centric Inference must be an integer, &quot;</span>
                <span class="s2">&quot;provided value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span><span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_lp_solver</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the lp solver.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if its pulp or None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="ow">in</span> <span class="n">Inference</span><span class="o">.</span><span class="n">LP_SOLVERS</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using LP Solver &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Setting LP Solver to None&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;LP Solver &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following LP Solvers: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Inference</span><span class="o">.</span><span class="n">LP_SOLVERS</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_parsimony_shared_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the shared peptides parameter.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if its all, best, or none</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="ow">in</span> <span class="n">Inference</span><span class="o">.</span><span class="n">SHARED_PEPTIDE_TYPES</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using Shared Peptide types &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Setting Shared Peptide type to None&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Shared Peptide types &#39;</span><span class="si">{}</span><span class="s2">&#39; not supported, please use one of the following &quot;</span>
                    <span class="s2">&quot;Shared Peptide types: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span><span class="p">,</span> <span class="n">Inference</span><span class="o">.</span><span class="n">SHARED_PEPTIDE_TYPES</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate the decoy symbol, isoform symbol,</span>
<span class="sd">        and reviewed identifier symbol.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Decoy Symbol set to: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Decoy Symbol must be a string, provided value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">))</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Isoform Symbol set to: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">))</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Isoform Symbol set to None&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Isoform Symbol set to None&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Isoform Symbol must be a string, provided value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span><span class="p">))</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reviewed Identifier Symbol set to: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">))</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reviewed Identifier Symbol set to None&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reviewed Identifier Symbol set to None&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Reviewed Identifier Symbol must be a string, provided value: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_parameter_shape</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">yaml_params</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to validate shape of the parameter file by checking to make sure</span>
<span class="sd">         that all necessary main parameter fields are defined.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span> <span class="ow">in</span> <span class="n">yaml_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Main Parameter Key is Present&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Key </span><span class="si">{}</span><span class="s2"> needs to be defined as the outermost parameter group&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">PARAMETER_MAIN_KEYS</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">]):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;All Sub Parameter Keys Present&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;All of the following values: </span><span class="si">{}</span><span class="s2">. Need to be Sub Parameters in the Yaml Parameter file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">PARAMETER_MAIN_KEYS</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">general_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">gkey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gkey</span> <span class="ow">in</span> <span class="n">general_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;General Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the general parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gkey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;general&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">data_res_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">drkey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">drkey</span> <span class="ow">in</span> <span class="n">data_res_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Data Restriction Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the data_restriction parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">drkey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;data_restriction&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">score_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">skey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">skey</span> <span class="ow">in</span> <span class="n">score_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Score Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the score parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">skey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;score&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">id_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">ikey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIER_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ikey</span> <span class="ow">in</span> <span class="n">id_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Identifiers Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the identifiers parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ikey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;identifiers&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">inf_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">infkey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">infkey</span> <span class="ow">in</span> <span class="n">inf_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Inference Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the inference parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">infkey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;inference&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">digest_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">dkey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">dkey</span> <span class="ow">in</span> <span class="n">digest_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Digest Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the digest parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dkey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;digest&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">parsimony_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">pkey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pkey</span> <span class="ow">in</span> <span class="n">parsimony_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Parsimony Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the parsimony parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pkey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;parsimony&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">pep_cen_params</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC_PARAMETER_KEY</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">pckey</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC_SUB_KEYS</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pckey</span> <span class="ow">in</span> <span class="n">pep_cen_params</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Peptide Centric Sub Parameter &#39;</span><span class="si">{}</span><span class="s2">&#39; is not found in the parameter file. &quot;</span>
                        <span class="s2">&quot;Please add it as a sub parameter of the peptide_centric parameter field&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pckey</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;peptide_centric&#39; sub Parameter not defined in the parameter file&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">override_q_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        ProteinInferenceParameter method to override restrict_q if the input data does not contain q values.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data_has_q</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_q</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">data_has_q</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No Q values found in the input data, overriding parameters to not filter on Q value&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">override_pep_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        ProteinInferenceParameter method to override restrict_pep if the input data does not contain pep values.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data_has_pep</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_pep</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">data_has_pep</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;No Pep values found in the input data, overriding parameters to not filter on Pep value&quot;</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">override_custom_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        ProteinInferenceParameter method to override restrict_custom if</span>
<span class="sd">        the input data does not contain custom score values.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data_has_custom</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_custom</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">data_has_custom</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;No Custom values found in the input data, overriding parameters to not filter on Custom value&quot;</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">fix_parameters_from_datastore</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        ProteinInferenceParameter method to override restriction values in the</span>
<span class="sd">        parameter file if those scores do not exist in the input files.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">override_q_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">override_pep_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">override_custom_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_fix_none_parameters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to fix parameters that have been defined as None.</span>
<span class="sd">        These get read in as strings with YAML reader and need to be converted to None type.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_grouping_type</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_lp_solver</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_shared_peptides</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_fix_grouping_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to override grouping type for None value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_fix_lp_solver</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to override lp_solver for None value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_fix_shared_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal ProteinInferenceParameter method to override shared_peptides for None value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">





































































  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">yaml_param_filepath</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Class to store Protein Inference parameter information as an object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>yaml_param_filepath</strong> (<code>str</code>)  path to properly formatted parameter file specific to Protein Inference.</p></li>
            <li><p><strong>validate</strong> (<code>bool</code>)  True/False on whether to validate the parameter file of interest.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ProteinInferenceParameter</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">yaml_param_filepath</span> <span class="o">=</span> <span class="s2">&quot;/path/to/pyproteininference_params.yaml&quot;</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">yaml_param_filepath</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Class to store Protein Inference parameter information as an object.</span>

<span class="sd">    Args:</span>
<span class="sd">        yaml_param_filepath (str): path to properly formatted parameter file specific to Protein Inference.</span>
<span class="sd">        validate (bool): True/False on whether to validate the parameter file of interest.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pyproteininference.parameters.ProteinInferenceParameter(</span>
<span class="sd">        &gt;&gt;&gt;     yaml_param_filepath = &quot;/path/to/pyproteininference_params.yaml&quot;, validate=True</span>
<span class="sd">        &gt;&gt;&gt; )</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span> <span class="o">=</span> <span class="n">yaml_param_filepath</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DIGEST_TYPE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">export</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_EXPORT</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">fdr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_FDR</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MISSED_CLEAVAGES</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PICKER</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEP</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEPTIDE_LENGTH</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_Q</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_CUSTOM</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PROTEIN_SCORE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE_TYPE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DECOY_SYMBOL</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_ISOFORM_SYMBOL</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_REVIEWED_IDENTIFIER_SYMBOL</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_INFERENCE_TYPE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">tag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_TAG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_GROUPING_TYPE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MAX_IDENTIFIERS_PEPTIDE_CENTRIC</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_LP_SOLVER</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_SHARED_PEPTIDES</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">validate</span> <span class="o">=</span> <span class="n">validate</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">convert_to_object</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">validate</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">validate_parameters</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_fix_none_parameters</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.convert_to_object">
<code class="highlight language-python"><span class="n">convert_to_object</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Function that takes a Protein Inference parameter file and converts it into a ProteinInferenceParameter object
by assigning all Attributes of the ProteinInferenceParameter object.</p>
<p>If no parameter filepath is supplied the parameter object will be loaded with default params.</p>
<p>This function gets ran in the initialization of the ProteinInferenceParameter object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">convert_to_object</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function that takes a Protein Inference parameter file and converts it into a ProteinInferenceParameter object</span>
<span class="sd">    by assigning all Attributes of the ProteinInferenceParameter object.</span>

<span class="sd">    If no parameter filepath is supplied the parameter object will be loaded with default params.</span>

<span class="sd">    This function gets ran in the initialization of the ProteinInferenceParameter object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yaml_param_filepath</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
            <span class="n">yaml_params</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="n">Loader</span><span class="o">=</span><span class="n">yaml</span><span class="o">.</span><span class="n">Loader</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">digest_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_TYPE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;digest_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DIGEST_TYPE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">export</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">EXPORT_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;export set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_EXPORT</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">fdr</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">FDR_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;fdr set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_FDR</span><span class="p">))</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">missed_cleavages</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DIGEST_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">MISSED_CLEAV_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;missed_cleavages set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MISSED_CLEAVAGES</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">picker</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PICKER_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;picker set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PICKER</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">PEP_RESTRICT_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_pep set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEP</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_peptide_length</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span>
            <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_LENGTH_RESTRICT_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;restrict_peptide_length set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_PEPTIDE_LENGTH</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE_RESTRICT_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_q set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_Q</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">DATA_RESTRICTION_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">CUSTOM_RESTRICT_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;restrict_custom set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_RESTRICT_CUSTOM</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">PROTEIN_SCORE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;protein_score set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PROTEIN_SCORE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">PSM_SCORE_TYPE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;psm_score_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE_TYPE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">decoy_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">DECOY_SYMBOL_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;decoy_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_DECOY_SYMBOL</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">isoform_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ISOFORM_SYMBOL_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;isoform_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_ISOFORM_SYMBOL</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">IDENTIFIERS_PARAMETER_KEY</span>
            <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">REVIEWED_IDENTIFIER_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;reviewed_identifier_symbol set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_REVIEWED_IDENTIFIER_SYMBOL</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_TYPE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;inference_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_INFERENCE_TYPE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tag</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">GENERAL_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">TAG_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;tag set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_TAG</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">psm_score</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">PSM_SCORE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;psm_score set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_PSM_SCORE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">grouping_type</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">INFERENCE_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">GROUPING_TYPE_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;grouping_type set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_GROUPING_TYPE</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_identifiers_peptide_centric</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_CENTRIC_PARAMETER_KEY</span>
            <span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_IDENTIFIERS_PARAMETER</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;max_identifiers_peptide_centric set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_MAX_IDENTIFIERS_PEPTIDE_CENTRIC</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lp_solver</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">LP_SOLVER_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;lp_solver set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_LP_SOLVER</span><span class="p">))</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Do try except here to make old param files backwards compatible</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">shared_peptides</span> <span class="o">=</span> <span class="n">yaml_params</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PARENT_PARAMETER_KEY</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">PARSIMONY_PARAMETER_KEY</span><span class="p">][</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">SHARED_PEPTIDES_PARAMETER</span>
            <span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;shared_peptides set to default of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DEFAULT_SHARED_PEPTIDES</span><span class="p">))</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Yaml parameter file not found, all parameters set to default&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.fix_parameters_from_datastore">
<code class="highlight language-python"><span class="n">fix_parameters_from_datastore</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>ProteinInferenceParameter method to override restriction values in the
parameter file if those scores do not exist in the input files.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">fix_parameters_from_datastore</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    ProteinInferenceParameter method to override restriction values in the</span>
<span class="sd">    parameter file if those scores do not exist in the input files.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">override_q_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">override_pep_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">override_custom_restrict</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.override_custom_restrict">
<code class="highlight language-python"><span class="n">override_custom_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>ProteinInferenceParameter method to override restrict_custom if
the input data does not contain custom score values.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">override_custom_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    ProteinInferenceParameter method to override restrict_custom if</span>
<span class="sd">    the input data does not contain custom score values.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_has_custom</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_custom</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">data_has_custom</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;No Custom values found in the input data, overriding parameters to not filter on Custom value&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_custom</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.override_pep_restrict">
<code class="highlight language-python"><span class="n">override_pep_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>ProteinInferenceParameter method to override restrict_pep if the input data does not contain pep values.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">override_pep_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    ProteinInferenceParameter method to override restrict_pep if the input data does not contain pep values.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_has_pep</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_pep</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">data_has_pep</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;No Pep values found in the input data, overriding parameters to not filter on Pep value&quot;</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_pep</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.override_q_restrict">
<code class="highlight language-python"><span class="n">override_q_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>ProteinInferenceParameter method to override restrict_q if the input data does not contain q values.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">override_q_restrict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    ProteinInferenceParameter method to override restrict_q if the input data does not contain q values.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_has_q</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">input_has_q</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">data_has_q</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No Q values found in the input data, overriding parameters to not filter on Q value&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">restrict_q</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.parameters.ProteinInferenceParameter.validate_parameters">
<code class="highlight language-python"><span class="n">validate_parameters</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Class method to validate all parameters.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>None</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/parameters.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">validate_parameters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class method to validate all parameters.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Run all of the parameter validations</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_digest_type</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_export_type</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_floats</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_bools</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_type</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_method</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_score_combination</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_inference_type</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_grouping_type</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_max_id</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_lp_solver</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_identifiers</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_parsimony_shared_peptides</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.physical">
        <code>physical</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">







  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.physical.Protein">
        <code>
Protein        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class is a representation of a Protein that stores characteristics/attributes of a protein for the
    entire analysis.
We use <strong>slots</strong> to predefine the attributes the Protein Object can have.
This is done to speed up runtime of the PI algorithm.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>identifier</code></td>
        <td><code>str</code></td>
        <td><p>String identifier for the Protein object.</p></td>
      </tr>
      <tr>
        <td><code>score</code></td>
        <td><code>float</code></td>
        <td><p>Float that represents the protein score as output from
<a class="autorefs autorefs-internal" href="#pyproteininference.scoring.Score">Score object</a> methods.</p></td>
      </tr>
      <tr>
        <td><code>psms</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p></td>
      </tr>
      <tr>
        <td><code>group_identification</code></td>
        <td><code>set</code></td>
        <td><p>Set of group Identifiers that the protein belongs to (int).</p></td>
      </tr>
      <tr>
        <td><code>reviewed</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on if the identifier is reviewed.</p></td>
      </tr>
      <tr>
        <td><code>unreviewed</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on if the identifier is reviewed.</p></td>
      </tr>
      <tr>
        <td><code>peptides</code></td>
        <td><code>list</code></td>
        <td><p>List of non flanking peptide sequences.</p></td>
      </tr>
      <tr>
        <td><code>peptide_scores</code></td>
        <td><code>list</code></td>
        <td><p>List of Psm scores associated with the protein.</p></td>
      </tr>
      <tr>
        <td><code>picked</code></td>
        <td><code>bool</code></td>
        <td><p>True/False if the protein passes the picker algo. True if passes. False if does not pass.</p></td>
      </tr>
      <tr>
        <td><code>num_peptides</code></td>
        <td><code>int</code></td>
        <td><p>Number of peptides that map to the given Protein.</p></td>
      </tr>
      <tr>
        <td><code>unique_peptides</code></td>
        <td><code>list</code></td>
        <td><p>List of peptide strings that are unique to this protein across the analysis.</p></td>
      </tr>
      <tr>
        <td><code>num_unique_peptides</code></td>
        <td><code>int</code></td>
        <td><p>Number of unique peptides.</p></td>
      </tr>
      <tr>
        <td><code>raw_peptides</code></td>
        <td><code>list</code></td>
        <td><p>List of raw peptides. Includes flanking AA and Mods.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Protein</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class is a representation of a Protein that stores characteristics/attributes of a protein for the</span>
<span class="sd">        entire analysis.</span>
<span class="sd">    We use __slots__ to predefine the attributes the Protein Object can have.</span>
<span class="sd">    This is done to speed up runtime of the PI algorithm.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        identifier (str): String identifier for the Protein object.</span>
<span class="sd">        score (float): Float that represents the protein score as output from</span>
<span class="sd">            [Score object][pyproteininference.scoring.Score] methods.</span>
<span class="sd">        psms (list): List of [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">        group_identification (set): Set of group Identifiers that the protein belongs to (int).</span>
<span class="sd">        reviewed (bool): True/False on if the identifier is reviewed.</span>
<span class="sd">        unreviewed (bool): True/False on if the identifier is reviewed.</span>
<span class="sd">        peptides (list): List of non flanking peptide sequences.</span>
<span class="sd">        peptide_scores (list): List of Psm scores associated with the protein.</span>
<span class="sd">        picked (bool): True/False if the protein passes the picker algo. True if passes. False if does not pass.</span>
<span class="sd">        num_peptides (int): Number of peptides that map to the given Protein.</span>
<span class="sd">        unique_peptides (list): List of peptide strings that are unique to this protein across the analysis.</span>
<span class="sd">        num_unique_peptides (int): Number of unique peptides.</span>
<span class="sd">        raw_peptides (list): List of raw peptides. Includes flanking AA and Mods.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;identifier&quot;</span><span class="p">,</span>
        <span class="s2">&quot;score&quot;</span><span class="p">,</span>
        <span class="s2">&quot;psms&quot;</span><span class="p">,</span>
        <span class="s2">&quot;group_identification&quot;</span><span class="p">,</span>
        <span class="s2">&quot;reviewed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;unreviewed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;peptides&quot;</span><span class="p">,</span>
        <span class="s2">&quot;peptide_scores&quot;</span><span class="p">,</span>
        <span class="s2">&quot;picked&quot;</span><span class="p">,</span>
        <span class="s2">&quot;num_peptides&quot;</span><span class="p">,</span>
        <span class="s2">&quot;unique_peptides&quot;</span><span class="p">,</span>
        <span class="s2">&quot;num_unique_peptides&quot;</span><span class="p">,</span>
        <span class="s2">&quot;raw_peptides&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method for Protein object.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifier (str): String identifier for the Protein object.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; protein = pyproteininference.physical.Protein(identifier = &quot;PRKDC_HUMAN|P78527&quot;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of psm objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">group_identification</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reviewed</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unreviewed</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Sequence info without flanking</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptide_scores</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unique_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_unique_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>  <span class="c1"># Includes Flanking Seq Info</span>

    <span class="k">def</span> <span class="nf">get_psm_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves psm scores for a given protein.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of psm scores for the given protein.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">score_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">main_score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">score_list</span>

    <span class="k">def</span> <span class="nf">get_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves a list of Psm identifiers.</span>

<span class="sd">         Returns:</span>
<span class="sd">             list: List of Psm identifiers.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">psms</span>

    <span class="k">def</span> <span class="nf">get_stripped_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves a list of Psm identifiers that have had mods removed and flanking AAs removed.</span>

<span class="sd">         Returns:</span>
<span class="sd">             list: List of Psm identifiers that have no mods or flanking AAs.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">psms</span>

    <span class="k">def</span> <span class="nf">get_unique_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the unique set of peptides for a protein.</span>

<span class="sd">         Returns:</span>
<span class="sd">             set: Set of peptide strings.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">unique_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_psm_identifiers</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">unique_peptides</span>

    <span class="k">def</span> <span class="nf">get_unique_stripped_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the unique set of peptides for a protein that are stripped.</span>

<span class="sd">         Returns:</span>
<span class="sd">             set: Set of peptide strings that are stripped of mods and flanking AAs.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stripped_peptide_identifiers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_stripped_psm_identifiers</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">stripped_peptide_identifiers</span>

    <span class="k">def</span> <span class="nf">get_num_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the number of Psms.</span>

<span class="sd">         Returns:</span>
<span class="sd">             int: Number of Psms.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">num_psms</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_psm_identifiers</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">num_psms</span>

    <span class="k">def</span> <span class="nf">get_num_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the number of peptides.</span>

<span class="sd">         Returns:</span>
<span class="sd">             int: Number of peptides.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_unique_peptide_identifiers</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">num_peptides</span>

    <span class="k">def</span> <span class="nf">get_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the Psm Ids.</span>

<span class="sd">         Returns:</span>
<span class="sd">            list: List of Psm Ids.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">psm_id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">psm_ids</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">










  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method for Protein object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>identifier</strong> (<code>str</code>)  String identifier for the Protein object.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">protein</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">physical</span><span class="o">.</span><span class="n">Protein</span><span class="p">(</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;PRKDC_HUMAN|P78527&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method for Protein object.</span>

<span class="sd">    Args:</span>
<span class="sd">        identifier (str): String identifier for the Protein object.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; protein = pyproteininference.physical.Protein(identifier = &quot;PRKDC_HUMAN|P78527&quot;)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># List of psm objects</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">group_identification</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reviewed</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">unreviewed</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Sequence info without flanking</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">peptide_scores</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">picked</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">unique_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_unique_peptides</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># remove</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">raw_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>  <span class="c1"># Includes Flanking Seq Info</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_num_peptides">
<code class="highlight language-python"><span class="n">get_num_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves the number of peptides.</p>
<p>!!! returns
     int: Number of peptides.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_num_peptides</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves the number of peptides.</span>

<span class="sd">     Returns:</span>
<span class="sd">         int: Number of peptides.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">num_peptides</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_unique_peptide_identifiers</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">num_peptides</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_num_psms">
<code class="highlight language-python"><span class="n">get_num_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves the number of Psms.</p>
<p>!!! returns
     int: Number of Psms.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_num_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves the number of Psms.</span>

<span class="sd">     Returns:</span>
<span class="sd">         int: Number of Psms.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">num_psms</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_psm_identifiers</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">num_psms</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_psm_identifiers">
<code class="highlight language-python"><span class="n">get_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves a list of Psm identifiers.</p>
<p>!!! returns
     list: List of Psm identifiers.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves a list of Psm identifiers.</span>

<span class="sd">     Returns:</span>
<span class="sd">         list: List of Psm identifiers.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">identifier</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">psms</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_psm_ids">
<code class="highlight language-python"><span class="n">get_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves the Psm Ids.</p>
<p>Returns:
    list: List of Psm Ids.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_psm_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves the Psm Ids.</span>

<span class="sd">     Returns:</span>
<span class="sd">        list: List of Psm Ids.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psm_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">psm_id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">psm_ids</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_psm_scores">
<code class="highlight language-python"><span class="n">get_psm_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves psm scores for a given protein.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>list</code>  List of psm scores for the given protein.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_psm_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves psm scores for a given protein.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of psm scores for the given protein.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">score_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">main_score</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">score_list</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_stripped_psm_identifiers">
<code class="highlight language-python"><span class="n">get_stripped_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves a list of Psm identifiers that have had mods removed and flanking AAs removed.</p>
<p>!!! returns
     list: List of Psm identifiers that have no mods or flanking AAs.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_stripped_psm_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves a list of Psm identifiers that have had mods removed and flanking AAs removed.</span>

<span class="sd">     Returns:</span>
<span class="sd">         list: List of Psm identifiers that have no mods or flanking AAs.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">psms</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">psms</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_unique_peptide_identifiers">
<code class="highlight language-python"><span class="n">get_unique_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves the unique set of peptides for a protein.</p>
<p>!!! returns
     set: Set of peptide strings.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_unique_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves the unique set of peptides for a protein.</span>

<span class="sd">     Returns:</span>
<span class="sd">         set: Set of peptide strings.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">unique_peptides</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_psm_identifiers</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">unique_peptides</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Protein.get_unique_stripped_peptide_identifiers">
<code class="highlight language-python"><span class="n">get_unique_stripped_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Retrieves the unique set of peptides for a protein that are stripped.</p>
<p>!!! returns
     set: Set of peptide strings that are stripped of mods and flanking AAs.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_unique_stripped_peptide_identifiers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves the unique set of peptides for a protein that are stripped.</span>

<span class="sd">     Returns:</span>
<span class="sd">         set: Set of peptide strings that are stripped of mods and flanking AAs.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stripped_peptide_identifiers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_stripped_psm_identifiers</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">stripped_peptide_identifiers</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.physical.ProteinGroup">
        <code>
ProteinGroup        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class is a physical Protein Group class that stores characteristics of a Protein Group for the entire
    analysis.
We use <strong>slots</strong> to predefine the attributes the Psm Object can have.
This is done to speed up runtime of the PI algorithm.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>number_id</code></td>
        <td><code>int</code></td>
        <td><p>unique Integer to represent a group.</p></td>
      </tr>
      <tr>
        <td><code>proteins</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p></td>
      </tr>
      <tr>
        <td><code>q_value</code></td>
        <td><code>float</code></td>
        <td><p>Q value for the protein group that is calculated with method
<a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.calculate_q_values">calculate_q_values</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ProteinGroup</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class is a physical Protein Group class that stores characteristics of a Protein Group for the entire</span>
<span class="sd">        analysis.</span>
<span class="sd">    We use __slots__ to predefine the attributes the Psm Object can have.</span>
<span class="sd">    This is done to speed up runtime of the PI algorithm.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        number_id (int): unique Integer to represent a group.</span>
<span class="sd">        proteins (list): List of [Protein][pyproteininference.physical.Protein] objects.</span>
<span class="sd">        q_value (float): Q value for the protein group that is calculated with method</span>
<span class="sd">            [calculate_q_values][pyproteininference.datastore.DataStore.calculate_q_values].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;proteins&quot;</span><span class="p">,</span> <span class="s2">&quot;number_id&quot;</span><span class="p">,</span> <span class="s2">&quot;q_value&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method for ProteinGroup object.</span>

<span class="sd">        Args:</span>
<span class="sd">            number_id (int): unique Integer to represent a group.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pg = pyproteininference.physical.ProteinGroup(number_id = 1)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">number_id</span> <span class="o">=</span> <span class="n">number_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">q_value</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">










  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.ProteinGroup.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_id</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method for ProteinGroup object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>number_id</strong> (<code>int</code>)  unique Integer to represent a group.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pg</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">physical</span><span class="o">.</span><span class="n">ProteinGroup</span><span class="p">(</span><span class="n">number_id</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">number_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method for ProteinGroup object.</span>

<span class="sd">    Args:</span>
<span class="sd">        number_id (int): unique Integer to represent a group.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pg = pyproteininference.physical.ProteinGroup(number_id = 1)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">proteins</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">number_id</span> <span class="o">=</span> <span class="n">number_id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">q_value</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.physical.Psm">
        <code>
Psm        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class is a physical Psm class that stores characteristics of a psm for the entire analysis.
We use <strong>slots</strong> to predefine the attributes the Psm Object can have.
This is done to speed up runtime of the PI algorithm.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>identifier</code></td>
        <td><code>str</code></td>
        <td><p>Peptide Identifier: IE "K.DLIDEGH#AATQLVNQLHDVVVENNLSDK.Q".</p></td>
      </tr>
      <tr>
        <td><code>percscore</code></td>
        <td><code>float</code></td>
        <td><p>Percolator Score from input file if it exists.</p></td>
      </tr>
      <tr>
        <td><code>qvalue</code></td>
        <td><code>float</code></td>
        <td><p>Q value from input file if it exists.</p></td>
      </tr>
      <tr>
        <td><code>pepvalue</code></td>
        <td><code>float</code></td>
        <td><p>Pep value from input file if it exists.</p></td>
      </tr>
      <tr>
        <td><code>possible_proteins</code></td>
        <td><code>list</code></td>
        <td><p>List of protein strings that the Psm maps to based on the digest.</p></td>
      </tr>
      <tr>
        <td><code>psm_id</code></td>
        <td><code>str</code></td>
        <td><p>String that represents a global identifier for the Psm. Should come from input files.</p></td>
      </tr>
      <tr>
        <td><code>custom_score</code></td>
        <td><code>float</code></td>
        <td><p>Score that comes from a custom column in the input files.</p></td>
      </tr>
      <tr>
        <td><code>main_score</code></td>
        <td><code>float</code></td>
        <td><p>The Psm score to be used as the scoring variable for protein scoring. can be
percscore,qvalue,pepvalue, or custom_score.</p></td>
      </tr>
      <tr>
        <td><code>stripped_peptide</code></td>
        <td><code>str</code></td>
        <td><p>This is the identifier attribute that has had mods removed and flanking AAs
removed IE: DLIDEGHAATQLVNQLHDVVVENNLSDK.</p></td>
      </tr>
      <tr>
        <td><code>non_flanking_peptide</code></td>
        <td><code>str</code></td>
        <td><p>This is the identifier attribute that has had flanking AAs
removed IE: DLIDEGH#AATQLVNQLHDVVVENNLSDK. #NOTE Mods are still present here.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Psm</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class is a physical Psm class that stores characteristics of a psm for the entire analysis.</span>
<span class="sd">    We use __slots__ to predefine the attributes the Psm Object can have.</span>
<span class="sd">    This is done to speed up runtime of the PI algorithm.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        identifier (str): Peptide Identifier: IE &quot;K.DLIDEGH#AATQLVNQLHDVVVENNLSDK.Q&quot;.</span>
<span class="sd">        percscore (float): Percolator Score from input file if it exists.</span>
<span class="sd">        qvalue (float): Q value from input file if it exists.</span>
<span class="sd">        pepvalue (float): Pep value from input file if it exists.</span>
<span class="sd">        possible_proteins (list): List of protein strings that the Psm maps to based on the digest.</span>
<span class="sd">        psm_id (str): String that represents a global identifier for the Psm. Should come from input files.</span>
<span class="sd">        custom_score (float): Score that comes from a custom column in the input files.</span>
<span class="sd">        main_score (float): The Psm score to be used as the scoring variable for protein scoring. can be</span>
<span class="sd">            percscore,qvalue,pepvalue, or custom_score.</span>
<span class="sd">        stripped_peptide (str): This is the identifier attribute that has had mods removed and flanking AAs</span>
<span class="sd">            removed IE: DLIDEGHAATQLVNQLHDVVVENNLSDK.</span>
<span class="sd">        non_flanking_peptide (str): This is the identifier attribute that has had flanking AAs</span>
<span class="sd">            removed IE: DLIDEGH#AATQLVNQLHDVVVENNLSDK. #NOTE Mods are still present here.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="vm">__slots__</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;identifier&quot;</span><span class="p">,</span>
        <span class="s2">&quot;percscore&quot;</span><span class="p">,</span>
        <span class="s2">&quot;qvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pepvalue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;possible_proteins&quot;</span><span class="p">,</span>
        <span class="s2">&quot;psm_id&quot;</span><span class="p">,</span>
        <span class="s2">&quot;custom_score&quot;</span><span class="p">,</span>
        <span class="s2">&quot;main_score&quot;</span><span class="p">,</span>
        <span class="s2">&quot;stripped_peptide&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_flanking_peptide&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># The regex removes anything between parantheses including parenthases - \([^()]*\)</span>
    <span class="c1"># The regex removes anything between brackets including parenthases - \[.*?\]</span>
    <span class="c1"># And the regex removes anything that is not an A-Z character [^A-Z]</span>
    <span class="n">MOD_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;\([^()]*\)|\[.*?\]|[^A-Z]&quot;</span><span class="p">)</span>  <span class="c1"># noqa W605</span>

    <span class="n">FRONT_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;^[A-Z|-][.]&quot;</span><span class="p">)</span>
    <span class="n">BACK_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;[.][A-Z|-]$&quot;</span><span class="p">)</span>

    <span class="n">SCORE_ATTRIBUTE_NAMES</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;pepvalue&quot;</span><span class="p">,</span> <span class="s2">&quot;qvalue&quot;</span><span class="p">,</span> <span class="s2">&quot;percscore&quot;</span><span class="p">,</span> <span class="s2">&quot;custom_score&quot;</span><span class="p">])</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method for the Psm object.</span>
<span class="sd">        This method also initializes the `stripped_peptide` and `non_flanking_peptide` attributes.</span>

<span class="sd">        Args:</span>
<span class="sd">            identifier (str): Peptide Identifier: IE &quot;&quot;K.DLIDEGH#AATQLVNQLHDVVVENNLSDK.Q&quot;.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; psm = pyproteininference.physical.Psm(identifier = &quot;K.DLIDEGHAATQLVNQLHDVVVENNLSDK.Q&quot;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">main_score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Add logic to split the peptide and strip it of mods</span>
        <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="o">=</span> <span class="n">current_peptide</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
            <span class="c1"># If we have mods remove them...</span>
            <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

        <span class="c1"># Set stripped_peptide variable</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">current_peptide</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">remove_peptide_mods</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class method takes a string and uses a `MOD_REGEX` to remove mods from peptide strings.</span>

<span class="sd">        Args:</span>
<span class="sd">            peptide_string (str): Peptide string to have mods removed from.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: a peptide string with mods removed.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">MOD_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stripped_peptide</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">split_peptide</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class method takes a peptide string with flanking AAs and removes them from the peptide string.</span>
<span class="sd">        This method uses string splitting and if the method produces a faulty peptide the method</span>
<span class="sd">            [split_peptide_pro][pyproteininference.physical.Psm.split_peptide_pro] will be called.</span>

<span class="sd">        Args:</span>
<span class="sd">            peptide_string (str): Peptide string to have mods removed from.</span>
<span class="sd">            delimiter (str): a string to indicate what separates a leading/trailing (flanking) AA from the</span>
<span class="sd">                peptide sequence.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: a peptide string with flanking AAs removed.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">peptide_split</span> <span class="o">=</span> <span class="n">peptide_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptide_split</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="c1"># If we get 3 chunks it will usually be [&#39;A&#39;, &#39;ADGSDFGSS&#39;, &#39;F&#39;]</span>
            <span class="c1"># So take index 1</span>
            <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide_split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptide_split</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># If we get 1 chunk it should just be [&#39;ADGSDFGSS&#39;]</span>
            <span class="c1"># So take index 0</span>
            <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide_split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If we split the peptide and it is not length 1 or 3 then try to split with pro</span>
            <span class="n">peptide</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">split_peptide_pro</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">peptide</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">split_peptide_pro</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class method takes a peptide string with flanking AAs and removes them from the peptide string.</span>
<span class="sd">        This is a specialized method of [split_peptide][pyproteininference.physical.Psm.split_peptide] that uses</span>
<span class="sd">         regex identifiers to replace flanking AAs as opposed to string splitting.</span>


<span class="sd">        Args:</span>
<span class="sd">            peptide_string (str): Peptide string to have mods removed from.</span>
<span class="sd">            delimiter (str): a string to indicate what separates a leading/trailing (flanking) AA from the peptide</span>
<span class="sd">                sequence.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: a peptide string with flanking AAs removed.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">delimiter</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
            <span class="n">front_regex</span> <span class="o">=</span> <span class="s2">&quot;^[A-Z|-][</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">FRONT_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">front_regex</span><span class="p">)</span>
            <span class="n">back_regex</span> <span class="o">=</span> <span class="s2">&quot;[</span><span class="si">{}</span><span class="s2">][A-Z|-]$&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">BACK_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">back_regex</span><span class="p">)</span>

        <span class="c1"># Replace the front flanking with nothing</span>
        <span class="n">peptide_string</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">FRONT_FLANKING_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>

        <span class="c1"># Replace the back flanking with nothing</span>
        <span class="n">peptide_string</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">BACK_FLANKING_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">peptide_string</span>

    <span class="k">def</span> <span class="nf">assign_main_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method takes in a score type and assigns the variable main_score for a given Psm based on the score type.</span>

<span class="sd">        Args:</span>
<span class="sd">            score (str): This is a string representation of the Psm attribute that will get assigned to the main_score</span>
<span class="sd">                variable.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Assign a main score based on user input</span>
        <span class="k">if</span> <span class="n">score</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_ATTRIBUTE_NAMES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Scores must either be one of: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_ATTRIBUTE_NAMES</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">main_score</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">














  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Psm.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method for the Psm object.
This method also initializes the <code>stripped_peptide</code> and <code>non_flanking_peptide</code> attributes.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>identifier</strong> (<code>str</code>)  Peptide Identifier: IE ""K.DLIDEGH#AATQLVNQLHDVVVENNLSDK.Q".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">psm</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">physical</span><span class="o">.</span><span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;K.DLIDEGHAATQLVNQLHDVVVENNLSDK.Q&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method for the Psm object.</span>
<span class="sd">    This method also initializes the `stripped_peptide` and `non_flanking_peptide` attributes.</span>

<span class="sd">    Args:</span>
<span class="sd">        identifier (str): Peptide Identifier: IE &quot;&quot;K.DLIDEGH#AATQLVNQLHDVVVENNLSDK.Q&quot;.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; psm = pyproteininference.physical.Psm(identifier = &quot;K.DLIDEGHAATQLVNQLHDVVVENNLSDK.Q&quot;)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">main_score</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Add logic to split the peptide and strip it of mods</span>
    <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">non_flanking_peptide</span> <span class="o">=</span> <span class="n">current_peptide</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
        <span class="c1"># If we have mods remove them...</span>
        <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
        <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

    <span class="c1"># Set stripped_peptide variable</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">current_peptide</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Psm.assign_main_score">
<code class="highlight language-python"><span class="n">assign_main_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method takes in a score type and assigns the variable main_score for a given Psm based on the score type.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>score</strong> (<code>str</code>)  This is a string representation of the Psm attribute that will get assigned to the main_score
variable.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">assign_main_score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method takes in a score type and assigns the variable main_score for a given Psm based on the score type.</span>

<span class="sd">    Args:</span>
<span class="sd">        score (str): This is a string representation of the Psm attribute that will get assigned to the main_score</span>
<span class="sd">            variable.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Assign a main score based on user input</span>
    <span class="k">if</span> <span class="n">score</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_ATTRIBUTE_NAMES</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Scores must either be one of: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_ATTRIBUTE_NAMES</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">main_score</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Psm.remove_peptide_mods">
<code class="highlight language-python"><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>This class method takes a string and uses a <code>MOD_REGEX</code> to remove mods from peptide strings.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>peptide_string</strong> (<code>str</code>)  Peptide string to have mods removed from.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>str</code>  a peptide string with mods removed.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">remove_peptide_mods</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class method takes a string and uses a `MOD_REGEX` to remove mods from peptide strings.</span>

<span class="sd">    Args:</span>
<span class="sd">        peptide_string (str): Peptide string to have mods removed from.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: a peptide string with mods removed.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">MOD_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stripped_peptide</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Psm.split_peptide">
<code class="highlight language-python"><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>This class method takes a peptide string with flanking AAs and removes them from the peptide string.
This method uses string splitting and if the method produces a faulty peptide the method
    <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm.split_peptide_pro">split_peptide_pro</a> will be called.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>peptide_string</strong> (<code>str</code>)  Peptide string to have mods removed from.</p></li>
            <li><p><strong>delimiter</strong> (<code>str</code>)  a string to indicate what separates a leading/trailing (flanking) AA from the
peptide sequence.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>str</code>  a peptide string with flanking AAs removed.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">split_peptide</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class method takes a peptide string with flanking AAs and removes them from the peptide string.</span>
<span class="sd">    This method uses string splitting and if the method produces a faulty peptide the method</span>
<span class="sd">        [split_peptide_pro][pyproteininference.physical.Psm.split_peptide_pro] will be called.</span>

<span class="sd">    Args:</span>
<span class="sd">        peptide_string (str): Peptide string to have mods removed from.</span>
<span class="sd">        delimiter (str): a string to indicate what separates a leading/trailing (flanking) AA from the</span>
<span class="sd">            peptide sequence.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: a peptide string with flanking AAs removed.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">peptide_split</span> <span class="o">=</span> <span class="n">peptide_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptide_split</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="c1"># If we get 3 chunks it will usually be [&#39;A&#39;, &#39;ADGSDFGSS&#39;, &#39;F&#39;]</span>
        <span class="c1"># So take index 1</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide_split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptide_split</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="c1"># If we get 1 chunk it should just be [&#39;ADGSDFGSS&#39;]</span>
        <span class="c1"># So take index 0</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide_split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># If we split the peptide and it is not length 1 or 3 then try to split with pro</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">split_peptide_pro</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">peptide</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.physical.Psm.split_peptide_pro">
<code class="highlight language-python"><span class="n">split_peptide_pro</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-classmethod"><code>classmethod</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>This class method takes a peptide string with flanking AAs and removes them from the peptide string.
This is a specialized method of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm.split_peptide">split_peptide</a> that uses
 regex identifiers to replace flanking AAs as opposed to string splitting.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>peptide_string</strong> (<code>str</code>)  Peptide string to have mods removed from.</p></li>
            <li><p><strong>delimiter</strong> (<code>str</code>)  a string to indicate what separates a leading/trailing (flanking) AA from the peptide
sequence.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>str</code>  a peptide string with flanking AAs removed.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/physical.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">split_peptide_pro</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class method takes a peptide string with flanking AAs and removes them from the peptide string.</span>
<span class="sd">    This is a specialized method of [split_peptide][pyproteininference.physical.Psm.split_peptide] that uses</span>
<span class="sd">     regex identifiers to replace flanking AAs as opposed to string splitting.</span>


<span class="sd">    Args:</span>
<span class="sd">        peptide_string (str): Peptide string to have mods removed from.</span>
<span class="sd">        delimiter (str): a string to indicate what separates a leading/trailing (flanking) AA from the peptide</span>
<span class="sd">            sequence.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: a peptide string with flanking AAs removed.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">delimiter</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
        <span class="n">front_regex</span> <span class="o">=</span> <span class="s2">&quot;^[A-Z|-][</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">FRONT_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">front_regex</span><span class="p">)</span>
        <span class="n">back_regex</span> <span class="o">=</span> <span class="s2">&quot;[</span><span class="si">{}</span><span class="s2">][A-Z|-]$&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">delimiter</span><span class="p">)</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">BACK_FLANKING_REGEX</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">back_regex</span><span class="p">)</span>

    <span class="c1"># Replace the front flanking with nothing</span>
    <span class="n">peptide_string</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">FRONT_FLANKING_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>

    <span class="c1"># Replace the back flanking with nothing</span>
    <span class="n">peptide_string</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">BACK_FLANKING_REGEX</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">peptide_string</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">peptide_string</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.pipeline">
        <code>pipeline</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.pipeline.ProteinInferencePipeline">
        <code>
ProteinInferencePipeline        </code>



</h3>

    <div class="doc doc-contents ">

      <p>This is the main Protein Inference class which houses the logic of the entire data analysis pipeline.
Logic is executed in the <a class="autorefs autorefs-internal" href="#pyproteininference.pipeline.ProteinInferencePipeline.execute">execute</a> method.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>parameter_file</code></td>
        <td><code>str</code></td>
        <td><p>Path to Protein Inference Yaml Parameter File.</p></td>
      </tr>
      <tr>
        <td><code>database_file</code></td>
        <td><code>str</code></td>
        <td><p>Path to Fasta database used in proteomics search.</p></td>
      </tr>
      <tr>
        <td><code>target_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Target Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>decoy_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Decoy Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>combined_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Combined Psm File (Or a list of files).</p></td>
      </tr>
      <tr>
        <td><code>target_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Target Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>decoy_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Decoy Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>combined_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory containing Combined Psm Files.</p></td>
      </tr>
      <tr>
        <td><code>output_directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to Directory where output will be written.</p></td>
      </tr>
      <tr>
        <td><code>output_filename</code></td>
        <td><code>str</code></td>
        <td><p>Path to Filename where output will be written. Will override output_directory.</p></td>
      </tr>
      <tr>
        <td><code>id_splitting</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether to split protein IDs in the digest. Advanced usage only.</p></td>
      </tr>
      <tr>
        <td><code>append_alt_from_db</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether to append alternative proteins from the DB digestion in
Reader class.</p></td>
      </tr>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/pipeline.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ProteinInferencePipeline</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is the main Protein Inference class which houses the logic of the entire data analysis pipeline.</span>
<span class="sd">    Logic is executed in the [execute][pyproteininference.pipeline.ProteinInferencePipeline.execute] method.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">        database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">        target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">        decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">        combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">        target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">        decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">        combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">        output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">        output_filename (str): Path to Filename where output will be written. Will override output_directory.</span>
<span class="sd">        id_splitting (bool): True/False on whether to split protein IDs in the digest. Advanced usage only.</span>
<span class="sd">        append_alt_from_db (bool): True/False on whether to append alternative proteins from the DB digestion in</span>
<span class="sd">            Reader class.</span>
<span class="sd">        data (DataStore): [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">parameter_file</span><span class="p">,</span>
        <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">            database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">            target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">            decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">            combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">            target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">            decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">            combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">            output_filename (str): Path to Filename where output will be written. Will override output_directory.</span>
<span class="sd">            output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">            id_splitting (bool): True/False on whether to split protein IDs in the digest. Advanced usage only.</span>
<span class="sd">            append_alt_from_db (bool): True/False on whether to append alternative proteins from the DB digestion in</span>
<span class="sd">                Reader class.</span>

<span class="sd">        Returns:</span>
<span class="sd">            object:</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pipeline = pyproteininference.pipeline.ProteinInferencePipeline(</span>
<span class="sd">            &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">            &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">            &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">            &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">            &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">            &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">            &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">            &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">            &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span> <span class="o">=</span> <span class="n">parameter_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span> <span class="o">=</span> <span class="n">database_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span> <span class="o">=</span> <span class="n">target_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span> <span class="o">=</span> <span class="n">decoy_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span> <span class="o">=</span> <span class="n">target_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span> <span class="o">=</span> <span class="n">decoy_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span> <span class="o">=</span> <span class="n">combined_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">output_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="o">=</span> <span class="n">output_filename</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_output_directory</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_log_append_alt_from_db</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_log_id_splitting</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method is the main driver of the data analysis for the protein inference package.</span>
<span class="sd">        This method calls other classes and methods that make up the protein inference pipeline.</span>
<span class="sd">        This includes but is not limited to:</span>

<span class="sd">        This method sets the data [DataStore Object][pyproteininference.datastore.DataStore] and digest</span>
<span class="sd">            [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">        1. Parameter file management.</span>
<span class="sd">        2. Digesting Fasta Database (Optional).</span>
<span class="sd">        3. Reading in input Psm Files.</span>
<span class="sd">        4. Initializing the [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">        5. Restricting Psms.</span>
<span class="sd">        6. Creating Protein objects/scoring input.</span>
<span class="sd">        7. Scoring Proteins.</span>
<span class="sd">        8. Running Protein Picker.</span>
<span class="sd">        9. Running Inference Methods/Grouping.</span>
<span class="sd">        10. Calculating Q Values.</span>
<span class="sd">        11. Exporting Proteins to filesystem.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pipeline = pyproteininference.pipeline.ProteinInferencePipeline(</span>
<span class="sd">            &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">            &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">            &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">            &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">            &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">            &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">            &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">            &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">            &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">            &gt;&gt;&gt; )</span>
<span class="sd">            &gt;&gt;&gt; pipeline.execute()</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># STEP 1: Load parameter file #</span>
        <span class="c1"># STEP 1: Load parameter file #</span>
        <span class="c1"># STEP 1: Load parameter file #</span>
        <span class="n">pyproteininference_parameters</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ProteinInferenceParameter</span><span class="p">(</span>
            <span class="n">yaml_param_filepath</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span>
        <span class="p">)</span>

        <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
        <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
        <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
        <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
            <span class="n">database_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">,</span>
            <span class="n">digest_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">digest_type</span><span class="p">,</span>
            <span class="n">missed_cleavages</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
            <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
            <span class="n">max_peptide_length</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">,</span>
            <span class="n">id_splitting</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running In Silico Database Digest on file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">))</span>
            <span class="n">digest</span><span class="o">.</span><span class="n">digest_fasta_database</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;No Database File provided, Skipping database digest and only taking protein-peptide mapping from the &quot;</span>
                <span class="s2">&quot;input files.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># STEP 3: Read PSM Data #</span>
        <span class="c1"># STEP 3: Read PSM Data #</span>
        <span class="c1"># STEP 3: Read PSM Data #</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span>
            <span class="n">target_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">target_files</span><span class="p">,</span>
            <span class="n">decoy_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span><span class="p">,</span>
            <span class="n">combined_files</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span>
            <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="p">,</span>
            <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span>
            <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>

        <span class="c1"># STEP 4: Initiate the datastore object #</span>
        <span class="c1"># STEP 4: Initiate the datastore object #</span>
        <span class="c1"># STEP 4: Initiate the datastore object #</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

        <span class="c1"># Step 5: Restrict the PSM data</span>
        <span class="c1"># Step 5: Restrict the PSM data</span>
        <span class="c1"># Step 5: Restrict the PSM data</span>
        <span class="n">data</span><span class="o">.</span><span class="n">restrict_psm_data</span><span class="p">()</span>

        <span class="n">data</span><span class="o">.</span><span class="n">recover_mapping</span><span class="p">()</span>
        <span class="c1"># Step 6: Generate protein scoring input</span>
        <span class="c1"># Step 6: Generate protein scoring input</span>
        <span class="c1"># Step 6: Generate protein scoring input</span>
        <span class="n">data</span><span class="o">.</span><span class="n">create_scoring_input</span><span class="p">()</span>

        <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
        <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
        <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
        <span class="k">if</span> <span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
            <span class="c1"># This gets ran if we run exclusion...</span>
            <span class="n">data</span><span class="o">.</span><span class="n">exclude_non_distinguishing_peptides</span><span class="p">()</span>

        <span class="c1"># STEP 8: Score our PSMs given a score method</span>
        <span class="c1"># STEP 8: Score our PSMs given a score method</span>
        <span class="c1"># STEP 8: Score our PSMs given a score method</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        <span class="n">score</span><span class="o">.</span><span class="n">score_psms</span><span class="p">(</span><span class="n">score_method</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">protein_score</span><span class="p">)</span>

        <span class="c1"># STEP 9: Run protein picker on the data</span>
        <span class="c1"># STEP 9: Run protein picker on the data</span>
        <span class="c1"># STEP 9: Run protein picker on the data</span>
        <span class="k">if</span> <span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">picker</span><span class="p">:</span>
            <span class="n">data</span><span class="o">.</span><span class="n">protein_picker</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="c1"># STEP 10: Apply Inference</span>
        <span class="c1"># STEP 10: Apply Inference</span>
        <span class="c1"># STEP 10: Apply Inference</span>
        <span class="n">pyproteininference</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">Inference</span><span class="o">.</span><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

        <span class="c1"># STEP 11: Q value Calculations</span>
        <span class="c1"># STEP 11: Q value Calculations</span>
        <span class="c1"># STEP 11: Q value Calculations</span>
        <span class="n">data</span><span class="o">.</span><span class="n">calculate_q_values</span><span class="p">()</span>

        <span class="c1"># STEP 12: Export to CSV</span>
        <span class="c1"># STEP 12: Export to CSV</span>
        <span class="c1"># STEP 12: Export to CSV</span>
        <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
        <span class="n">export</span><span class="o">.</span><span class="n">export_to_csv</span><span class="p">(</span>
            <span class="n">output_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">,</span>
            <span class="n">directory</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
            <span class="n">export_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">export</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Protein Inference Finished&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate_input</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method that validates whether the proper input files have been defined.</span>

<span class="sd">        One of the following combinations must be selected as input. No more and no less:</span>

<span class="sd">        1. either one or multiple target_files and decoy_files.</span>
<span class="sd">        2. either one or multiple combined_files that include target and decoy data.</span>
<span class="sd">        3. a directory that contains target files (target_directory) as well as a directory that contains decoy files</span>
<span class="sd">            (decoy_directory).</span>
<span class="sd">        4. a directory that contains combined target/decoy files (combined_directory).</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: ValueError will occur if an improper combination of input.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as target_files and decoy_files&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as combined_files&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span>
            <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as target_directory and decoy_directory&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_transform_directory_to_files</span><span class="p">()</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as combined_directory&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_transform_directory_to_files</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;To run Protein inference please supply either: &quot;</span>
                <span class="s2">&quot;(1) either one or multiple target_files and decoy_files, &quot;</span>
                <span class="s2">&quot;(2) either one or multiple combined_files that include target and decoy data&quot;</span>
                <span class="s2">&quot;(3) a directory that contains target files (target_directory) as well as a directory that &quot;</span>
                <span class="s2">&quot;contains decoy files (decoy_directory)&quot;</span>
                <span class="s2">&quot;(4) a directory that contains combined target/decoy files (combined_directory)&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_transform_directory_to_files</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This internal method takes files that are in the target_directory, decoy_directory, or combined_directory and</span>
<span class="sd">        reassigns these files to the target_files, decoy_files, and combined_files to be used in</span>
<span class="sd">         [Reader][pyproteininference.reader.Reader] object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Transforming target_directory and decoy_directory into files&quot;</span><span class="p">)</span>
            <span class="n">target_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span><span class="p">)</span>
            <span class="n">target_files_full</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">target_files</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="n">decoy_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span><span class="p">)</span>
            <span class="n">decoy_files_full</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">decoy_files</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span> <span class="o">=</span> <span class="n">target_files_full</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span> <span class="o">=</span> <span class="n">decoy_files_full</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Transforming combined_directory into files&quot;</span><span class="p">)</span>
            <span class="n">combined_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span><span class="p">)</span>
            <span class="n">combined_files_full</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">combined_files</span>
                <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files_full</span>

    <span class="k">def</span> <span class="nf">_set_output_directory</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method for setting the output directory.</span>
<span class="sd">        If the output_directory argument is not supplied the output directory is set as the cwd.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">_log_append_alt_from_db</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method for logging whether the user sets alternative protein append to True or False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Append Alternative Proteins from Database set to True&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Append Alternative Proteins from Database set to False&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_log_id_splitting</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method for logging whether the user sets ID splitting to True or False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;ID Splitting for Database Digestion set to True&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;ID Splitting for Database Digestion set to False&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.pipeline.ProteinInferencePipeline.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameter_file</span><span class="p">,</span> <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>parameter_file</strong> (<code>str</code>)  Path to Protein Inference Yaml Parameter File.</p></li>
            <li><p><strong>database_file</strong> (<code>str</code>)  Path to Fasta database used in proteomics search.</p></li>
            <li><p><strong>target_files</strong> (<code>str/list</code>)  Path to Target Psm File (Or a list of files).</p></li>
            <li><p><strong>decoy_files</strong> (<code>str/list</code>)  Path to Decoy Psm File (Or a list of files).</p></li>
            <li><p><strong>combined_files</strong> (<code>str/list</code>)  Path to Combined Psm File (Or a list of files).</p></li>
            <li><p><strong>target_directory</strong> (<code>str</code>)  Path to Directory containing Target Psm Files.</p></li>
            <li><p><strong>decoy_directory</strong> (<code>str</code>)  Path to Directory containing Decoy Psm Files.</p></li>
            <li><p><strong>combined_directory</strong> (<code>str</code>)  Path to Directory containing Combined Psm Files.</p></li>
            <li><p><strong>output_filename</strong> (<code>str</code>)  Path to Filename where output will be written. Will override output_directory.</p></li>
            <li><p><strong>output_directory</strong> (<code>str</code>)  Path to Directory where output will be written.</p></li>
            <li><p><strong>id_splitting</strong> (<code>bool</code>)  True/False on whether to split protein IDs in the digest. Advanced usage only.</p></li>
            <li><p><strong>append_alt_from_db</strong> (<code>bool</code>)  True/False on whether to append alternative proteins from the DB digestion in
Reader class.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>object</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pipeline</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">pipeline</span><span class="o">.</span><span class="n">ProteinInferencePipeline</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">parameter_file</span><span class="o">=</span><span class="n">yaml_params</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_file</span><span class="o">=</span><span class="n">database</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_files</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_files</span><span class="o">=</span><span class="n">decoy</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_files</span><span class="o">=</span><span class="n">combined_files</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_directory</span><span class="o">=</span><span class="n">target_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_directory</span><span class="o">=</span><span class="n">decoy_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_directory</span><span class="o">=</span><span class="n">combined_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_directory</span><span class="o">=</span><span class="n">dir_name</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_filename</span><span class="o">=</span><span class="n">output_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">append_alt_from_db</span><span class="o">=</span><span class="n">append_alt</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/pipeline.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">parameter_file</span><span class="p">,</span>
    <span class="n">database_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">target_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">target_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">output_directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">output_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">id_splitting</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        parameter_file (str): Path to Protein Inference Yaml Parameter File.</span>
<span class="sd">        database_file (str): Path to Fasta database used in proteomics search.</span>
<span class="sd">        target_files (str/list): Path to Target Psm File (Or a list of files).</span>
<span class="sd">        decoy_files (str/list): Path to Decoy Psm File (Or a list of files).</span>
<span class="sd">        combined_files (str/list): Path to Combined Psm File (Or a list of files).</span>
<span class="sd">        target_directory (str): Path to Directory containing Target Psm Files.</span>
<span class="sd">        decoy_directory (str): Path to Directory containing Decoy Psm Files.</span>
<span class="sd">        combined_directory (str): Path to Directory containing Combined Psm Files.</span>
<span class="sd">        output_filename (str): Path to Filename where output will be written. Will override output_directory.</span>
<span class="sd">        output_directory (str): Path to Directory where output will be written.</span>
<span class="sd">        id_splitting (bool): True/False on whether to split protein IDs in the digest. Advanced usage only.</span>
<span class="sd">        append_alt_from_db (bool): True/False on whether to append alternative proteins from the DB digestion in</span>
<span class="sd">            Reader class.</span>

<span class="sd">    Returns:</span>
<span class="sd">        object:</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pipeline = pyproteininference.pipeline.ProteinInferencePipeline(</span>
<span class="sd">        &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">        &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">        &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">        &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">        &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">        &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">        &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">        &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">        &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span> <span class="o">=</span> <span class="n">parameter_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span> <span class="o">=</span> <span class="n">database_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_files</span> <span class="o">=</span> <span class="n">target_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span> <span class="o">=</span> <span class="n">decoy_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_directory</span> <span class="o">=</span> <span class="n">target_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_directory</span> <span class="o">=</span> <span class="n">decoy_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_directory</span> <span class="o">=</span> <span class="n">combined_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">output_directory</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span> <span class="o">=</span> <span class="n">output_filename</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span> <span class="o">=</span> <span class="n">id_splitting</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_set_output_directory</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_log_append_alt_from_db</span><span class="p">()</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_log_id_splitting</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.pipeline.ProteinInferencePipeline.execute">
<code class="highlight language-python"><span class="n">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method is the main driver of the data analysis for the protein inference package.
This method calls other classes and methods that make up the protein inference pipeline.
This includes but is not limited to:</p>
<p>This method sets the data <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a> and digest
    <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p>
<ol>
<li>Parameter file management.</li>
<li>Digesting Fasta Database (Optional).</li>
<li>Reading in input Psm Files.</li>
<li>Initializing the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore Object</a>.</li>
<li>Restricting Psms.</li>
<li>Creating Protein objects/scoring input.</li>
<li>Scoring Proteins.</li>
<li>Running Protein Picker.</li>
<li>Running Inference Methods/Grouping.</li>
<li>Calculating Q Values.</li>
<li>Exporting Proteins to filesystem.</li>
</ol>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pipeline</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">pipeline</span><span class="o">.</span><span class="n">ProteinInferencePipeline</span><span class="p">(</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">parameter_file</span><span class="o">=</span><span class="n">yaml_params</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">database_file</span><span class="o">=</span><span class="n">database</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_files</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_files</span><span class="o">=</span><span class="n">decoy</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_files</span><span class="o">=</span><span class="n">combined_files</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">target_directory</span><span class="o">=</span><span class="n">target_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_directory</span><span class="o">=</span><span class="n">decoy_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">combined_directory</span><span class="o">=</span><span class="n">combined_directory</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_directory</span><span class="o">=</span><span class="n">dir_name</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">output_filename</span><span class="o">=</span><span class="n">output_filename</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">append_alt_from_db</span><span class="o">=</span><span class="n">append_alt</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">execute</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/pipeline.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">execute</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method is the main driver of the data analysis for the protein inference package.</span>
<span class="sd">    This method calls other classes and methods that make up the protein inference pipeline.</span>
<span class="sd">    This includes but is not limited to:</span>

<span class="sd">    This method sets the data [DataStore Object][pyproteininference.datastore.DataStore] and digest</span>
<span class="sd">        [Digest Object][pyproteininference.in_silico_digest.Digest].</span>

<span class="sd">    1. Parameter file management.</span>
<span class="sd">    2. Digesting Fasta Database (Optional).</span>
<span class="sd">    3. Reading in input Psm Files.</span>
<span class="sd">    4. Initializing the [DataStore Object][pyproteininference.datastore.DataStore].</span>
<span class="sd">    5. Restricting Psms.</span>
<span class="sd">    6. Creating Protein objects/scoring input.</span>
<span class="sd">    7. Scoring Proteins.</span>
<span class="sd">    8. Running Protein Picker.</span>
<span class="sd">    9. Running Inference Methods/Grouping.</span>
<span class="sd">    10. Calculating Q Values.</span>
<span class="sd">    11. Exporting Proteins to filesystem.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pipeline = pyproteininference.pipeline.ProteinInferencePipeline(</span>
<span class="sd">        &gt;&gt;&gt;     parameter_file=yaml_params,</span>
<span class="sd">        &gt;&gt;&gt;     database_file=database,</span>
<span class="sd">        &gt;&gt;&gt;     target_files=target,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_files=decoy,</span>
<span class="sd">        &gt;&gt;&gt;     combined_files=combined_files,</span>
<span class="sd">        &gt;&gt;&gt;     target_directory=target_directory,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_directory=decoy_directory,</span>
<span class="sd">        &gt;&gt;&gt;     combined_directory=combined_directory,</span>
<span class="sd">        &gt;&gt;&gt;     output_directory=dir_name,</span>
<span class="sd">        &gt;&gt;&gt;     output_filename=output_filename,</span>
<span class="sd">        &gt;&gt;&gt;     append_alt_from_db=append_alt,</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">        &gt;&gt;&gt; pipeline.execute()</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># STEP 1: Load parameter file #</span>
    <span class="c1"># STEP 1: Load parameter file #</span>
    <span class="c1"># STEP 1: Load parameter file #</span>
    <span class="n">pyproteininference_parameters</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ProteinInferenceParameter</span><span class="p">(</span>
        <span class="n">yaml_param_filepath</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file</span>
    <span class="p">)</span>

    <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
    <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
    <span class="c1"># STEP 2: Start with running an In Silico Digestion #</span>
    <span class="n">digest</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">in_silico_digest</span><span class="o">.</span><span class="n">PyteomicsDigest</span><span class="p">(</span>
        <span class="n">database_path</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">,</span>
        <span class="n">digest_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">digest_type</span><span class="p">,</span>
        <span class="n">missed_cleavages</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">missed_cleavages</span><span class="p">,</span>
        <span class="n">reviewed_identifier_symbol</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">reviewed_identifier_symbol</span><span class="p">,</span>
        <span class="n">max_peptide_length</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">restrict_peptide_length</span><span class="p">,</span>
        <span class="n">id_splitting</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">id_splitting</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running In Silico Database Digest on file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">database_file</span><span class="p">))</span>
        <span class="n">digest</span><span class="o">.</span><span class="n">digest_fasta_database</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;No Database File provided, Skipping database digest and only taking protein-peptide mapping from the &quot;</span>
            <span class="s2">&quot;input files.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># STEP 3: Read PSM Data #</span>
    <span class="c1"># STEP 3: Read PSM Data #</span>
    <span class="c1"># STEP 3: Read PSM Data #</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span>
        <span class="n">target_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">target_files</span><span class="p">,</span>
        <span class="n">decoy_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_files</span><span class="p">,</span>
        <span class="n">combined_files</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span>
        <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="p">,</span>
        <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>

    <span class="c1"># STEP 4: Initiate the datastore object #</span>
    <span class="c1"># STEP 4: Initiate the datastore object #</span>
    <span class="c1"># STEP 4: Initiate the datastore object #</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">datastore</span><span class="o">.</span><span class="n">DataStore</span><span class="p">(</span><span class="n">reader</span><span class="o">=</span><span class="n">reader</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

    <span class="c1"># Step 5: Restrict the PSM data</span>
    <span class="c1"># Step 5: Restrict the PSM data</span>
    <span class="c1"># Step 5: Restrict the PSM data</span>
    <span class="n">data</span><span class="o">.</span><span class="n">restrict_psm_data</span><span class="p">()</span>

    <span class="n">data</span><span class="o">.</span><span class="n">recover_mapping</span><span class="p">()</span>
    <span class="c1"># Step 6: Generate protein scoring input</span>
    <span class="c1"># Step 6: Generate protein scoring input</span>
    <span class="c1"># Step 6: Generate protein scoring input</span>
    <span class="n">data</span><span class="o">.</span><span class="n">create_scoring_input</span><span class="p">()</span>

    <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
    <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
    <span class="c1"># Step 7: Remove non unique peptides if running exclusion</span>
    <span class="k">if</span> <span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">EXCLUSION</span><span class="p">:</span>
        <span class="c1"># This gets ran if we run exclusion...</span>
        <span class="n">data</span><span class="o">.</span><span class="n">exclude_non_distinguishing_peptides</span><span class="p">()</span>

    <span class="c1"># STEP 8: Score our PSMs given a score method</span>
    <span class="c1"># STEP 8: Score our PSMs given a score method</span>
    <span class="c1"># STEP 8: Score our PSMs given a score method</span>
    <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="n">score</span><span class="o">.</span><span class="n">score_psms</span><span class="p">(</span><span class="n">score_method</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">protein_score</span><span class="p">)</span>

    <span class="c1"># STEP 9: Run protein picker on the data</span>
    <span class="c1"># STEP 9: Run protein picker on the data</span>
    <span class="c1"># STEP 9: Run protein picker on the data</span>
    <span class="k">if</span> <span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">picker</span><span class="p">:</span>
        <span class="n">data</span><span class="o">.</span><span class="n">protein_picker</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="c1"># STEP 10: Apply Inference</span>
    <span class="c1"># STEP 10: Apply Inference</span>
    <span class="c1"># STEP 10: Apply Inference</span>
    <span class="n">pyproteininference</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">Inference</span><span class="o">.</span><span class="n">run_inference</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">)</span>

    <span class="c1"># STEP 11: Q value Calculations</span>
    <span class="c1"># STEP 11: Q value Calculations</span>
    <span class="c1"># STEP 11: Q value Calculations</span>
    <span class="n">data</span><span class="o">.</span><span class="n">calculate_q_values</span><span class="p">()</span>

    <span class="c1"># STEP 12: Export to CSV</span>
    <span class="c1"># STEP 12: Export to CSV</span>
    <span class="c1"># STEP 12: Export to CSV</span>
    <span class="n">export</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">Export</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="n">export</span><span class="o">.</span><span class="n">export_to_csv</span><span class="p">(</span>
        <span class="n">output_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_filename</span><span class="p">,</span>
        <span class="n">directory</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span>
        <span class="n">export_type</span><span class="o">=</span><span class="n">pyproteininference_parameters</span><span class="o">.</span><span class="n">export</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Protein Inference Finished&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.reader">
        <code>reader</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.reader.GenericReader">
        <code>
GenericReader            (<a class="autorefs autorefs-internal" title="pyproteininference.reader.Reader" href="#pyproteininference.reader.Reader">Reader</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class takes a percolator like target file and a percolator like decoy file
and creates standard <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p>
<p>Percolator Like Output is formatted as follows:
with each entry being tab delimited.</p>
<p>| PSMId                         | score    |  q-value    | posterior_error_prob  |  peptide                       | proteinIds          |                      |                      |                         | # noqa E501 W605
|-------------------------------|----------|-------------|-----------------------|--------------------------------|---------------------|----------------------|----------------------|-------------------------| # noqa E501 W605
|     116108.15139.15139.6.dta  |  3.44016 | 0.000479928 | 7.60258e-10           | K.MVVSMTLGLHPWIANIDDTQYLAAK.R  | CNDP1_HUMAN|Q96KN2 | B4E180_HUMAN|B4E180 | A8K1K1_HUMAN|A8K1K1 | J3KRP0_HUMAN|J3KRP0    | # noqa E501 W605</p>
<p>Custom columns can be added and used as scoring input. Please see package documentation for more information.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>target_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Target PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>decoy_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Decoy PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>combined_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Combined PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory containing combined PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>psms</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p></td>
      </tr>
      <tr>
        <td><code>load_custom_score</code></td>
        <td><code>bool</code></td>
        <td><p>True/False on whether or not to load a custom score. Depends on scoring_variable.</p></td>
      </tr>
      <tr>
        <td><code>scoring_variable</code></td>
        <td><code>str</code></td>
        <td><p>String to indicate which column in the input file is to be used as the scoring input.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>parameter_file_object</code></td>
        <td><code>ProteinInferenceParameter</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a> object</p></td>
      </tr>
      <tr>
        <td><code>append_alt_from_db</code></td>
        <td><code>bool</code></td>
        <td><p>Whether or not to append alternative proteins found in the database that
are not in the input files.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">GenericReader</span><span class="p">(</span><span class="n">Reader</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class takes a percolator like target file and a percolator like decoy file</span>
<span class="sd">    and creates standard [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    Percolator Like Output is formatted as follows:</span>
<span class="sd">    with each entry being tab delimited.</span>

<span class="sd">    | PSMId                         | score    |  q-value    | posterior_error_prob  |  peptide                       | proteinIds          |                      |                      |                         | # noqa E501 W605</span>
<span class="sd">    |-------------------------------|----------|-------------|-----------------------|--------------------------------|---------------------|----------------------|----------------------|-------------------------| # noqa E501 W605</span>
<span class="sd">    |     116108.15139.15139.6.dta  |  3.44016 | 0.000479928 | 7.60258e-10           | K.MVVSMTLGLHPWIANIDDTQYLAAK.R  | CNDP1_HUMAN\|Q96KN2 | B4E180_HUMAN\|B4E180 | A8K1K1_HUMAN\|A8K1K1 | J3KRP0_HUMAN\|J3KRP0    | # noqa E501 W605</span>

<span class="sd">    Custom columns can be added and used as scoring input. Please see package documentation for more information.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>
<span class="sd">        psms (list): List of [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">        load_custom_score (bool): True/False on whether or not to load a custom score. Depends on scoring_variable.</span>
<span class="sd">        scoring_variable (str): String to indicate which column in the input file is to be used as the scoring input.</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">            [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object</span>
<span class="sd">        append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that</span>
<span class="sd">            are not in the input files.</span>



<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">PSMID</span> <span class="o">=</span> <span class="s2">&quot;PSMId&quot;</span>
    <span class="n">SCORE</span> <span class="o">=</span> <span class="s2">&quot;score&quot;</span>
    <span class="n">Q_VALUE</span> <span class="o">=</span> <span class="s2">&quot;q-value&quot;</span>
    <span class="n">POSTERIOR_ERROR_PROB</span> <span class="o">=</span> <span class="s2">&quot;posterior_error_prob&quot;</span>
    <span class="n">PEPTIDE</span> <span class="o">=</span> <span class="s2">&quot;peptide&quot;</span>
    <span class="n">PROTEIN_IDS</span> <span class="o">=</span> <span class="s2">&quot;proteinIds&quot;</span>
    <span class="n">ALTERNATIVE_PROTEINS</span> <span class="o">=</span> <span class="s2">&quot;alternative_proteins&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">digest</span><span class="p">,</span>
        <span class="n">parameter_file_object</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">            parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">                [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>
<span class="sd">            append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that</span>
<span class="sd">                are not in the input files.</span>
<span class="sd">            target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">            decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">            combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">            directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Reader: [Reader][pyproteininference.reader.Reader] object.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pyproteininference.reader.GenericReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">=</span> <span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Pulling custom column based on parameter file input for score, Column: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Pulling no custom columns based on parameter file input for score, using standard Column: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># If we select to not run inference at all</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
            <span class="c1"># Only allow 1 Protein per PSM</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">        [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">        This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">        This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; reader = pyproteininference.reader.GenericReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">            &gt;&gt;&gt; reader.read_psms()</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading in Input Files using Generic Reader...&quot;</span><span class="p">)</span>
        <span class="c1"># Read in and split by line</span>
        <span class="c1"># If target_file is a list... read them all in and concatenate...</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">t_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
                    <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">t_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_target_file</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">t_files</span><span class="p">)</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                            <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">all_target</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">ptarg</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If not just read the file...</span>
                <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_target_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="n">ptarg</span>

            <span class="c1"># Repeat for decoy file</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">d_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
                    <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">d_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_decoy_file</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">d_files</span><span class="p">)</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                            <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">all_decoy</span> <span class="o">+</span> <span class="n">pdec</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_decoy_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">pdec</span>

            <span class="c1"># Combine the lists</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">all_decoy</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">c_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
                    <span class="n">c_all</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">c_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">c_files</span><span class="p">)</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                            <span class="n">c_all</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">c_all</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">c_all</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">c_all</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="n">c_all</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">all</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>
            <span class="n">all_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">)</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">all_files</span><span class="p">:</span>
                <span class="n">psm_per_file</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">psm_per_file</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">psm_per_file</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">all</span>

        <span class="n">psms_all_filtered</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">all_psms</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">])</span>
                    <span class="n">psms_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
                    <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">])</span>
                    <span class="n">psms_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
                    <span class="k">pass</span>

        <span class="c1"># Filter by pep</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting by </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">))</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">psms_all_filtered</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">]),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cannot Sort by </span><span class="si">{}</span><span class="s2"> the values do not exist&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">))</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting by </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">))</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE_SCORE_TYPE</span><span class="p">:</span>
                <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="n">psms_all_filtered</span><span class="p">,</span>
                    <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">]),</span>
                    <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span><span class="p">:</span>
                <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="n">psms_all_filtered</span><span class="p">,</span>
                    <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">]),</span>
                    <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
        <span class="c1"># We only want to get unique peptides... using all messes up scoring...</span>
        <span class="c1"># Create Psm objects with the identifier, percscore, qvalue, pepvalue, and possible proteins...</span>

        <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

        <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of PSMs in the input data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_psms</span><span class="p">)))</span>
        <span class="n">psms_with_alternative_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_psms_with_alternative_proteins</span><span class="p">(</span><span class="n">raw_psms</span><span class="o">=</span><span class="n">all_psms</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Number of PSMs that have alternative proteins in the input data </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">psms_with_alternative_proteins</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms_with_alternative_proteins</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;No PSMs in the input have alternative proteins. &quot;</span>
                <span class="s2">&quot;Make sure your input is properly formatted. &quot;</span>
                <span class="s2">&quot;Alternative Proteins will be retrieved from the fasta database&quot;</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">psm_info</span> <span class="ow">in</span> <span class="n">all_psms</span><span class="p">:</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE</span><span class="p">]</span>
            <span class="c1"># Define the Psm...</span>
            <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
                <span class="n">psm</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
                <span class="c1"># Attempt to add variables from PSM info...</span>
                <span class="c1"># If they do not exist in the psm info then we skip...</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="c1"># If user has a custom score IE not q-value or pep_value...</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span><span class="p">:</span>
                    <span class="c1"># Then we look for it...</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">])</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEIN_IDS</span><span class="p">])</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">+</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ALTERNATIVE_PROTEINS</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">]</span>
                <span class="c1"># Remove potential Repeats</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">!=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)))</span>

                <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>

                <span class="c1"># Get PSM ID</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PSMID</span><span class="p">]</span>

                <span class="c1"># Split peptide if flanking</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                    <span class="c1"># If we have mods remove them...</span>
                    <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                    <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>
                <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">]))</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB for Proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">current_peptide</span><span class="p">,</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                            <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                        <span class="p">)</span>

                <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
                <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                    <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                    <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                    <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="c1"># Restrict to 50 possible proteins</span>
                <span class="n">psm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                    <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                    <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                    <span class="n">psm</span><span class="o">=</span><span class="n">psm</span><span class="p">,</span>
                    <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm</span><span class="p">)</span>
                <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
            <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">)))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished GenericReader.read_psms...&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_find_psms_with_alternative_proteins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_psms</span><span class="p">):</span>

        <span class="n">psms_with_alternative_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">raw_psms</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]]</span>

        <span class="k">return</span> <span class="n">psms_with_alternative_proteins</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">
















  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.GenericReader.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="p">,</span> <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
            <li><p><strong>parameter_file_object</strong> (<code>ProteinInferenceParameter</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a> object.</p></li>
            <li><p><strong>append_alt_from_db</strong> (<code>bool</code>)  Whether or not to append alternative proteins found in the database that
are not in the input files.</p></li>
            <li><p><strong>target_file</strong> (<code>str/list</code>)  Path to Target PSM result files.</p></li>
            <li><p><strong>decoy_file</strong> (<code>str/list</code>)  Path to Decoy PSM result files.</p></li>
            <li><p><strong>combined_files</strong> (<code>str/list</code>)  Path to Combined PSM result files.</p></li>
            <li><p><strong>directory</strong> (<code>str</code>)  Path to directory containing combined PSM result files.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>Reader</code>  <a class="autorefs autorefs-internal" href="#pyproteininference.reader.Reader">Reader</a> object.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span><span class="n">target_file</span> <span class="o">=</span> <span class="s2">&quot;example_target.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_file</span> <span class="o">=</span> <span class="s2">&quot;example_decoy.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pi_params</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">digest</span><span class="p">,</span>
    <span class="n">parameter_file_object</span><span class="p">,</span>
    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">            [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>
<span class="sd">        append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that</span>
<span class="sd">            are not in the input files.</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Reader: [Reader][pyproteininference.reader.Reader] object.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pyproteininference.reader.GenericReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">=</span> <span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Pulling custom column based on parameter file input for score, Column: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Pulling no custom columns based on parameter file input for score, using standard Column: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># If we select to not run inference at all</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
        <span class="c1"># Only allow 1 Protein per PSM</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span> <span class="o">=</span> <span class="mi">1</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.GenericReader.read_psms">
<code class="highlight language-python"><span class="n">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to read psms from the input files and to transform them into a list of
<a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p>
<p>This method sets the <code>psms</code> variable. Which is a list of Psm objets.</p>
<p>This method must be ran before initializing <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">GenericReader</span><span class="p">(</span><span class="n">target_file</span> <span class="o">=</span> <span class="s2">&quot;example_target.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_file</span> <span class="o">=</span> <span class="s2">&quot;example_decoy.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pi_params</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">    [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">    This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; reader = pyproteininference.reader.GenericReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">        &gt;&gt;&gt; reader.read_psms()</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading in Input Files using Generic Reader...&quot;</span><span class="p">)</span>
    <span class="c1"># Read in and split by line</span>
    <span class="c1"># If target_file is a list... read them all in and concatenate...</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="n">all_target</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">t_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
                <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">t_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_target_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">t_files</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">ptarg</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If not just read the file...</span>
            <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_target_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">)</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="n">all_target</span> <span class="o">=</span> <span class="n">ptarg</span>

        <span class="c1"># Repeat for decoy file</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="n">all_decoy</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">d_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
                <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">d_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_decoy_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">d_files</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">all_decoy</span> <span class="o">+</span> <span class="n">pdec</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_decoy_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">)</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">pdec</span>

        <span class="c1"># Combine the lists</span>
        <span class="n">all_psms</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">all_decoy</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">c_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
                <span class="n">c_all</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">c_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">c_files</span><span class="p">)</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                        <span class="n">c_all</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">c_all</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">c_all</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">)</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">c_all</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="n">c_all</span>
        <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">all</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>
        <span class="n">all_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">)</span>
        <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">all_files</span><span class="p">:</span>
            <span class="n">psm_per_file</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">psm_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">psm_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="n">psm_per_file</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">psm_per_file</span>
        <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">all</span>

    <span class="n">psms_all_filtered</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">all_psms</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span> <span class="ow">in</span> <span class="n">psms</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">])</span>
                <span class="n">psms_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
                <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">])</span>
                <span class="n">psms_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
                <span class="k">pass</span>

    <span class="c1"># Filter by pep</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting by </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">))</span>
        <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">psms_all_filtered</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">]),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Cannot Sort by </span><span class="si">{}</span><span class="s2"> the values do not exist&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting by </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">))</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">ADDITIVE_SCORE_TYPE</span><span class="p">:</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">psms_all_filtered</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">]),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">psm_score_type</span> <span class="o">==</span> <span class="n">Score</span><span class="o">.</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span><span class="p">:</span>
            <span class="n">all_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">psms_all_filtered</span><span class="p">,</span>
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">]),</span>
                <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
    <span class="c1"># We only want to get unique peptides... using all messes up scoring...</span>
    <span class="c1"># Create Psm objects with the identifier, percscore, qvalue, pepvalue, and possible proteins...</span>

    <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

    <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of PSMs in the input data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_psms</span><span class="p">)))</span>
    <span class="n">psms_with_alternative_proteins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_psms_with_alternative_proteins</span><span class="p">(</span><span class="n">raw_psms</span><span class="o">=</span><span class="n">all_psms</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Number of PSMs that have alternative proteins in the input data </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">psms_with_alternative_proteins</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psms_with_alternative_proteins</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;No PSMs in the input have alternative proteins. &quot;</span>
            <span class="s2">&quot;Make sure your input is properly formatted. &quot;</span>
            <span class="s2">&quot;Alternative Proteins will be retrieved from the fasta database&quot;</span>
        <span class="p">)</span>
    <span class="k">for</span> <span class="n">psm_info</span> <span class="ow">in</span> <span class="n">all_psms</span><span class="p">:</span>
        <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE</span><span class="p">]</span>
        <span class="c1"># Define the Psm...</span>
        <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
            <span class="n">psm</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
            <span class="c1"># Attempt to add variables from PSM info...</span>
            <span class="c1"># If they do not exist in the psm info then we skip...</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="c1"># If user has a custom score IE not q-value or pep_value...</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_custom_score</span><span class="p">:</span>
                <span class="c1"># Then we look for it...</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">custom_score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">scoring_variable</span><span class="p">])</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEIN_IDS</span><span class="p">])</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">+</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ALTERNATIVE_PROTEINS</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">]</span>
            <span class="c1"># Remove potential Repeats</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">!=</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
                <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)))</span>

            <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>

            <span class="c1"># Get PSM ID</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PSMID</span><span class="p">]</span>

            <span class="c1"># Split peptide if flanking</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                <span class="c1"># If we have mods remove them...</span>
                <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>
            <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">]))</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB for Proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">current_peptide</span><span class="p">,</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                    <span class="p">)</span>

            <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
            <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Restrict to 50 possible proteins</span>
            <span class="n">psm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                <span class="n">psm</span><span class="o">=</span><span class="n">psm</span><span class="p">,</span>
                <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psm</span><span class="p">)</span>
            <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
        <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">)))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished GenericReader.read_psms...&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.reader.PercolatorReader">
        <code>
PercolatorReader            (<a class="autorefs autorefs-internal" title="pyproteininference.reader.Reader" href="#pyproteininference.reader.Reader">Reader</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>The following class takes a percolator target file and a percolator decoy file
or combined files/directory and creates standard <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.
This reader class is used as input for <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>
<p>Percolator Output is formatted as follows:
with each entry being tab delimited.</p>
<p>| PSMId                         | score    |  q-value    | posterior_error_prob  |  peptide                       | proteinIds          |                      |                      |                         | # noqa E501 W605
|-------------------------------|----------|-------------|-----------------------|--------------------------------|---------------------|----------------------|----------------------|-------------------------| # noqa E501 W605
|     116108.15139.15139.6.dta  |  3.44016 | 0.000479928 | 7.60258e-10           | K.MVVSMTLGLHPWIANIDDTQYLAAK.R  | CNDP1_HUMAN|Q96KN2 | B4E180_HUMAN|B4E180 | A8K1K1_HUMAN|A8K1K1 | J3KRP0_HUMAN|J3KRP0    | # noqa E501 W605</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>target_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Target PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>decoy_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Decoy PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>combined_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Combined PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory containing combined PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>PSMID_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the PSMId from the input files.</p></td>
      </tr>
      <tr>
        <td><code>PERC_SCORE_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the Percolator score from the input files.</p></td>
      </tr>
      <tr>
        <td><code>Q_VALUE_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the q-value from the input files.</p></td>
      </tr>
      <tr>
        <td><code>POSTERIOR_ERROR_PROB_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the posterior error probability from the input files.</p></td>
      </tr>
      <tr>
        <td><code>PEPTIDE_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the peptides from the input files.</p></td>
      </tr>
      <tr>
        <td><code>PROTEINIDS_INDEX</code></td>
        <td><code>int</code></td>
        <td><p>Index of the proteins from the input files.</p></td>
      </tr>
      <tr>
        <td><code>psms</code></td>
        <td><code>list</code></td>
        <td><p>List of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">PercolatorReader</span><span class="p">(</span><span class="n">Reader</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following class takes a percolator target file and a percolator decoy file</span>
<span class="sd">    or combined files/directory and creates standard [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">    This reader class is used as input for [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    Percolator Output is formatted as follows:</span>
<span class="sd">    with each entry being tab delimited.</span>

<span class="sd">    | PSMId                         | score    |  q-value    | posterior_error_prob  |  peptide                       | proteinIds          |                      |                      |                         | # noqa E501 W605</span>
<span class="sd">    |-------------------------------|----------|-------------|-----------------------|--------------------------------|---------------------|----------------------|----------------------|-------------------------| # noqa E501 W605</span>
<span class="sd">    |     116108.15139.15139.6.dta  |  3.44016 | 0.000479928 | 7.60258e-10           | K.MVVSMTLGLHPWIANIDDTQYLAAK.R  | CNDP1_HUMAN\|Q96KN2 | B4E180_HUMAN\|B4E180 | A8K1K1_HUMAN\|A8K1K1 | J3KRP0_HUMAN\|J3KRP0    | # noqa E501 W605</span>

<span class="sd">    Attributes:</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>
<span class="sd">        PSMID_INDEX (int): Index of the PSMId from the input files.</span>
<span class="sd">        PERC_SCORE_INDEX (int): Index of the Percolator score from the input files.</span>
<span class="sd">        Q_VALUE_INDEX (int): Index of the q-value from the input files.</span>
<span class="sd">        POSTERIOR_ERROR_PROB_INDEX (int): Index of the posterior error probability from the input files.</span>
<span class="sd">        PEPTIDE_INDEX (int): Index of the peptides from the input files.</span>
<span class="sd">        PROTEINIDS_INDEX (int): Index of the proteins from the input files.</span>
<span class="sd">        psms (list): List of [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">PSMID_INDEX</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">PERC_SCORE_INDEX</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">Q_VALUE_INDEX</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">POSTERIOR_ERROR_PROB_INDEX</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="n">PEPTIDE_INDEX</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">PROTEINIDS_INDEX</span> <span class="o">=</span> <span class="mi">5</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">digest</span><span class="p">,</span>
        <span class="n">parameter_file_object</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">            parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">                [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter].</span>
<span class="sd">            append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that</span>
<span class="sd">                are not in the input files.</span>
<span class="sd">            target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">            decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">            combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">            directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Reader: [Reader][pyproteininference.reader.Reader] object.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; pyproteininference.reader.PercolatorReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;, digest=digest,parameter_file_object=pi_params)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>
        <span class="c1"># Define Indicies based on input</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">        [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">        This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">        This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; reader = pyproteininference.reader.PercolatorReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">            &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">            &gt;&gt;&gt; reader.read_psms()</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Read in and split by line</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
            <span class="c1"># If target_file is a list... read them all in and concatenate...</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">t_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">t_files</span><span class="p">)</span>
                    <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">t_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_target_file</span><span class="p">:</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="k">del</span> <span class="n">ptarg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">all_target</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">ptarg</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
                <span class="c1"># If not just read the file...</span>
                <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_target_file</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">ptarg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="n">ptarg</span>

            <span class="c1"># Repeat for decoy file</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">d_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">d_files</span><span class="p">)</span>
                    <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">d_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_decoy_file</span><span class="p">:</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="k">del</span> <span class="n">pdec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">all_decoy</span> <span class="o">+</span> <span class="n">pdec</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
                <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_decoy_file</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">pdec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">pdec</span>

            <span class="c1"># Combine the lists</span>
            <span class="n">perc_all</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">all_decoy</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                    <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_files</span><span class="p">:</span>
                        <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_files</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                    <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">combined_psm_result_rows</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
                <span class="c1"># If not just read the file...</span>
                <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_files</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_files</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="n">combined_psm_result_rows</span>
            <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">all</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>

            <span class="n">all_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">)</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">all_files</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
                <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_file</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">combined_psm_result_rows</span>
            <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">all</span>

        <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

        <span class="n">perc_all_filtered</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">perc_all</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">])</span>
                <span class="n">perc_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
                <span class="k">pass</span>

        <span class="c1"># Filter by pep</span>
        <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">perc_all_filtered</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">]),</span>
            <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
        <span class="c1"># We only want to get unique peptides... using all messes up scoring...</span>
        <span class="c1"># Create Psm objects with the identifier, percscore, qvalue, pepvalue, and possible proteins...</span>

        <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">perc_all</span><span class="p">)))</span>
        <span class="k">for</span> <span class="n">psm_info</span> <span class="ow">in</span> <span class="n">perc_all</span><span class="p">:</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_INDEX</span><span class="p">]</span>
            <span class="c1"># Define the Psm...</span>
            <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
                <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
                <span class="c1"># Add all the attributes</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PERC_SCORE_INDEX</span><span class="p">])</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE_INDEX</span><span class="p">])</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">])</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
                    <span class="n">poss_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEINIDS_INDEX</span><span class="p">]]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">poss_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEINIDS_INDEX</span> <span class="p">:])))</span>  <span class="c1"># noqa E203</span>
                    <span class="n">poss_proteins</span> <span class="o">=</span> <span class="n">poss_proteins</span><span class="p">[:</span> <span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">]</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">poss_proteins</span>  <span class="c1"># Restrict to 50 total possible proteins...</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PSMID_INDEX</span><span class="p">]</span>
                <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">poss_proteins</span><span class="p">)</span>

                <span class="c1"># Split peptide if flanking</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                    <span class="c1"># If we have mods remove them...</span>
                    <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                    <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

                <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                        <span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">])</span>
                    <span class="p">)</span>  <span class="c1"># This peptide needs to be scrubbed of Mods...</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB with the following proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">current_peptide</span><span class="p">,</span>
                            <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">),</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                            <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                        <span class="p">)</span>

                <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
                <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                    <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                    <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                    <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="c1"># Restrict to 50 possible proteins</span>
                <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                    <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                    <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                    <span class="n">psm</span><span class="o">=</span><span class="n">combined_psm_result_rows</span><span class="p">,</span>
                    <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="c1"># Remove blank alt proteins</span>
                <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                <span class="p">]</span>

                <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">combined_psm_result_rows</span><span class="p">)</span>
                <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
            <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">)))</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">















  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.PercolatorReader.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="p">,</span> <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
            <li><p><strong>parameter_file_object</strong> (<code>ProteinInferenceParameter</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a>.</p></li>
            <li><p><strong>append_alt_from_db</strong> (<code>bool</code>)  Whether or not to append alternative proteins found in the database that
are not in the input files.</p></li>
            <li><p><strong>target_file</strong> (<code>str/list</code>)  Path to Target PSM result files.</p></li>
            <li><p><strong>decoy_file</strong> (<code>str/list</code>)  Path to Decoy PSM result files.</p></li>
            <li><p><strong>combined_files</strong> (<code>str/list</code>)  Path to Combined PSM result files.</p></li>
            <li><p><strong>directory</strong> (<code>str</code>)  Path to directory containing combined PSM result files.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>Reader</code>  <a class="autorefs autorefs-internal" href="#pyproteininference.reader.Reader">Reader</a> object.</p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">PercolatorReader</span><span class="p">(</span><span class="n">target_file</span> <span class="o">=</span> <span class="s2">&quot;example_target.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_file</span> <span class="o">=</span> <span class="s2">&quot;example_decoy.txt&quot;</span><span class="p">,</span> <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span><span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pi_params</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">digest</span><span class="p">,</span>
    <span class="n">parameter_file_object</span><span class="p">,</span>
    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">            [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter].</span>
<span class="sd">        append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that</span>
<span class="sd">            are not in the input files.</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Reader: [Reader][pyproteininference.reader.Reader] object.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; pyproteininference.reader.PercolatorReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;, digest=digest,parameter_file_object=pi_params)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>
    <span class="c1"># Define Indicies based on input</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_input</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.PercolatorReader.read_psms">
<code class="highlight language-python"><span class="n">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to read psms from the input files and to transform them into a list of
<a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p>
<p>This method sets the <code>psms</code> variable. Which is a list of Psm objets.</p>
<p>This method must be ran before initializing <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">reader</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">reader</span><span class="o">.</span><span class="n">PercolatorReader</span><span class="p">(</span><span class="n">target_file</span> <span class="o">=</span> <span class="s2">&quot;example_target.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">decoy_file</span> <span class="o">=</span> <span class="s2">&quot;example_decoy.txt&quot;</span><span class="p">,</span>
<span class="o">&gt;&gt;&gt;</span>     <span class="n">digest</span><span class="o">=</span><span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="o">=</span><span class="n">pi_params</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">reader</span><span class="o">.</span><span class="n">read_psms</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">    [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">    This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; reader = pyproteininference.reader.PercolatorReader(target_file = &quot;example_target.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     decoy_file = &quot;example_decoy.txt&quot;,</span>
<span class="sd">        &gt;&gt;&gt;     digest=digest, parameter_file_object=pi_params)</span>
<span class="sd">        &gt;&gt;&gt; reader.read_psms()</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Read in and split by line</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
        <span class="c1"># If target_file is a list... read them all in and concatenate...</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="n">all_target</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">t_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">t_files</span><span class="p">)</span>
                <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">t_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_target_file</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">ptarg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">all_target</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">ptarg</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
            <span class="c1"># If not just read the file...</span>
            <span class="n">ptarg</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_target_file</span><span class="p">:</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_target_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">ptarg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">ptarg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">all_target</span> <span class="o">=</span> <span class="n">ptarg</span>

        <span class="c1"># Repeat for decoy file</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="n">all_decoy</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">d_files</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">d_files</span><span class="p">)</span>
                <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">d_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_decoy_file</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">pdec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">all_decoy</span> <span class="o">+</span> <span class="n">pdec</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">:</span>
            <span class="n">pdec</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_decoy_file</span><span class="p">:</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_decoy_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">pdec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">pdec</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">all_decoy</span> <span class="o">=</span> <span class="n">pdec</span>

        <span class="c1"># Combine the lists</span>
        <span class="n">perc_all</span> <span class="o">=</span> <span class="n">all_target</span> <span class="o">+</span> <span class="n">all_decoy</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_files</span><span class="p">:</span>
                    <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_files</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                        <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
                <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">combined_psm_result_rows</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">:</span>
            <span class="c1"># If not just read the file...</span>
            <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_files</span><span class="p">:</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_files</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="n">combined_psm_result_rows</span>
        <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">all</span>

    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>

        <span class="n">all_files</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">)</span>
        <span class="nb">all</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">all_files</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
            <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">perc_file</span><span class="p">:</span>
                <span class="n">spamreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">perc_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spamreader</span><span class="p">:</span>
                    <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
            <span class="k">del</span> <span class="n">combined_psm_result_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="nb">all</span> <span class="o">=</span> <span class="nb">all</span> <span class="o">+</span> <span class="n">combined_psm_result_rows</span>
        <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">all</span>

    <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

    <span class="n">perc_all_filtered</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">perc_all</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="nb">float</span><span class="p">(</span><span class="n">psms</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">])</span>
            <span class="n">perc_all_filtered</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># noqa F841</span>
            <span class="k">pass</span>

    <span class="c1"># Filter by pep</span>
    <span class="n">perc_all</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="n">perc_all_filtered</span><span class="p">,</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">]),</span>
        <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
    <span class="c1"># We only want to get unique peptides... using all messes up scoring...</span>
    <span class="c1"># Create Psm objects with the identifier, percscore, qvalue, pepvalue, and possible proteins...</span>

    <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">perc_all</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">psm_info</span> <span class="ow">in</span> <span class="n">perc_all</span><span class="p">:</span>
        <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PEPTIDE_INDEX</span><span class="p">]</span>
        <span class="c1"># Define the Psm...</span>
        <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
            <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
            <span class="c1"># Add all the attributes</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PERC_SCORE_INDEX</span><span class="p">])</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">Q_VALUE_INDEX</span><span class="p">])</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">POSTERIOR_ERROR_PROB_INDEX</span><span class="p">])</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
                <span class="n">poss_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEINIDS_INDEX</span><span class="p">]]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">poss_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PROTEINIDS_INDEX</span> <span class="p">:])))</span>  <span class="c1"># noqa E203</span>
                <span class="n">poss_proteins</span> <span class="o">=</span> <span class="n">poss_proteins</span><span class="p">[:</span> <span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">]</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">poss_proteins</span>  <span class="c1"># Restrict to 50 total possible proteins...</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">psm_info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PSMID_INDEX</span><span class="p">]</span>
            <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">poss_proteins</span><span class="p">)</span>

            <span class="c1"># Split peptide if flanking</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                <span class="c1"># If we have mods remove them...</span>
                <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

            <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">])</span>
                <span class="p">)</span>  <span class="c1"># This peptide needs to be scrubbed of Mods...</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB with the following proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">current_peptide</span><span class="p">,</span>
                        <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                    <span class="p">)</span>

            <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
            <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Restrict to 50 possible proteins</span>
            <span class="n">combined_psm_result_rows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                <span class="n">psm</span><span class="o">=</span><span class="n">combined_psm_result_rows</span><span class="p">,</span>
                <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Remove blank alt proteins</span>
            <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">combined_psm_result_rows</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
            <span class="p">]</span>

            <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">combined_psm_result_rows</span><span class="p">)</span>
            <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
        <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of PSM Data: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">psms</span><span class="p">)))</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.reader.ProteologicPostSearchReader">
        <code>
ProteologicPostSearchReader            (<a class="autorefs autorefs-internal" title="pyproteininference.reader.Reader" href="#pyproteininference.reader.Reader">Reader</a>)
        </code>



</h3>

    <div class="doc doc-contents ">

      <p>This class is used to read from post processing proteologic logical object.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>proteologic_object</code></td>
        <td><code>list</code></td>
        <td><p>List of proteologic post search objects.</p></td>
      </tr>
      <tr>
        <td><code>search_id</code></td>
        <td><code>int</code></td>
        <td><p>Search ID or Search IDs associated with the data.</p></td>
      </tr>
      <tr>
        <td><code>postsearch_id</code></td>
        <td><code>int</code></td>
        <td><p>PostSearch ID or PostSearch IDs associated with the data.</p></td>
      </tr>
      <tr>
        <td><code>digest</code></td>
        <td><code>Digest</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></td>
      </tr>
      <tr>
        <td><code>parameter_file_object</code></td>
        <td><code>ProteinInferenceParameter</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a> object.</p></td>
      </tr>
      <tr>
        <td><code>append_alt_from_db</code></td>
        <td><code>bool</code></td>
        <td><p>Whether or not to append alternative proteins found in the database
that are not in the input files.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ProteologicPostSearchReader</span><span class="p">(</span><span class="n">Reader</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class is used to read from post processing proteologic logical object.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        proteologic_object (list): List of proteologic post search objects.</span>
<span class="sd">        search_id (int): Search ID or Search IDs associated with the data.</span>
<span class="sd">        postsearch_id (int): PostSearch ID or PostSearch IDs associated with the data.</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">            [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>
<span class="sd">        append_alt_from_db (bool): Whether or not to append alternative proteins found in the database</span>
<span class="sd">            that are not in the input files.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">proteologic_object</span><span class="p">,</span>
        <span class="n">search_id</span><span class="p">,</span>
        <span class="n">postsearch_id</span><span class="p">,</span>
        <span class="n">digest</span><span class="p">,</span>
        <span class="n">parameter_file_object</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            proteologic_object (list): List of proteologic post search objects.</span>
<span class="sd">            search_id (int): Search ID or Search IDs associated with the data.</span>
<span class="sd">            postsearch_id: PostSearch ID or PostSearch IDs associated with the data.</span>
<span class="sd">            digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">            parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">                [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>
<span class="sd">            append_alt_from_db (bool): Whether or not to append alternative proteins found in the database</span>
<span class="sd">                that are not in the input files.</span>


<span class="sd">        Returns:</span>
<span class="sd">            object:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span> <span class="o">=</span> <span class="n">proteologic_object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="n">search_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">postsearch_id</span> <span class="o">=</span> <span class="n">postsearch_id</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>

    <span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">        [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">        This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">        This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading in data from Proteologic...&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
            <span class="n">list_of_psms</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">p_objs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">p_objs</span><span class="o">.</span><span class="n">physical_object</span><span class="o">.</span><span class="n">psm_sets</span><span class="p">:</span>
                    <span class="n">list_of_psms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">list_of_psms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="o">.</span><span class="n">physical_object</span><span class="o">.</span><span class="n">psm_sets</span>

        <span class="c1"># Sort this by posterior error prob...</span>
        <span class="n">list_of_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">list_of_psms</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">))</span>

        <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

        <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
        <span class="c1"># Peptide tracker is used because we only want UNIQUE peptides...</span>
        <span class="c1"># The data is sorted by percolator score... or at least it should be...</span>
        <span class="c1"># Or sorted by posterior error probability</span>

        <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">list_of_psms</span><span class="p">:</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">sequence</span>
            <span class="c1"># Define the Psm...</span>
            <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
                <span class="n">p</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
                <span class="c1"># Add all the attributes</span>
                <span class="n">p</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># Will be stored in table in future I think...</span>
                <span class="n">p</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">peps</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
                <span class="n">p</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">peps</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span><span class="p">:</span>
                    <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">protein</span><span class="p">]</span> <span class="o">+</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span>

                <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">))</span>
                <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
                <span class="n">p</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">spectrum</span><span class="o">.</span><span class="n">spectrum_identifier</span>

                <span class="c1"># Split peptide if flanking</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                    <span class="c1"># If we have mods remove them...</span>
                    <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                    <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

                <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                        <span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">])</span>
                    <span class="p">)</span>  <span class="c1"># This peptide needs to be scrubbed of Mods...</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB with the following proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">current_peptide</span><span class="p">,</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                            <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                        <span class="p">)</span>

                <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
                <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                    <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                    <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                    <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="c1"># Restrict to 50 possible proteins... and append alt proteins from db</span>
                <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                    <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                    <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                    <span class="n">psm</span><span class="o">=</span><span class="n">p</span><span class="p">,</span>
                    <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
                <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

                <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
            <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished reading in data from Proteologic...&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">









  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.ProteologicPostSearchReader.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">proteologic_object</span><span class="p">,</span> <span class="n">search_id</span><span class="p">,</span> <span class="n">postsearch_id</span><span class="p">,</span> <span class="n">digest</span><span class="p">,</span> <span class="n">parameter_file_object</span><span class="p">,</span> <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>proteologic_object</strong> (<code>list</code>)  List of proteologic post search objects.</p></li>
            <li><p><strong>search_id</strong> (<code>int</code>)  Search ID or Search IDs associated with the data.</p></li>
            <li><p><strong>postsearch_id</strong>  PostSearch ID or PostSearch IDs associated with the data.</p></li>
            <li><p><strong>digest</strong> (<code>Digest</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.in_silico_digest.Digest">Digest Object</a>.</p></li>
            <li><p><strong>parameter_file_object</strong> (<code>ProteinInferenceParameter</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a> object.</p></li>
            <li><p><strong>append_alt_from_db</strong> (<code>bool</code>)  Whether or not to append alternative proteins found in the database
that are not in the input files.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
    <th class="field-name">Returns:</th>
    <td class="field-body">
        <ul class="first simple">
          <li><p><code>object</code>  </p></li>
        </ul>
    </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">proteologic_object</span><span class="p">,</span>
    <span class="n">search_id</span><span class="p">,</span>
    <span class="n">postsearch_id</span><span class="p">,</span>
    <span class="n">digest</span><span class="p">,</span>
    <span class="n">parameter_file_object</span><span class="p">,</span>
    <span class="n">append_alt_from_db</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        proteologic_object (list): List of proteologic post search objects.</span>
<span class="sd">        search_id (int): Search ID or Search IDs associated with the data.</span>
<span class="sd">        postsearch_id: PostSearch ID or PostSearch IDs associated with the data.</span>
<span class="sd">        digest (Digest): [Digest Object][pyproteininference.in_silico_digest.Digest].</span>
<span class="sd">        parameter_file_object (ProteinInferenceParameter):</span>
<span class="sd">            [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>
<span class="sd">        append_alt_from_db (bool): Whether or not to append alternative proteins found in the database</span>
<span class="sd">            that are not in the input files.</span>


<span class="sd">    Returns:</span>
<span class="sd">        object:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span> <span class="o">=</span> <span class="n">proteologic_object</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">search_id</span> <span class="o">=</span> <span class="n">search_id</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">postsearch_id</span> <span class="o">=</span> <span class="n">postsearch_id</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span> <span class="o">=</span> <span class="n">digest</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span> <span class="o">=</span> <span class="n">append_alt_from_db</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span> <span class="o">=</span> <span class="n">parameter_file_object</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.ProteologicPostSearchReader.read_psms">
<code class="highlight language-python"><span class="n">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to read psms from the input files and to transform them into a list of
<a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p>
<p>This method sets the <code>psms</code> variable. Which is a list of Psm objets.</p>
<p>This method must be ran before initializing <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">read_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to read psms from the input files and to transform them into a list of</span>
<span class="sd">    [Psm][pyproteininference.physical.Psm] objects.</span>

<span class="sd">    This method sets the `psms` variable. Which is a list of Psm objets.</span>

<span class="sd">    This method must be ran before initializing [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading in data from Proteologic...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
        <span class="n">list_of_psms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">p_objs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">psms</span> <span class="ow">in</span> <span class="n">p_objs</span><span class="o">.</span><span class="n">physical_object</span><span class="o">.</span><span class="n">psm_sets</span><span class="p">:</span>
                <span class="n">list_of_psms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">psms</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">list_of_psms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">proteologic_object</span><span class="o">.</span><span class="n">physical_object</span><span class="o">.</span><span class="n">psm_sets</span>

    <span class="c1"># Sort this by posterior error prob...</span>
    <span class="n">list_of_psms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">list_of_psms</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">))</span>

    <span class="n">peptide_to_protein_dictionary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span>

    <span class="n">list_of_psm_objects</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">peptide_tracker</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">all_sp_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">swiss_prot_protein_set</span><span class="p">)</span>
    <span class="c1"># Peptide tracker is used because we only want UNIQUE peptides...</span>
    <span class="c1"># The data is sorted by percolator score... or at least it should be...</span>
    <span class="c1"># Or sorted by posterior error probability</span>

    <span class="n">initial_poss_prots</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">peps</span> <span class="ow">in</span> <span class="n">list_of_psms</span><span class="p">:</span>
        <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">sequence</span>
        <span class="c1"># Define the Psm...</span>
        <span class="k">if</span> <span class="n">current_peptide</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peptide_tracker</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">Psm</span><span class="p">(</span><span class="n">identifier</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>
            <span class="c1"># Add all the attributes</span>
            <span class="n">p</span><span class="o">.</span><span class="n">percscore</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># Will be stored in table in future I think...</span>
            <span class="n">p</span><span class="o">.</span><span class="n">qvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">peps</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">q_value</span><span class="p">)</span>
            <span class="n">p</span><span class="o">.</span><span class="n">pepvalue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">peps</span><span class="o">.</span><span class="n">psm_filter</span><span class="o">.</span><span class="n">pepvalue</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span><span class="p">:</span>
                <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">peps</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">protein</span><span class="p">]</span> <span class="o">+</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">alternative_proteins</span>

            <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">))</span>
            <span class="n">input_poss_prots</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
            <span class="n">p</span><span class="o">.</span><span class="n">psm_id</span> <span class="o">=</span> <span class="n">peps</span><span class="o">.</span><span class="n">spectrum</span><span class="o">.</span><span class="n">spectrum_identifier</span>

            <span class="c1"># Split peptide if flanking</span>
            <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">split_peptide</span><span class="p">(</span><span class="n">peptide_string</span><span class="o">=</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">isalpha</span><span class="p">():</span>
                <span class="c1"># If we have mods remove them...</span>
                <span class="n">peptide_string</span> <span class="o">=</span> <span class="n">current_peptide</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">stripped_peptide</span> <span class="o">=</span> <span class="n">Psm</span><span class="o">.</span><span class="n">remove_peptide_mods</span><span class="p">(</span><span class="n">peptide_string</span><span class="p">)</span>
                <span class="n">current_peptide</span> <span class="o">=</span> <span class="n">stripped_peptide</span>

            <span class="c1"># Add the other possible_proteins from insilicodigest here...</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="n">peptide_to_protein_dictionary</span><span class="p">[</span><span class="n">current_peptide</span><span class="p">])</span>
                <span class="p">)</span>  <span class="c1"># This peptide needs to be scrubbed of Mods...</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">current_alt_proteins</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="s2">&quot;Peptide </span><span class="si">{}</span><span class="s2"> was not found in the supplied DB with the following proteins </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">current_peptide</span><span class="p">,</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">poss_prot</span> <span class="ow">in</span> <span class="n">p</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">peptide_to_protein_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">digest</span><span class="o">.</span><span class="n">protein_to_peptide_dictionary</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">poss_prot</span><span class="p">,</span> <span class="nb">set</span><span class="p">())</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                        <span class="s2">&quot;Adding Peptide </span><span class="si">{}</span><span class="s2"> and Protein </span><span class="si">{}</span><span class="s2"> to Digest dictionaries&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">,</span> <span class="n">poss_prot</span><span class="p">)</span>
                    <span class="p">)</span>

            <span class="c1"># Sort Alt Proteins by Swissprot then Trembl...</span>
            <span class="n">identifiers_sorted</span> <span class="o">=</span> <span class="n">DataStore</span><span class="o">.</span><span class="n">sort_protein_strings</span><span class="p">(</span>
                <span class="n">protein_string_list</span><span class="o">=</span><span class="n">current_alt_proteins</span><span class="p">,</span>
                <span class="n">sp_proteins</span><span class="o">=</span><span class="n">all_sp_proteins</span><span class="p">,</span>
                <span class="n">decoy_symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="o">.</span><span class="n">decoy_symbol</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Restrict to 50 possible proteins... and append alt proteins from db</span>
            <span class="n">p</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fix_alternative_proteins</span><span class="p">(</span>
                <span class="n">append_alt_from_db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">append_alt_from_db</span><span class="p">,</span>
                <span class="n">identifiers_sorted</span><span class="o">=</span><span class="n">identifiers_sorted</span><span class="p">,</span>
                <span class="n">max_proteins</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span><span class="p">,</span>
                <span class="n">psm</span><span class="o">=</span><span class="n">p</span><span class="p">,</span>
                <span class="n">parameter_file_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">parameter_file_object</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">list_of_psm_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
            <span class="n">peptide_tracker</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">current_peptide</span><span class="p">)</span>

            <span class="n">initial_poss_prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">input_poss_prots</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">psms</span> <span class="o">=</span> <span class="n">list_of_psm_objects</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_check_initial_database_overlap</span><span class="p">(</span>
        <span class="n">initial_possible_proteins</span><span class="o">=</span><span class="n">initial_poss_prots</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">initial_protein_peptide_map</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished reading in data from Proteologic...&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>



  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.reader.Reader">
        <code>
Reader        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Main Reader Class which is parent to all reader subclasses.</p>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>target_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Target PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>decoy_file</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Decoy PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>combined_files</code></td>
        <td><code>str/list</code></td>
        <td><p>Path to Combined PSM result files.</p></td>
      </tr>
      <tr>
        <td><code>directory</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory containing combined PSM result files.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Reader</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main Reader Class which is parent to all reader subclasses.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">MAX_ALLOWED_ALTERNATIVE_PROTEINS</span> <span class="o">=</span> <span class="mi">50</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">        Args:</span>
<span class="sd">            target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">            decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">            combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">            directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>

    <span class="k">def</span> <span class="nf">get_alternative_proteins_from_input</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to get the alternative proteins from the input files.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="kc">None</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="c1"># Sort the alternative proteins - when they are read in they become unsorted</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">return</span> <span class="n">row</span>

    <span class="k">def</span> <span class="nf">_validate_input</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method to validate the input to Reader.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as target_file and decoy_file&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as combined_files&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Validating input as combined_directory&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;To run Protein inference please supply either: &quot;</span>
                <span class="s2">&quot;(1) either one or multiple target_files and decoy_files, &quot;</span>
                <span class="s2">&quot;(2) either one or multiple combined_files that include target and decoy data&quot;</span>
                <span class="s2">&quot;(3) a combined_directory that contains combined target/decoy files (combined_directory)&quot;</span>
            <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_fix_alternative_proteins</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span>
        <span class="n">append_alt_from_db</span><span class="p">,</span>
        <span class="n">identifiers_sorted</span><span class="p">,</span>
        <span class="n">max_proteins</span><span class="p">,</span>
        <span class="n">psm</span><span class="p">,</span>
        <span class="n">parameter_file_object</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method to fix the alternative proteins variable for a given</span>
<span class="sd">         [Psm][pyproteininference.physical.Psm] object.</span>

<span class="sd">        Args:</span>
<span class="sd">            append_alt_from_db (bool): Whether or not to append alternative proteins found in the database that are</span>
<span class="sd">                not in the input files.</span>
<span class="sd">            identifiers_sorted (list): List of sorted Protein Strings for the given Psm.</span>
<span class="sd">            max_proteins (int): Maximum number of proteins that a [Psm][pyproteininference.physical.Psm]</span>
<span class="sd">                is allowed to map to.</span>
<span class="sd">            psm: (Psm): [Psm][pyproteininference.physical.Psm] object of interest.</span>
<span class="sd">            parameter_file_object: (ProteinInferenceParameter):</span>
<span class="sd">                [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter].</span>

<span class="sd">        Returns:</span>
<span class="sd">            pyproteininference.physical.Psm: [Psm][pyproteininference.physical.Psm] with alternative proteins fixed.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># If we are appending alternative proteins from the db</span>
        <span class="k">if</span> <span class="n">append_alt_from_db</span><span class="p">:</span>
            <span class="c1"># Loop over the Identifiers from the DB These are identifiers that contain the current peptide</span>
            <span class="k">for</span> <span class="n">alt_proteins</span> <span class="ow">in</span> <span class="n">identifiers_sorted</span><span class="p">[:</span><span class="n">max_proteins</span><span class="p">]:</span>
                <span class="c1"># If the identifier is not already in possible proteins</span>
                <span class="c1"># and if then len of poss prot is less than the max...</span>
                <span class="c1"># Then append</span>
                <span class="k">if</span> <span class="n">alt_proteins</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">max_proteins</span><span class="p">:</span>
                    <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt_proteins</span><span class="p">)</span>
        <span class="c1"># Next if the len of possible proteins is greater than max then restrict the list length...</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_proteins</span><span class="p">:</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_proteins</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="c1"># If no inference only select first poss protein</span>
        <span class="k">if</span> <span class="n">parameter_file_object</span><span class="o">.</span><span class="n">inference_type</span> <span class="o">==</span> <span class="n">Inference</span><span class="o">.</span><span class="n">FIRST_PROTEIN</span><span class="p">:</span>
            <span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">psm</span><span class="o">.</span><span class="n">possible_proteins</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

        <span class="k">return</span> <span class="n">psm</span>

    <span class="k">def</span> <span class="nf">_check_initial_database_overlap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">initial_possible_proteins</span><span class="p">,</span> <span class="n">initial_protein_peptide_map</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Internal method that checks to make sure there is at least some overlap between proteins in the input files</span>
<span class="sd">        And the proteins in the database digestion.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">initial_protein_peptide_map</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">input_protein_ids_flat</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">protein</span> <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">initial_possible_proteins</span> <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">group</span><span class="p">])</span>

            <span class="n">digest_proteins</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">initial_protein_peptide_map</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

            <span class="n">intersection</span> <span class="o">=</span> <span class="n">input_protein_ids_flat</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">digest_proteins</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">intersection</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;The Intersection of Protein Identifiers between the database digest &quot;</span>
                    <span class="s2">&quot;and the input files is zero. Please consider setting id_splitting to True. &quot;</span>
                    <span class="s2">&quot;Or make sure that the identifiers in the input files and database file match. &quot;</span>
                    <span class="s2">&quot;Example Protein Identifier from input file &#39;</span><span class="si">{}</span><span class="s2">&#39;.&quot;</span>
                    <span class="s2">&quot;Example Protein Identifier from database file &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">list</span><span class="p">(</span><span class="n">input_protein_ids_flat</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">digest_proteins</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of matching proteins from database and input files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">intersection</span><span class="p">)))</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of proteins from database file: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">digest_proteins</span><span class="p">)))</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of proteins from input files: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">input_protein_ids_flat</span><span class="p">)))</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">










  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.Reader.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>target_file</strong> (<code>str/list</code>)  Path to Target PSM result files.</p></li>
            <li><p><strong>decoy_file</strong> (<code>str/list</code>)  Path to Decoy PSM result files.</p></li>
            <li><p><strong>combined_files</strong> (<code>str/list</code>)  Path to Combined PSM result files.</p></li>
            <li><p><strong>directory</strong> (<code>str</code>)  Path to directory containing combined PSM result files.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">decoy_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">combined_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        target_file (str/list): Path to Target PSM result files.</span>
<span class="sd">        decoy_file (str/list): Path to Decoy PSM result files.</span>
<span class="sd">        combined_files (str/list): Path to Combined PSM result files.</span>
<span class="sd">        directory (str): Path to directory containing combined PSM result files.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">target_file</span> <span class="o">=</span> <span class="n">target_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">decoy_file</span> <span class="o">=</span> <span class="n">decoy_file</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">combined_files</span> <span class="o">=</span> <span class="n">combined_files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="n">directory</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.reader.Reader.get_alternative_proteins_from_input">
<code class="highlight language-python"><span class="n">get_alternative_proteins_from_input</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>Method to get the alternative proteins from the input files.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/reader.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">get_alternative_proteins_from_input</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Method to get the alternative proteins from the input files.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="kc">None</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="c1"># Sort the alternative proteins - when they are read in they become unsorted</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alternative_proteins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">return</span> <span class="n">row</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>



  <div class="doc doc-object doc-module">



<h2 class="doc doc-heading" id="pyproteininference.scoring">
        <code>scoring</code>



</h2>

    <div class="doc doc-contents ">




  <div class="doc doc-children">








  <div class="doc doc-object doc-class">



<h3 class="doc doc-heading" id="pyproteininference.scoring.Score">
        <code>
Score        </code>



</h3>

    <div class="doc doc-contents ">

      <p>Score class that contains methods to do a variety of scoring methods on the
<a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects
contained inside of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects.</p>
<p>Methods in the class loop over each Protein object and creates a protein "score" variable using the Psm object
scores.</p>
<p>Methods score all proteins from <code>scoring_input</code> from <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.
The PSM score that is used is determined from
<a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore.create_scoring_input">create_scoring_input</a>.</p>
<p>Each scoring method will set the following attributes for
the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore object</a>.</p>
<ol>
<li><code>score_method</code>; This is the full name of the score method.</li>
<li><code>short_score_method</code>; This is the short name of the score method.</li>
<li><code>scored_proteins</code>; This is a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
that have been scored.</li>
</ol>

<p><strong>Attributes:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>pre_score_data</code></td>
        <td><code>list</code></td>
        <td><p>This is a list of <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Protein">Protein</a> objects
that contain <a class="autorefs autorefs-internal" href="#pyproteininference.physical.Psm">Psm</a> objects.</p></td>
      </tr>
      <tr>
        <td><code>data</code></td>
        <td><code>DataStore</code></td>
        <td><p><a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore</a> object.</p></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">Score</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Score class that contains methods to do a variety of scoring methods on the</span>
<span class="sd">    [Psm][pyproteininference.physical.Psm] objects</span>
<span class="sd">    contained inside of [Protein][pyproteininference.physical.Protein] objects.</span>

<span class="sd">    Methods in the class loop over each Protein object and creates a protein &quot;score&quot; variable using the Psm object</span>
<span class="sd">    scores.</span>

<span class="sd">    Methods score all proteins from `scoring_input` from [DataStore object][pyproteininference.datastore.DataStore].</span>
<span class="sd">    The PSM score that is used is determined from</span>
<span class="sd">    [create_scoring_input][pyproteininference.datastore.DataStore.create_scoring_input].</span>

<span class="sd">    Each scoring method will set the following attributes for</span>
<span class="sd">    the [DataStore object][pyproteininference.datastore.DataStore].</span>

<span class="sd">    1. `score_method`; This is the full name of the score method.</span>
<span class="sd">    2. `short_score_method`; This is the short name of the score method.</span>
<span class="sd">    3. `scored_proteins`; This is a list of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">    that have been scored.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        pre_score_data (list): This is a list of [Protein][pyproteininference.physical.Protein] objects</span>
<span class="sd">            that contain [Psm][pyproteininference.physical.Psm] objects.</span>
<span class="sd">        data (DataStore): [DataStore][pyproteininference.datastore.DataStore] object.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">BEST_PEPTIDE_PER_PROTEIN</span> <span class="o">=</span> <span class="s2">&quot;best_peptide_per_protein&quot;</span>
    <span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span> <span class="o">=</span> <span class="s2">&quot;iterative_downweighted_log&quot;</span>
    <span class="n">MULTIPLICATIVE_LOG</span> <span class="o">=</span> <span class="s2">&quot;multiplicative_log&quot;</span>
    <span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span> <span class="o">=</span> <span class="s2">&quot;downweighted_multiplicative_log&quot;</span>
    <span class="n">DOWNWEIGHTED_VERSION2</span> <span class="o">=</span> <span class="s2">&quot;downweighted_version2&quot;</span>
    <span class="n">TOP_TWO_COMBINED</span> <span class="o">=</span> <span class="s2">&quot;top_two_combined&quot;</span>
    <span class="n">GEOMETRIC_MEAN</span> <span class="o">=</span> <span class="s2">&quot;geometric_mean&quot;</span>
    <span class="n">ADDITIVE</span> <span class="o">=</span> <span class="s2">&quot;additive&quot;</span>

    <span class="n">SCORE_METHODS</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">BEST_PEPTIDE_PER_PROTEIN</span><span class="p">,</span>
        <span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span><span class="p">,</span>
        <span class="n">MULTIPLICATIVE_LOG</span><span class="p">,</span>
        <span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span><span class="p">,</span>
        <span class="n">DOWNWEIGHTED_VERSION2</span><span class="p">,</span>
        <span class="n">TOP_TWO_COMBINED</span><span class="p">,</span>
        <span class="n">GEOMETRIC_MEAN</span><span class="p">,</span>
        <span class="n">ADDITIVE</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">SHORT_BEST_PEPTIDE_PER_PROTEIN</span> <span class="o">=</span> <span class="s2">&quot;bppp&quot;</span>
    <span class="n">SHORT_ITERATIVE_DOWNWEIGHTED_LOG</span> <span class="o">=</span> <span class="s2">&quot;idwl&quot;</span>
    <span class="n">SHORT_MULTIPLICATIVE_LOG</span> <span class="o">=</span> <span class="s2">&quot;ml&quot;</span>
    <span class="n">SHORT_DOWNWEIGHTED_MULTIPLICATIVE_LOG</span> <span class="o">=</span> <span class="s2">&quot;dwml&quot;</span>
    <span class="n">SHORT_DOWNWEIGHTED_VERSION2</span> <span class="o">=</span> <span class="s2">&quot;dw2&quot;</span>
    <span class="n">SHORT_TOP_TWO_COMBINED</span> <span class="o">=</span> <span class="s2">&quot;ttc&quot;</span>
    <span class="n">SHORT_GEOMETRIC_MEAN</span> <span class="o">=</span> <span class="s2">&quot;gm&quot;</span>
    <span class="n">SHORT_ADDITIVE</span> <span class="o">=</span> <span class="s2">&quot;add&quot;</span>

    <span class="n">SHORT_SCORE_METHODS</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">SHORT_BEST_PEPTIDE_PER_PROTEIN</span><span class="p">,</span>
        <span class="n">SHORT_ITERATIVE_DOWNWEIGHTED_LOG</span><span class="p">,</span>
        <span class="n">SHORT_MULTIPLICATIVE_LOG</span><span class="p">,</span>
        <span class="n">SHORT_DOWNWEIGHTED_MULTIPLICATIVE_LOG</span><span class="p">,</span>
        <span class="n">SHORT_DOWNWEIGHTED_VERSION2</span><span class="p">,</span>
        <span class="n">SHORT_TOP_TWO_COMBINED</span><span class="p">,</span>
        <span class="n">SHORT_GEOMETRIC_MEAN</span><span class="p">,</span>
        <span class="n">SHORT_ADDITIVE</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">MULTIPLICATIVE_SCORE_TYPE</span> <span class="o">=</span> <span class="s2">&quot;multiplicative&quot;</span>
    <span class="n">ADDITIVE_SCORE_TYPE</span> <span class="o">=</span> <span class="s2">&quot;additive&quot;</span>

    <span class="n">SCORE_TYPES</span> <span class="o">=</span> <span class="p">[</span><span class="n">MULTIPLICATIVE_SCORE_TYPE</span><span class="p">,</span> <span class="n">ADDITIVE_SCORE_TYPE</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialization method for the Score class.</span>

<span class="sd">        Args:</span>
<span class="sd">            data (DataStore): [DataStore][pyproteininference.datastore.DataStore] object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If the variable `scoring_input` for the [DataStore][pyproteininference.datastore.DataStore]</span>
<span class="sd">                object is Empty &quot;[]&quot; or does not exist &quot;None&quot;.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">data</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">scoring_input</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;scoring input not found in data object - Please run &#39;create_scoring_input&#39; method from &quot;</span>
                <span class="s2">&quot;DataStore to run any scoring type&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">score_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score_method</span><span class="o">=</span><span class="s2">&quot;multiplicative_log&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method dispatches to the actual scoring method given a string input that is defined in the</span>
<span class="sd">        [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>

<span class="sd">        Args:</span>
<span class="sd">            score_method (str): This is a string that represents which scoring method to call.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: Will Error out if the score_method is not present in the constant `SCORE_METHODS`.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.score_psms(score_method=&quot;best_peptide_per_protein&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">score_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;score method &#39;</span><span class="si">{}</span><span class="s2">&#39; is not a proper method. Score method must be one of the following: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">score_method</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">BEST_PEPTIDE_PER_PROTEIN</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">best_peptide_per_protein</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">iterative_down_weighted_log</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">MULTIPLICATIVE_LOG</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">multiplicative_log</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">down_weighted_multiplicative_log</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_VERSION2</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">down_weighted_v2</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">TOP_TWO_COMBINED</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">top_two_combied</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">GEOMETRIC_MEAN</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">geometric_mean_log</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ADDITIVE</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">additive</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">best_peptide_per_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a best peptide per protein scoring scheme.</span>
<span class="sd">        The top scoring Psm for each protein is selected as the overall Protein object score.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.best_peptide_per_protein()</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with BPPP&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="n">score</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">val_list</span><span class="p">])</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
        <span class="c1"># Here do ascending sorting because a lower pep or q value is better</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">BEST_PEPTIDE_PER_PROTEIN</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_BEST_PEPTIDE_PER_PROTEIN</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">fishers_method</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a fishers method scoring scheme.</span>
<span class="sd">\</span>
<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.fishers_method()</span>

<span class="sd">         &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with fishers method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span> <span class="o">*</span> <span class="nb">sum</span><span class="p">([</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">val_list</span><span class="p">])</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
        <span class="c1"># Here reverse the sorting to descending because a higher score is better</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="s2">&quot;fishers_method&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="s2">&quot;fm&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Multiplicative Log scoring scheme.</span>
<span class="sd">        The selected Psm score from all the peptides per protein are multiplied together and we take -Log(X)</span>
<span class="sd">        of the multiplied Peptide scores.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.multiplicative_log()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Instead of making all_scores a list... make it a generator??</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with Multiplicative Log Method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="c1"># We create a generator of val_list...</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span><span class="p">)</span>
            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">MULTIPLICATIVE_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_MULTIPLICATIVE_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">down_weighted_multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Multiplicative Log scoring scheme.</span>
<span class="sd">        The selected PSM score from all the peptides per protein are multiplied together and</span>
<span class="sd">        then this number is divided by the set PSM scores mean raised to the number of peptides for that protein</span>
<span class="sd">        then we take -Log(X) of the following value.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.down_weighted_multiplicative_log()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">score_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">cur_scores</span> <span class="o">=</span> <span class="n">proteins</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">cur_scores</span><span class="p">:</span>
                <span class="n">score_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
        <span class="n">score_mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with DWML method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="c1"># Divide by the score mean raised to the length of the number of unique peptides for the protein</span>
            <span class="c1"># This is an attempt to normalize for number of peptides per protein</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span> <span class="o">/</span> <span class="p">(</span><span class="n">score_mean</span> <span class="o">**</span> <span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">)))</span>
            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_DOWNWEIGHTED_MULTIPLICATIVE_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">top_two_combied</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Top Two scoring scheme.</span>
<span class="sd">        The top two scores for each protein are multiplied together and we take -Log(X) of the multiplied value.</span>
<span class="sd">        If a protein only has 1 score/peptide, then we only do -Log(X) of the 1 peptide score.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.top_two_combied()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with Top Two Method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Try to combine the top two scores</span>
                <span class="c1"># Divide by 2 to attempt to normalize the value</span>
                <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">((</span><span class="n">val_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">val_list</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                <span class="c1"># If there is only 1 score/1 peptide then just use the 1 peptide provided</span>
                <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">val_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">TOP_TWO_COMBINED</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_TOP_TWO_COMBINED</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Downweighted Multiplicative Log scoring scheme.</span>
<span class="sd">        Each peptide is iteratively downweighted by raising the peptide QValue or PepValue to the</span>
<span class="sd">        following power (1/(1+index_number)).</span>
<span class="sd">        Where index_number is the peptide number per protein.</span>
<span class="sd">        Each score for a protein provides less and less weight iteratively.</span>

<span class="sd">        We also take -Log(X) of the final score here.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.down_weighted_v2()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with down weighted v2 method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="c1"># Here take each score and raise it to the power of (1/(1+index_number)).</span>
            <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
            <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
            <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">))</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_VERSION2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_DOWNWEIGHTED_VERSION2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">iterative_down_weighted_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Downweighted Multiplicative Log scoring scheme.</span>
<span class="sd">        Each peptide is iteratively downweighted by multiplying the peptide QValue or PepValue to</span>
<span class="sd">        the following (1+index_number).</span>
<span class="sd">        Where index_number is the peptide number per protein.</span>
<span class="sd">        Each score for a protein provides less and less weight iteratively.</span>

<span class="sd">        We also take -Log(X) of the final score here.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.iterative_down_weighted_log()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with IDWL method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="c1"># Here take each score and multiply it by its index number).</span>
            <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
            <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
            <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span><span class="p">)</span>
            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_ITERATIVE_DOWNWEIGHTED_LOG</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">geometric_mean_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a Geometric Mean scoring scheme.</span>

<span class="sd">        We also take -Log(X) of the final score here.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.geometric_mean_log()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins. with GML method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">psm_scores</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">psm_scores</span><span class="p">:</span>
                <span class="n">val_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">vals</span><span class="p">))</span>
                <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
                <span class="n">pre_log_score</span> <span class="o">=</span> <span class="n">combine</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">)))</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">pre_log_score</span><span class="p">)</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">GEOMETRIC_MEAN</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_GEOMETRIC_MEAN</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">iterative_down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The following method is an experimental method essentially used for future development of potential scoring</span>
<span class="sd">        schemes.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with iterative down weighted v2 method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="c1"># Here take each score and raise it to the power of (1/(1+index_number)).</span>
            <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
            <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
            <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">))</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="s2">&quot;iterative_downweighting2&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="s2">&quot;idw2&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>

    <span class="k">def</span> <span class="nf">additive</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses an additive scoring scheme.</span>
<span class="sd">        The method can only be used if a larger PSM score is a better PSM score such as the Percolator score.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.additive()</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with additive method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

            <span class="c1"># Take the sum of our scores</span>
            <span class="n">score</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">val_list</span><span class="p">)</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

        <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ADDITIVE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_ADDITIVE</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>



  <div class="doc doc-children">






























  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span></code>

  <span class="doc doc-properties">
      <small class="doc doc-property doc-property-special"><code>special</code></small>
  </span>

</h4>

    <div class="doc doc-contents ">

      <p>Initialization method for the Score class.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>data</strong> (<code>DataStore</code>)  <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore</a> object.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Exceptions:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><code>ValueError</code>  If the variable <code>scoring_input</code> for the <a class="autorefs autorefs-internal" href="#pyproteininference.datastore.DataStore">DataStore</a>
object is Empty "[]" or does not exist "None".</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialization method for the Score class.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (DataStore): [DataStore][pyproteininference.datastore.DataStore] object.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the variable `scoring_input` for the [DataStore][pyproteininference.datastore.DataStore]</span>
<span class="sd">            object is Empty &quot;[]&quot; or does not exist &quot;None&quot;.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data</span><span class="o">.</span><span class="n">scoring_input</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">scoring_input</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;scoring input not found in data object - Please run &#39;create_scoring_input&#39; method from &quot;</span>
            <span class="s2">&quot;DataStore to run any scoring type&quot;</span>
        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.additive">
<code class="highlight language-python"><span class="n">additive</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses an additive scoring scheme.
The method can only be used if a larger PSM score is a better PSM score such as the Percolator score.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">additive</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">additive</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses an additive scoring scheme.</span>
<span class="sd">    The method can only be used if a larger PSM score is a better PSM score such as the Percolator score.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.additive()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with additive method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="c1"># Take the sum of our scores</span>
        <span class="n">score</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">val_list</span><span class="p">)</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ADDITIVE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_ADDITIVE</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.best_peptide_per_protein">
<code class="highlight language-python"><span class="n">best_peptide_per_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a best peptide per protein scoring scheme.
The top scoring Psm for each protein is selected as the overall Protein object score.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">best_peptide_per_protein</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">best_peptide_per_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a best peptide per protein scoring scheme.</span>
<span class="sd">    The top scoring Psm for each protein is selected as the overall Protein object score.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.best_peptide_per_protein()</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with BPPP&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
        <span class="n">score</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">val_list</span><span class="p">])</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
    <span class="c1"># Here do ascending sorting because a lower pep or q value is better</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">BEST_PEPTIDE_PER_PROTEIN</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_BEST_PEPTIDE_PER_PROTEIN</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.down_weighted_multiplicative_log">
<code class="highlight language-python"><span class="n">down_weighted_multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Multiplicative Log scoring scheme.
The selected PSM score from all the peptides per protein are multiplied together and
then this number is divided by the set PSM scores mean raised to the number of peptides for that protein
then we take -Log(X) of the following value.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">down_weighted_multiplicative_log</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">down_weighted_multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Multiplicative Log scoring scheme.</span>
<span class="sd">    The selected PSM score from all the peptides per protein are multiplied together and</span>
<span class="sd">    then this number is divided by the set PSM scores mean raised to the number of peptides for that protein</span>
<span class="sd">    then we take -Log(X) of the following value.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.down_weighted_multiplicative_log()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">score_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">proteins</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">cur_scores</span> <span class="o">=</span> <span class="n">proteins</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">cur_scores</span><span class="p">:</span>
            <span class="n">score_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
    <span class="n">score_mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with DWML method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
        <span class="c1"># Divide by the score mean raised to the length of the number of unique peptides for the protein</span>
        <span class="c1"># This is an attempt to normalize for number of peptides per protein</span>
        <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span> <span class="o">/</span> <span class="p">(</span><span class="n">score_mean</span> <span class="o">**</span> <span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">)))</span>
        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_DOWNWEIGHTED_MULTIPLICATIVE_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.down_weighted_v2">
<code class="highlight language-python"><span class="n">down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Downweighted Multiplicative Log scoring scheme.
Each peptide is iteratively downweighted by raising the peptide QValue or PepValue to the
following power (1/(1+index_number)).
Where index_number is the peptide number per protein.
Each score for a protein provides less and less weight iteratively.</p>
<p>We also take -Log(X) of the final score here.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">down_weighted_v2</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Downweighted Multiplicative Log scoring scheme.</span>
<span class="sd">    Each peptide is iteratively downweighted by raising the peptide QValue or PepValue to the</span>
<span class="sd">    following power (1/(1+index_number)).</span>
<span class="sd">    Where index_number is the peptide number per protein.</span>
<span class="sd">    Each score for a protein provides less and less weight iteratively.</span>

<span class="sd">    We also take -Log(X) of the final score here.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.down_weighted_v2()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with down weighted v2 method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="c1"># Here take each score and raise it to the power of (1/(1+index_number)).</span>
        <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
        <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
        <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">))</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_VERSION2</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_DOWNWEIGHTED_VERSION2</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.fishers_method">
<code class="highlight language-python"><span class="n">fishers_method</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a fishers method scoring scheme.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">fishers_method</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code>    <span class="k">def</span> <span class="nf">fishers_method</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method uses a fishers method scoring scheme.</span>
<span class="sd">\</span>
<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">            &gt;&gt;&gt; score.fishers_method()</span>

<span class="sd">         &quot;&quot;&quot;</span>

        <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with fishers method&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
            <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span> <span class="o">*</span> <span class="nb">sum</span><span class="p">([</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">val_list</span><span class="p">])</span>

            <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

            <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
        <span class="c1"># Here reverse the sorting to descending because a higher score is better</span>
        <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="s2">&quot;fishers_method&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="s2">&quot;fm&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.geometric_mean_log">
<code class="highlight language-python"><span class="n">geometric_mean_log</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Geometric Mean scoring scheme.</p>
<p>We also take -Log(X) of the final score here.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">geometric_mean_log</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">geometric_mean_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Geometric Mean scoring scheme.</span>

<span class="sd">    We also take -Log(X) of the final score here.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.geometric_mean_log()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins. with GML method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">psm_scores</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">psm_scores</span><span class="p">:</span>
            <span class="n">val_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">vals</span><span class="p">))</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
            <span class="n">pre_log_score</span> <span class="o">=</span> <span class="n">combine</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">)))</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">pre_log_score</span><span class="p">)</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">GEOMETRIC_MEAN</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_GEOMETRIC_MEAN</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.iterative_down_weighted_log">
<code class="highlight language-python"><span class="n">iterative_down_weighted_log</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Downweighted Multiplicative Log scoring scheme.
Each peptide is iteratively downweighted by multiplying the peptide QValue or PepValue to
the following (1+index_number).
Where index_number is the peptide number per protein.
Each score for a protein provides less and less weight iteratively.</p>
<p>We also take -Log(X) of the final score here.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">iterative_down_weighted_log</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">iterative_down_weighted_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Downweighted Multiplicative Log scoring scheme.</span>
<span class="sd">    Each peptide is iteratively downweighted by multiplying the peptide QValue or PepValue to</span>
<span class="sd">    the following (1+index_number).</span>
<span class="sd">    Where index_number is the peptide number per protein.</span>
<span class="sd">    Each score for a protein provides less and less weight iteratively.</span>

<span class="sd">    We also take -Log(X) of the final score here.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.iterative_down_weighted_log()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with IDWL method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="c1"># Here take each score and multiply it by its index number).</span>
        <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
        <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
        <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
        <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span><span class="p">)</span>
        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_ITERATIVE_DOWNWEIGHTED_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.iterative_down_weighted_v2">
<code class="highlight language-python"><span class="n">iterative_down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>The following method is an experimental method essentially used for future development of potential scoring
schemes.</p>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">iterative_down_weighted_v2</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The following method is an experimental method essentially used for future development of potential scoring</span>
<span class="sd">    schemes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with iterative down weighted v2 method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="c1"># Here take each score and raise it to the power of (1/(1+index_number)).</span>
        <span class="c1"># This downweights each successive score by reducing its weight in a decreasing fashion</span>
        <span class="c1"># Basically, each score for a protein will provide less and less weight iteratively</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">val_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">val_list</span><span class="p">))]</span>
        <span class="c1"># val_list = [val_list[x]**(1/float(1+(float(x)/10))) for x in range(len(val_list))]</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">))</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="s2">&quot;iterative_downweighting2&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="s2">&quot;idw2&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.multiplicative_log">
<code class="highlight language-python"><span class="n">multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Multiplicative Log scoring scheme.
The selected Psm score from all the peptides per protein are multiplied together and we take -Log(X)
of the multiplied Peptide scores.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">multiplicative_log</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">multiplicative_log</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Multiplicative Log scoring scheme.</span>
<span class="sd">    The selected Psm score from all the peptides per protein are multiplied together and we take -Log(X)</span>
<span class="sd">    of the multiplied Peptide scores.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.multiplicative_log()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Instead of making all_scores a list... make it a generator??</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with Multiplicative Log Method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="c1"># We create a generator of val_list...</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="n">combine</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="n">y</span><span class="p">,</span> <span class="n">val_list</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">combine</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">combine</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">float_info</span><span class="o">.</span><span class="n">min</span>
        <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">combine</span><span class="p">)</span>
        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>

        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">MULTIPLICATIVE_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_MULTIPLICATIVE_LOG</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.score_psms">
<code class="highlight language-python"><span class="n">score_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;multiplicative_log&#39;</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method dispatches to the actual scoring method given a string input that is defined in the
<a class="autorefs autorefs-internal" href="#pyproteininference.parameters.ProteinInferenceParameter">ProteinInferenceParameter</a> object.</p>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><strong>score_method</strong> (<code>str</code>)  This is a string that represents which scoring method to call.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Exceptions:</th>
      <td class="field-body">
        <ul class="first simple">
            <li><p><code>ValueError</code>  Will Error out if the score_method is not present in the constant <code>SCORE_METHODS</code>.</p></li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">score_psms</span><span class="p">(</span><span class="n">score_method</span><span class="o">=</span><span class="s2">&quot;best_peptide_per_protein&quot;</span><span class="p">)</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">score_psms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score_method</span><span class="o">=</span><span class="s2">&quot;multiplicative_log&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method dispatches to the actual scoring method given a string input that is defined in the</span>
<span class="sd">    [ProteinInferenceParameter][pyproteininference.parameters.ProteinInferenceParameter] object.</span>

<span class="sd">    Args:</span>
<span class="sd">        score_method (str): This is a string that represents which scoring method to call.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: Will Error out if the score_method is not present in the constant `SCORE_METHODS`.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.score_psms(score_method=&quot;best_peptide_per_protein&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">score_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;score method &#39;</span><span class="si">{}</span><span class="s2">&#39; is not a proper method. Score method must be one of the following: &#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">score_method</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">SCORE_METHODS</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">BEST_PEPTIDE_PER_PROTEIN</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">best_peptide_per_protein</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ITERATIVE_DOWNWEIGHTED_LOG</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iterative_down_weighted_log</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">MULTIPLICATIVE_LOG</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">multiplicative_log</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_MULTIPLICATIVE_LOG</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">down_weighted_multiplicative_log</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">DOWNWEIGHTED_VERSION2</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">down_weighted_v2</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">TOP_TWO_COMBINED</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">top_two_combied</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">GEOMETRIC_MEAN</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">geometric_mean_log</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">score_method</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">ADDITIVE</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">additive</span><span class="p">()</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-method">



<h4 class="doc doc-heading" id="pyproteininference.scoring.Score.top_two_combied">
<code class="highlight language-python"><span class="n">top_two_combied</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></code>


</h4>

    <div class="doc doc-contents ">

      <p>This method uses a Top Two scoring scheme.
The top two scores for each protein are multiplied together and we take -Log(X) of the multiplied value.
If a protein only has 1 score/peptide, then we only do -Log(X) of the 1 peptide score.</p>

<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span class="o">&gt;&gt;&gt;</span> <span class="n">score</span> <span class="o">=</span> <span class="n">pyproteininference</span><span class="o">.</span><span class="n">scoring</span><span class="o">.</span><span class="n">Score</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">score</span><span class="o">.</span><span class="n">top_two_combied</span><span class="p">()</span>
</code></pre></div>

        <details class="quote">
          <summary>Source code in <code>pyproteininference/scoring.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">top_two_combied</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This method uses a Top Two scoring scheme.</span>
<span class="sd">    The top two scores for each protein are multiplied together and we take -Log(X) of the multiplied value.</span>
<span class="sd">    If a protein only has 1 score/peptide, then we only do -Log(X) of the 1 peptide score.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; score = pyproteininference.scoring.Score(data=data)</span>
<span class="sd">        &gt;&gt;&gt; score.top_two_combied()</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">all_scores</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Scoring Proteins with Top Two Method&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pre_score_data</span><span class="p">:</span>
        <span class="n">val_list</span> <span class="o">=</span> <span class="n">protein</span><span class="o">.</span><span class="n">get_psm_scores</span><span class="p">()</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Try to combine the top two scores</span>
            <span class="c1"># Divide by 2 to attempt to normalize the value</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">((</span><span class="n">val_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">val_list</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="c1"># If there is only 1 score/1 peptide then just use the 1 peptide provided</span>
            <span class="n">score</span> <span class="o">=</span> <span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">val_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="n">protein</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="n">all_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>

    <span class="c1"># Higher score is better as a smaller q or pep in a -log will give a larger value</span>
    <span class="n">all_scores</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_scores</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">k</span><span class="o">.</span><span class="n">score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">TOP_TWO_COMBINED</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">short_protein_score</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SHORT_TOP_TWO_COMBINED</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">scored_proteins</span> <span class="o">=</span> <span class="n">all_scores</span>
</code></pre></div>
        </details>
    </div>

  </div>





  </div>

    </div>

  </div>







  </div>

    </div>

  </div>




  </div>

    </div>

  </div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../supplementary/" class="btn btn-neutral float-left" title="Supplementary Information"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../supplementary/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
